{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Summary:\n",
    "\n",
    "This notebook is the testing ground for a script to analyze snRNA data. Specifically, this will be designed for nPOD samples.\n",
    "\n",
    "This is a cleaned up copy of the third version of the sample specific cell QC filtering methods we are establishing. Here we filter the raw matrix by thresholds (RNA genes >= 500, ATAC fragments >= 1000), remove multiplets (called by CellRanger), and then perform SoupX to adjust the RNA counts based on background contamination. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "Module(leidenalg)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#always run this first!!\n",
    "#essential reticulate functions that allow us to use python packages in R\n",
    "#you'll need a conda env with 'leidenalg' and 'pandas' installed to do this\n",
    "#then route reticulate to the python installed in that conda env with the below functions\n",
    "Sys.setenv(RETICULATE_PYTHON=\"/home/.conda/envs/decontX_reticulate/bin/python\")\n",
    "library(reticulate)\n",
    "reticulate::use_python(\"/home/.conda/envs/decontX_reticulate/bin/python\")\n",
    "reticulate::use_condaenv(\"/home/.conda/envs/decontX_reticulate\")\n",
    "reticulate::py_module_available(module='leidenalg') #needs to be TRUE\n",
    "reticulate::import('leidenalg') #good to make sure this doesn't error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(Matrix)\n",
    "library(harmony)\n",
    "library(stringr)\n",
    "library(gridExtra)\n",
    "\n",
    "source('231121_WE_Graphing_Clustering_Functions.R')\n",
    "\n",
    "#suppressMessages(library(hdf5r))\n",
    "#suppressMessages(library(data.table))\n",
    "#suppressMessages(library(ggpubr))\n",
    "#\n",
    "#\n",
    "#warnLevel <- getOption('warn')\n",
    "#options(warn = -1)\n",
    "#\n",
    "#library(DoubletFinder)\n",
    "#library(scater)\n",
    "##library(celda)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.1 (2023-06-16)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.2 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /home/welison/.conda/envs/mamba/envs/seurat5.0_decontx/lib/libopenblasp-r0.3.21.so;  LAPACK version 3.9.0\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: America/Los_Angeles\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       "[1] stats4    stats     graphics  grDevices utils     datasets  methods  \n",
       "[8] base     \n",
       "\n",
       "other attached packages:\n",
       " [1] gridExtra_2.3             stringr_1.5.1            \n",
       " [3] harmony_1.2.0             Rcpp_1.0.11              \n",
       " [5] Matrix_1.6-4              ggplot2_3.4.4            \n",
       " [7] dplyr_1.1.4               EnsDb.Hsapiens.v86_2.99.0\n",
       " [9] ensembldb_2.26.0          AnnotationFilter_1.26.0  \n",
       "[11] GenomicFeatures_1.54.1    AnnotationDbi_1.64.1     \n",
       "[13] Biobase_2.62.0            GenomicRanges_1.54.1     \n",
       "[15] GenomeInfoDb_1.38.1       IRanges_2.36.0           \n",
       "[17] S4Vectors_0.40.2          BiocGenerics_0.48.1      \n",
       "[19] Signac_1.12.0             Seurat_5.0.1             \n",
       "[21] SeuratObject_5.0.1        sp_2.1-2                 \n",
       "[23] reticulate_1.34.0        \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.21            splines_4.3.1              \n",
       "  [3] later_1.3.2                 BiocIO_1.12.0              \n",
       "  [5] pbdZMQ_0.3-10               bitops_1.0-7               \n",
       "  [7] filelock_1.0.3              tibble_3.2.1               \n",
       "  [9] polyclip_1.10-6             XML_3.99-0.16              \n",
       " [11] fastDummies_1.7.3           lifecycle_1.0.4            \n",
       " [13] globals_0.16.2              lattice_0.22-5             \n",
       " [15] MASS_7.3-60                 magrittr_2.0.3             \n",
       " [17] plotly_4.10.3               yaml_2.3.7                 \n",
       " [19] httpuv_1.6.13               sctransform_0.4.1          \n",
       " [21] spam_2.10-0                 spatstat.sparse_3.0-3      \n",
       " [23] cowplot_1.1.1               pbapply_1.7-2              \n",
       " [25] DBI_1.1.3                   RColorBrewer_1.1-3         \n",
       " [27] abind_1.4-5                 zlibbioc_1.48.0            \n",
       " [29] Rtsne_0.17                  purrr_1.0.2                \n",
       " [31] RCurl_1.98-1.13             rappdirs_0.3.3             \n",
       " [33] GenomeInfoDbData_1.2.11     ggrepel_0.9.4              \n",
       " [35] irlba_2.3.5.1               listenv_0.9.0              \n",
       " [37] spatstat.utils_3.0-4        goftest_1.2-3              \n",
       " [39] RSpectra_0.16-1             spatstat.random_3.2-2      \n",
       " [41] fitdistrplus_1.1-11         parallelly_1.36.0          \n",
       " [43] DelayedArray_0.28.0         leiden_0.4.3.1             \n",
       " [45] codetools_0.2-19            RcppRoll_0.3.0             \n",
       " [47] xml2_1.3.6                  tidyselect_1.2.0           \n",
       " [49] matrixStats_1.1.0           BiocFileCache_2.10.1       \n",
       " [51] base64enc_0.1-3             spatstat.explore_3.2-5     \n",
       " [53] GenomicAlignments_1.38.0    jsonlite_1.8.8             \n",
       " [55] ellipsis_0.3.2              progressr_0.14.0           \n",
       " [57] ggridges_0.5.4              survival_3.5-7             \n",
       " [59] tools_4.3.1                 progress_1.2.3             \n",
       " [61] ica_1.0-3                   glue_1.6.2                 \n",
       " [63] SparseArray_1.2.2           MatrixGenerics_1.14.0      \n",
       " [65] IRdisplay_1.1               withr_2.5.2                \n",
       " [67] fastmap_1.1.1               fansi_1.0.6                \n",
       " [69] digest_0.6.33               R6_2.5.1                   \n",
       " [71] mime_0.12                   colorspace_2.1-0           \n",
       " [73] scattermore_1.2             tensor_1.5                 \n",
       " [75] spatstat.data_3.0-3         biomaRt_2.58.0             \n",
       " [77] RSQLite_2.3.4               utf8_1.2.4                 \n",
       " [79] tidyr_1.3.0                 generics_0.1.3             \n",
       " [81] data.table_1.14.10          rtracklayer_1.62.0         \n",
       " [83] S4Arrays_1.2.0              prettyunits_1.2.0          \n",
       " [85] httr_1.4.7                  htmlwidgets_1.6.4          \n",
       " [87] uwot_0.1.16                 pkgconfig_2.0.3            \n",
       " [89] gtable_0.3.4                blob_1.2.4                 \n",
       " [91] lmtest_0.9-40               XVector_0.42.0             \n",
       " [93] htmltools_0.5.7             dotCall64_1.1-1            \n",
       " [95] ProtGenerics_1.34.0         scales_1.3.0               \n",
       " [97] png_0.1-8                   rjson_0.2.21               \n",
       " [99] reshape2_1.4.4              uuid_1.1-1                 \n",
       "[101] nlme_3.1-164                curl_5.2.0                 \n",
       "[103] repr_1.1.6                  zoo_1.8-12                 \n",
       "[105] cachem_1.0.8                KernSmooth_2.23-22         \n",
       "[107] parallel_4.3.1              miniUI_0.1.1.1             \n",
       "[109] restfulr_0.0.15             pillar_1.9.0               \n",
       "[111] grid_4.3.1                  vctrs_0.6.5                \n",
       "[113] RANN_2.6.1                  promises_1.2.1             \n",
       "[115] dbplyr_2.4.0                xtable_1.8-4               \n",
       "[117] cluster_2.1.6               evaluate_0.23              \n",
       "[119] cli_3.6.2                   compiler_4.3.1             \n",
       "[121] Rsamtools_2.18.0            rlang_1.1.2                \n",
       "[123] crayon_1.5.2                future.apply_1.11.0        \n",
       "[125] plyr_1.8.9                  stringi_1.8.2              \n",
       "[127] viridisLite_0.4.2           deldir_2.0-2               \n",
       "[129] BiocParallel_1.36.0         munsell_0.5.0              \n",
       "[131] Biostrings_2.70.1           lazyeval_0.2.2             \n",
       "[133] spatstat.geom_3.2-7         IRkernel_1.3.2             \n",
       "[135] RcppHNSW_0.5.0              hms_1.1.3                  \n",
       "[137] patchwork_1.1.3             bit64_4.0.5                \n",
       "[139] future_1.33.0               KEGGREST_1.42.0            \n",
       "[141] shiny_1.8.0                 SummarizedExperiment_1.32.0\n",
       "[143] ROCR_1.0-11                 igraph_1.6.0               \n",
       "[145] memoise_2.0.1               fastmatch_1.1-4            \n",
       "[147] bit_4.0.5                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read in data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is the latest pooled data for nPOD  \n",
    "\n",
    "Cellranger and initial clean up (500 Features RNA, 1000 atac fragments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'MM_1349_1_2_MM_1345_1_2'</li><li>'MM_1350_1_2_MM_1346_1_2'</li><li>'MM_1351_1_2_MM_1347_1_2'</li><li>'MM_1352_1_2_MM_1348_1_2'</li><li>'QY_2312_1_2_QY_2311_1_2'</li><li>'QY_2314_1_2_QY_2313_1_2'</li><li>'QY_2316_1_2_QY_2315_1_2'</li><li>'QY_2318_1_2_QY_2317_1_2'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'MM\\_1349\\_1\\_2\\_MM\\_1345\\_1\\_2'\n",
       "\\item 'MM\\_1350\\_1\\_2\\_MM\\_1346\\_1\\_2'\n",
       "\\item 'MM\\_1351\\_1\\_2\\_MM\\_1347\\_1\\_2'\n",
       "\\item 'MM\\_1352\\_1\\_2\\_MM\\_1348\\_1\\_2'\n",
       "\\item 'QY\\_2312\\_1\\_2\\_QY\\_2311\\_1\\_2'\n",
       "\\item 'QY\\_2314\\_1\\_2\\_QY\\_2313\\_1\\_2'\n",
       "\\item 'QY\\_2316\\_1\\_2\\_QY\\_2315\\_1\\_2'\n",
       "\\item 'QY\\_2318\\_1\\_2\\_QY\\_2317\\_1\\_2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'MM_1349_1_2_MM_1345_1_2'\n",
       "2. 'MM_1350_1_2_MM_1346_1_2'\n",
       "3. 'MM_1351_1_2_MM_1347_1_2'\n",
       "4. 'MM_1352_1_2_MM_1348_1_2'\n",
       "5. 'QY_2312_1_2_QY_2311_1_2'\n",
       "6. 'QY_2314_1_2_QY_2313_1_2'\n",
       "7. 'QY_2316_1_2_QY_2315_1_2'\n",
       "8. 'QY_2318_1_2_QY_2317_1_2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"MM_1349_1_2_MM_1345_1_2\" \"MM_1350_1_2_MM_1346_1_2\"\n",
       "[3] \"MM_1351_1_2_MM_1347_1_2\" \"MM_1352_1_2_MM_1348_1_2\"\n",
       "[5] \"QY_2312_1_2_QY_2311_1_2\" \"QY_2314_1_2_QY_2313_1_2\"\n",
       "[7] \"QY_2316_1_2_QY_2315_1_2\" \"QY_2318_1_2_QY_2317_1_2\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Set some values and load sample list\n",
    "dataset <- 'nPOD'\n",
    "modality <- 'Pooled'\n",
    "\n",
    "samp_dir <- paste0(\"/nfs/lab/projects/mega_pancreas/data/cellranger/\",dataset,\"/\", modality, \"/\")\n",
    "\n",
    "\n",
    "samples <- list.dirs(samp_dir, full.names = FALSE, recursive = FALSE)\n",
    "samples = samples[samples != \"intermediates\"]\n",
    "\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>34</li><li>12</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 34\n",
       "\\item 12\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 34\n",
       "2. 12\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 34 12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 34 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>donor_demux</th><th scope=col>Donor.Type</th><th scope=col>AAb.status</th><th scope=col>Age..yrs.</th><th scope=col>NoAbs</th><th scope=col>Gender</th><th scope=col>Race</th><th scope=col>BMI</th><th scope=col>HbA1C</th><th scope=col>GRS1..European.Hispanic</th><th scope=col>GRS.High.Low...0.28</th><th scope=col>Bullshit.Percentiles</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>6301  </td><td>Autoab Pos </td><td>GADA+       </td><td>26.00</td><td>1</td><td>Male  </td><td>African Am     </td><td>32.08</td><td>5.5</td><td>0.187200</td><td>low </td><td>0.0200</td></tr>\n",
       "\t<tr><td>6303  </td><td>Autoab Pos </td><td>GADA+       </td><td>22.00</td><td>1</td><td>Male  </td><td>Caucasian      </td><td>31.90</td><td>5.4</td><td>0.282700</td><td>high</td><td>0.6000</td></tr>\n",
       "\t<tr><td>6314  </td><td>Autoab Pos </td><td>GADA+       </td><td>21.00</td><td>1</td><td>Male  </td><td>Caucasian      </td><td>23.80</td><td>   </td><td>0.244000</td><td>low </td><td>0.1300</td></tr>\n",
       "\t<tr><td>6347  </td><td>Autoab Pos </td><td>mIAA+*      </td><td> 8.50</td><td>1</td><td>Male  </td><td>Caucasian      </td><td>19.50</td><td>   </td><td>0.160670</td><td>low </td><td>0.0100</td></tr>\n",
       "\t<tr><td>6397  </td><td>Autoab Pos </td><td>GADA+       </td><td>21.16</td><td>1</td><td>Female</td><td>Caucasian      </td><td>29.60</td><td>6  </td><td>0.174830</td><td>low </td><td>0.0150</td></tr>\n",
       "\t<tr><td>6400  </td><td>Autoab Pos </td><td>GADA+       </td><td>25.15</td><td>1</td><td>Male  </td><td>Hispanic/Latino</td><td>22.20</td><td>5.5</td><td>0.268833</td><td>low </td><td>0.3500</td></tr>\n",
       "\t<tr><td>6421  </td><td>Autoab Pos </td><td>GADA+       </td><td> 6.73</td><td>1</td><td>Male  </td><td>Hispanic/Latino</td><td>17.90</td><td>5.6</td><td>0.220333</td><td>low </td><td>0.0400</td></tr>\n",
       "\t<tr><td>6424  </td><td>Autoab Pos </td><td>GADA+ mIAA+*</td><td>17.65</td><td>2</td><td>Male  </td><td>Caucasian      </td><td>51.40</td><td>5.8</td><td>0.344830</td><td>high</td><td>0.9900</td></tr>\n",
       "\t<tr><td>6433  </td><td>Autoab Pos </td><td>GADA+       </td><td>23.96</td><td>1</td><td>Male  </td><td>Hispanic/Latino</td><td>30.80</td><td>5.3</td><td>0.268330</td><td>low </td><td>0.3450</td></tr>\n",
       "\t<tr><td>6437  </td><td>Autoab Pos </td><td>GADA+       </td><td>24.73</td><td>1</td><td>Male  </td><td>Caucasian      </td><td>24.17</td><td>5.7</td><td>0.266800</td><td>low </td><td>0.3200</td></tr>\n",
       "\t<tr><td>6483  </td><td>Autoab Pos </td><td>GADA+       </td><td>30.46</td><td>1</td><td>Female</td><td>Caucasian      </td><td>20.80</td><td>5.5</td><td>0.263500</td><td>low </td><td>0.3000</td></tr>\n",
       "\t<tr><td>6494  </td><td>Autoab Pos </td><td>mIAA+*      </td><td> 8.83</td><td>1</td><td>Male  </td><td>African Am     </td><td>15.20</td><td>4.9</td><td>0.205500</td><td>low </td><td>0.0250</td></tr>\n",
       "\t<tr><td>6496  </td><td>Autoab Pos </td><td>GADA+       </td><td>26.11</td><td>1</td><td>Male  </td><td>African Am     </td><td>46.10</td><td>5.7</td><td>0.225000</td><td>low </td><td>0.0450</td></tr>\n",
       "\t<tr><td>6517  </td><td>Autoab Pos </td><td>GADA+       </td><td>22.14</td><td>1</td><td>Male  </td><td>Caucasian      </td><td>26.60</td><td>5.6</td><td>0.156167</td><td>low </td><td>0.0010</td></tr>\n",
       "\t<tr><td>6532  </td><td>Autoab Pos </td><td>GADA+       </td><td>20.04</td><td>1</td><td>Male  </td><td>Hispanic/Latino</td><td>23.90</td><td>5.9</td><td>0.212667</td><td>low </td><td>0.0300</td></tr>\n",
       "\t<tr><td>6538  </td><td>Autoab Pos </td><td>GADA+       </td><td>19.14</td><td>1</td><td>Male  </td><td>Caucasian      </td><td>32.90</td><td>5.8</td><td>0.257500</td><td>low </td><td>0.2500</td></tr>\n",
       "\t<tr><td>6553  </td><td>Autoab Pos </td><td>mIAA+*      </td><td>12.34</td><td>1</td><td>Female</td><td>Hispanic/Latino</td><td>24.80</td><td>8.4</td><td>0.333667</td><td>high</td><td>0.9800</td></tr>\n",
       "\t<tr><td>6558  </td><td>Autoab Pos </td><td>GADA+       </td><td>21.69</td><td>1</td><td>Female</td><td>African Am     </td><td>27.80</td><td>4.4</td><td>0.268833</td><td>low </td><td>0.3500</td></tr>\n",
       "\t<tr><td>6562  </td><td>Autoab Pos </td><td>GADA+       </td><td>29.77</td><td>1</td><td>Female</td><td>Caucasian      </td><td>18.20</td><td>5.6</td><td>0.253330</td><td>low </td><td>0.2400</td></tr>\n",
       "\t<tr><td>CV02  </td><td>Autoab Pos </td><td>GADA+ ZnT8A+</td><td>15.93</td><td>2</td><td>Male  </td><td>Caucasian      </td><td>22.00</td><td>5.1</td><td>0.220167</td><td>low </td><td>0.0400</td></tr>\n",
       "\t<tr><td>HDL011</td><td>Autoab Pos </td><td>GADA+       </td><td> 8.00</td><td>1</td><td>Female</td><td>African Am     </td><td>12.00</td><td>n/a</td><td>0.210000</td><td>low </td><td>0.0250</td></tr>\n",
       "\t<tr><td>6569  </td><td>Autoab Pos </td><td>GADA+       </td><td>20.00</td><td>1</td><td>Female</td><td>HIspanic/Latino</td><td>24.10</td><td>   </td><td>0.236167</td><td>low </td><td>0.0500</td></tr>\n",
       "\t<tr><td>HDL112</td><td>No diabetes</td><td>Aab-        </td><td> 3.72</td><td>0</td><td>Female</td><td>Caucasian      </td><td>17.00</td><td>5.1</td><td>0.291000</td><td>high</td><td>0.6000</td></tr>\n",
       "\t<tr><td>6543  </td><td>No diabetes</td><td>Aab-        </td><td> 3.50</td><td>0</td><td>Male  </td><td>Caucasian      </td><td>17.40</td><td>5.3</td><td>0.211667</td><td>low </td><td>0.0275</td></tr>\n",
       "\t<tr><td>6504  </td><td>No diabetes</td><td>Aab-        </td><td> 7.61</td><td>0</td><td>Female</td><td>Hispanic/Latino</td><td>16.60</td><td>5.2</td><td>0.318667</td><td>high</td><td>0.8500</td></tr>\n",
       "\t<tr><td>6232  </td><td>No diabetes</td><td>Aab-        </td><td>14.00</td><td>0</td><td>Female</td><td>Caucasian      </td><td>20.80</td><td>n/a</td><td>0.131833</td><td>low </td><td>0.0001</td></tr>\n",
       "\t<tr><td>6516  </td><td>No diabetes</td><td>Aab-        </td><td>20.75</td><td>0</td><td>Male  </td><td>Caucasian      </td><td>28.80</td><td>5.5</td><td>0.196167</td><td>low </td><td>0.0250</td></tr>\n",
       "\t<tr><td>6509  </td><td>No diabetes</td><td>Aab-        </td><td>18.72</td><td>0</td><td>Male  </td><td>Caucasian      </td><td>25.80</td><td>5.4</td><td>0.198667</td><td>low </td><td>0.0250</td></tr>\n",
       "\t<tr><td>6474  </td><td>No diabetes</td><td>Aab-        </td><td>21.12</td><td>0</td><td>Male  </td><td>African Am     </td><td>29.00</td><td>5.3</td><td>0.260833</td><td>low </td><td>0.2800</td></tr>\n",
       "\t<tr><td>6548  </td><td>No diabetes</td><td>Aab-        </td><td>20.24</td><td>0</td><td>Male  </td><td>Caucasian      </td><td>23.80</td><td>5.7</td><td>0.324500</td><td>high</td><td>0.9000</td></tr>\n",
       "\t<tr><td>6539  </td><td>No diabetes</td><td>Aab-        </td><td>24.60</td><td>0</td><td>Male  </td><td>Hispanic/Latino</td><td>19.30</td><td>5.7</td><td>0.276833</td><td>low </td><td>0.4500</td></tr>\n",
       "\t<tr><td>6455  </td><td>No diabetes</td><td>Aab-        </td><td> 6.76</td><td>0</td><td>Male  </td><td>Caucasian      </td><td>17.69</td><td>5.3</td><td>0.254833</td><td>low </td><td>0.2400</td></tr>\n",
       "\t<tr><td>6560  </td><td>No diabetes</td><td>Aab-        </td><td>34.36</td><td>0</td><td>Female</td><td>Caucasian      </td><td>26.60</td><td>5.9</td><td>0.300000</td><td>high</td><td>0.7500</td></tr>\n",
       "\t<tr><td>6392  </td><td>No diabetes</td><td>Aab-        </td><td>14.10</td><td>0</td><td>Male  </td><td>African Am     </td><td>23.60</td><td>5.7</td><td>0.224833</td><td>low </td><td>0.0450</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 34 × 12\n",
       "\\begin{tabular}{llllllllllll}\n",
       " donor\\_demux & Donor.Type & AAb.status & Age..yrs. & NoAbs & Gender & Race & BMI & HbA1C & GRS1..European.Hispanic & GRS.High.Low...0.28 & Bullshit.Percentiles\\\\\n",
       " <chr> & <chr> & <chr> & <dbl> & <int> & <chr> & <chr> & <dbl> & <chr> & <dbl> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\t 6301   & Autoab Pos  & GADA+        & 26.00 & 1 & Male   & African Am      & 32.08 & 5.5 & 0.187200 & low  & 0.0200\\\\\n",
       "\t 6303   & Autoab Pos  & GADA+        & 22.00 & 1 & Male   & Caucasian       & 31.90 & 5.4 & 0.282700 & high & 0.6000\\\\\n",
       "\t 6314   & Autoab Pos  & GADA+        & 21.00 & 1 & Male   & Caucasian       & 23.80 &     & 0.244000 & low  & 0.1300\\\\\n",
       "\t 6347   & Autoab Pos  & mIAA+*       &  8.50 & 1 & Male   & Caucasian       & 19.50 &     & 0.160670 & low  & 0.0100\\\\\n",
       "\t 6397   & Autoab Pos  & GADA+        & 21.16 & 1 & Female & Caucasian       & 29.60 & 6   & 0.174830 & low  & 0.0150\\\\\n",
       "\t 6400   & Autoab Pos  & GADA+        & 25.15 & 1 & Male   & Hispanic/Latino & 22.20 & 5.5 & 0.268833 & low  & 0.3500\\\\\n",
       "\t 6421   & Autoab Pos  & GADA+        &  6.73 & 1 & Male   & Hispanic/Latino & 17.90 & 5.6 & 0.220333 & low  & 0.0400\\\\\n",
       "\t 6424   & Autoab Pos  & GADA+ mIAA+* & 17.65 & 2 & Male   & Caucasian       & 51.40 & 5.8 & 0.344830 & high & 0.9900\\\\\n",
       "\t 6433   & Autoab Pos  & GADA+        & 23.96 & 1 & Male   & Hispanic/Latino & 30.80 & 5.3 & 0.268330 & low  & 0.3450\\\\\n",
       "\t 6437   & Autoab Pos  & GADA+        & 24.73 & 1 & Male   & Caucasian       & 24.17 & 5.7 & 0.266800 & low  & 0.3200\\\\\n",
       "\t 6483   & Autoab Pos  & GADA+        & 30.46 & 1 & Female & Caucasian       & 20.80 & 5.5 & 0.263500 & low  & 0.3000\\\\\n",
       "\t 6494   & Autoab Pos  & mIAA+*       &  8.83 & 1 & Male   & African Am      & 15.20 & 4.9 & 0.205500 & low  & 0.0250\\\\\n",
       "\t 6496   & Autoab Pos  & GADA+        & 26.11 & 1 & Male   & African Am      & 46.10 & 5.7 & 0.225000 & low  & 0.0450\\\\\n",
       "\t 6517   & Autoab Pos  & GADA+        & 22.14 & 1 & Male   & Caucasian       & 26.60 & 5.6 & 0.156167 & low  & 0.0010\\\\\n",
       "\t 6532   & Autoab Pos  & GADA+        & 20.04 & 1 & Male   & Hispanic/Latino & 23.90 & 5.9 & 0.212667 & low  & 0.0300\\\\\n",
       "\t 6538   & Autoab Pos  & GADA+        & 19.14 & 1 & Male   & Caucasian       & 32.90 & 5.8 & 0.257500 & low  & 0.2500\\\\\n",
       "\t 6553   & Autoab Pos  & mIAA+*       & 12.34 & 1 & Female & Hispanic/Latino & 24.80 & 8.4 & 0.333667 & high & 0.9800\\\\\n",
       "\t 6558   & Autoab Pos  & GADA+        & 21.69 & 1 & Female & African Am      & 27.80 & 4.4 & 0.268833 & low  & 0.3500\\\\\n",
       "\t 6562   & Autoab Pos  & GADA+        & 29.77 & 1 & Female & Caucasian       & 18.20 & 5.6 & 0.253330 & low  & 0.2400\\\\\n",
       "\t CV02   & Autoab Pos  & GADA+ ZnT8A+ & 15.93 & 2 & Male   & Caucasian       & 22.00 & 5.1 & 0.220167 & low  & 0.0400\\\\\n",
       "\t HDL011 & Autoab Pos  & GADA+        &  8.00 & 1 & Female & African Am      & 12.00 & n/a & 0.210000 & low  & 0.0250\\\\\n",
       "\t 6569   & Autoab Pos  & GADA+        & 20.00 & 1 & Female & HIspanic/Latino & 24.10 &     & 0.236167 & low  & 0.0500\\\\\n",
       "\t HDL112 & No diabetes & Aab-         &  3.72 & 0 & Female & Caucasian       & 17.00 & 5.1 & 0.291000 & high & 0.6000\\\\\n",
       "\t 6543   & No diabetes & Aab-         &  3.50 & 0 & Male   & Caucasian       & 17.40 & 5.3 & 0.211667 & low  & 0.0275\\\\\n",
       "\t 6504   & No diabetes & Aab-         &  7.61 & 0 & Female & Hispanic/Latino & 16.60 & 5.2 & 0.318667 & high & 0.8500\\\\\n",
       "\t 6232   & No diabetes & Aab-         & 14.00 & 0 & Female & Caucasian       & 20.80 & n/a & 0.131833 & low  & 0.0001\\\\\n",
       "\t 6516   & No diabetes & Aab-         & 20.75 & 0 & Male   & Caucasian       & 28.80 & 5.5 & 0.196167 & low  & 0.0250\\\\\n",
       "\t 6509   & No diabetes & Aab-         & 18.72 & 0 & Male   & Caucasian       & 25.80 & 5.4 & 0.198667 & low  & 0.0250\\\\\n",
       "\t 6474   & No diabetes & Aab-         & 21.12 & 0 & Male   & African Am      & 29.00 & 5.3 & 0.260833 & low  & 0.2800\\\\\n",
       "\t 6548   & No diabetes & Aab-         & 20.24 & 0 & Male   & Caucasian       & 23.80 & 5.7 & 0.324500 & high & 0.9000\\\\\n",
       "\t 6539   & No diabetes & Aab-         & 24.60 & 0 & Male   & Hispanic/Latino & 19.30 & 5.7 & 0.276833 & low  & 0.4500\\\\\n",
       "\t 6455   & No diabetes & Aab-         &  6.76 & 0 & Male   & Caucasian       & 17.69 & 5.3 & 0.254833 & low  & 0.2400\\\\\n",
       "\t 6560   & No diabetes & Aab-         & 34.36 & 0 & Female & Caucasian       & 26.60 & 5.9 & 0.300000 & high & 0.7500\\\\\n",
       "\t 6392   & No diabetes & Aab-         & 14.10 & 0 & Male   & African Am      & 23.60 & 5.7 & 0.224833 & low  & 0.0450\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 34 × 12\n",
       "\n",
       "| donor_demux &lt;chr&gt; | Donor.Type &lt;chr&gt; | AAb.status &lt;chr&gt; | Age..yrs. &lt;dbl&gt; | NoAbs &lt;int&gt; | Gender &lt;chr&gt; | Race &lt;chr&gt; | BMI &lt;dbl&gt; | HbA1C &lt;chr&gt; | GRS1..European.Hispanic &lt;dbl&gt; | GRS.High.Low...0.28 &lt;chr&gt; | Bullshit.Percentiles &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 6301   | Autoab Pos  | GADA+        | 26.00 | 1 | Male   | African Am      | 32.08 | 5.5 | 0.187200 | low  | 0.0200 |\n",
       "| 6303   | Autoab Pos  | GADA+        | 22.00 | 1 | Male   | Caucasian       | 31.90 | 5.4 | 0.282700 | high | 0.6000 |\n",
       "| 6314   | Autoab Pos  | GADA+        | 21.00 | 1 | Male   | Caucasian       | 23.80 | <!----> | 0.244000 | low  | 0.1300 |\n",
       "| 6347   | Autoab Pos  | mIAA+*       |  8.50 | 1 | Male   | Caucasian       | 19.50 | <!----> | 0.160670 | low  | 0.0100 |\n",
       "| 6397   | Autoab Pos  | GADA+        | 21.16 | 1 | Female | Caucasian       | 29.60 | 6   | 0.174830 | low  | 0.0150 |\n",
       "| 6400   | Autoab Pos  | GADA+        | 25.15 | 1 | Male   | Hispanic/Latino | 22.20 | 5.5 | 0.268833 | low  | 0.3500 |\n",
       "| 6421   | Autoab Pos  | GADA+        |  6.73 | 1 | Male   | Hispanic/Latino | 17.90 | 5.6 | 0.220333 | low  | 0.0400 |\n",
       "| 6424   | Autoab Pos  | GADA+ mIAA+* | 17.65 | 2 | Male   | Caucasian       | 51.40 | 5.8 | 0.344830 | high | 0.9900 |\n",
       "| 6433   | Autoab Pos  | GADA+        | 23.96 | 1 | Male   | Hispanic/Latino | 30.80 | 5.3 | 0.268330 | low  | 0.3450 |\n",
       "| 6437   | Autoab Pos  | GADA+        | 24.73 | 1 | Male   | Caucasian       | 24.17 | 5.7 | 0.266800 | low  | 0.3200 |\n",
       "| 6483   | Autoab Pos  | GADA+        | 30.46 | 1 | Female | Caucasian       | 20.80 | 5.5 | 0.263500 | low  | 0.3000 |\n",
       "| 6494   | Autoab Pos  | mIAA+*       |  8.83 | 1 | Male   | African Am      | 15.20 | 4.9 | 0.205500 | low  | 0.0250 |\n",
       "| 6496   | Autoab Pos  | GADA+        | 26.11 | 1 | Male   | African Am      | 46.10 | 5.7 | 0.225000 | low  | 0.0450 |\n",
       "| 6517   | Autoab Pos  | GADA+        | 22.14 | 1 | Male   | Caucasian       | 26.60 | 5.6 | 0.156167 | low  | 0.0010 |\n",
       "| 6532   | Autoab Pos  | GADA+        | 20.04 | 1 | Male   | Hispanic/Latino | 23.90 | 5.9 | 0.212667 | low  | 0.0300 |\n",
       "| 6538   | Autoab Pos  | GADA+        | 19.14 | 1 | Male   | Caucasian       | 32.90 | 5.8 | 0.257500 | low  | 0.2500 |\n",
       "| 6553   | Autoab Pos  | mIAA+*       | 12.34 | 1 | Female | Hispanic/Latino | 24.80 | 8.4 | 0.333667 | high | 0.9800 |\n",
       "| 6558   | Autoab Pos  | GADA+        | 21.69 | 1 | Female | African Am      | 27.80 | 4.4 | 0.268833 | low  | 0.3500 |\n",
       "| 6562   | Autoab Pos  | GADA+        | 29.77 | 1 | Female | Caucasian       | 18.20 | 5.6 | 0.253330 | low  | 0.2400 |\n",
       "| CV02   | Autoab Pos  | GADA+ ZnT8A+ | 15.93 | 2 | Male   | Caucasian       | 22.00 | 5.1 | 0.220167 | low  | 0.0400 |\n",
       "| HDL011 | Autoab Pos  | GADA+        |  8.00 | 1 | Female | African Am      | 12.00 | n/a | 0.210000 | low  | 0.0250 |\n",
       "| 6569   | Autoab Pos  | GADA+        | 20.00 | 1 | Female | HIspanic/Latino | 24.10 | <!----> | 0.236167 | low  | 0.0500 |\n",
       "| HDL112 | No diabetes | Aab-         |  3.72 | 0 | Female | Caucasian       | 17.00 | 5.1 | 0.291000 | high | 0.6000 |\n",
       "| 6543   | No diabetes | Aab-         |  3.50 | 0 | Male   | Caucasian       | 17.40 | 5.3 | 0.211667 | low  | 0.0275 |\n",
       "| 6504   | No diabetes | Aab-         |  7.61 | 0 | Female | Hispanic/Latino | 16.60 | 5.2 | 0.318667 | high | 0.8500 |\n",
       "| 6232   | No diabetes | Aab-         | 14.00 | 0 | Female | Caucasian       | 20.80 | n/a | 0.131833 | low  | 0.0001 |\n",
       "| 6516   | No diabetes | Aab-         | 20.75 | 0 | Male   | Caucasian       | 28.80 | 5.5 | 0.196167 | low  | 0.0250 |\n",
       "| 6509   | No diabetes | Aab-         | 18.72 | 0 | Male   | Caucasian       | 25.80 | 5.4 | 0.198667 | low  | 0.0250 |\n",
       "| 6474   | No diabetes | Aab-         | 21.12 | 0 | Male   | African Am      | 29.00 | 5.3 | 0.260833 | low  | 0.2800 |\n",
       "| 6548   | No diabetes | Aab-         | 20.24 | 0 | Male   | Caucasian       | 23.80 | 5.7 | 0.324500 | high | 0.9000 |\n",
       "| 6539   | No diabetes | Aab-         | 24.60 | 0 | Male   | Hispanic/Latino | 19.30 | 5.7 | 0.276833 | low  | 0.4500 |\n",
       "| 6455   | No diabetes | Aab-         |  6.76 | 0 | Male   | Caucasian       | 17.69 | 5.3 | 0.254833 | low  | 0.2400 |\n",
       "| 6560   | No diabetes | Aab-         | 34.36 | 0 | Female | Caucasian       | 26.60 | 5.9 | 0.300000 | high | 0.7500 |\n",
       "| 6392   | No diabetes | Aab-         | 14.10 | 0 | Male   | African Am      | 23.60 | 5.7 | 0.224833 | low  | 0.0450 |\n",
       "\n"
      ],
      "text/plain": [
       "   donor_demux Donor.Type  AAb.status   Age..yrs. NoAbs Gender Race           \n",
       "1  6301        Autoab Pos  GADA+        26.00     1     Male   African Am     \n",
       "2  6303        Autoab Pos  GADA+        22.00     1     Male   Caucasian      \n",
       "3  6314        Autoab Pos  GADA+        21.00     1     Male   Caucasian      \n",
       "4  6347        Autoab Pos  mIAA+*        8.50     1     Male   Caucasian      \n",
       "5  6397        Autoab Pos  GADA+        21.16     1     Female Caucasian      \n",
       "6  6400        Autoab Pos  GADA+        25.15     1     Male   Hispanic/Latino\n",
       "7  6421        Autoab Pos  GADA+         6.73     1     Male   Hispanic/Latino\n",
       "8  6424        Autoab Pos  GADA+ mIAA+* 17.65     2     Male   Caucasian      \n",
       "9  6433        Autoab Pos  GADA+        23.96     1     Male   Hispanic/Latino\n",
       "10 6437        Autoab Pos  GADA+        24.73     1     Male   Caucasian      \n",
       "11 6483        Autoab Pos  GADA+        30.46     1     Female Caucasian      \n",
       "12 6494        Autoab Pos  mIAA+*        8.83     1     Male   African Am     \n",
       "13 6496        Autoab Pos  GADA+        26.11     1     Male   African Am     \n",
       "14 6517        Autoab Pos  GADA+        22.14     1     Male   Caucasian      \n",
       "15 6532        Autoab Pos  GADA+        20.04     1     Male   Hispanic/Latino\n",
       "16 6538        Autoab Pos  GADA+        19.14     1     Male   Caucasian      \n",
       "17 6553        Autoab Pos  mIAA+*       12.34     1     Female Hispanic/Latino\n",
       "18 6558        Autoab Pos  GADA+        21.69     1     Female African Am     \n",
       "19 6562        Autoab Pos  GADA+        29.77     1     Female Caucasian      \n",
       "20 CV02        Autoab Pos  GADA+ ZnT8A+ 15.93     2     Male   Caucasian      \n",
       "21 HDL011      Autoab Pos  GADA+         8.00     1     Female African Am     \n",
       "22 6569        Autoab Pos  GADA+        20.00     1     Female HIspanic/Latino\n",
       "23 HDL112      No diabetes Aab-          3.72     0     Female Caucasian      \n",
       "24 6543        No diabetes Aab-          3.50     0     Male   Caucasian      \n",
       "25 6504        No diabetes Aab-          7.61     0     Female Hispanic/Latino\n",
       "26 6232        No diabetes Aab-         14.00     0     Female Caucasian      \n",
       "27 6516        No diabetes Aab-         20.75     0     Male   Caucasian      \n",
       "28 6509        No diabetes Aab-         18.72     0     Male   Caucasian      \n",
       "29 6474        No diabetes Aab-         21.12     0     Male   African Am     \n",
       "30 6548        No diabetes Aab-         20.24     0     Male   Caucasian      \n",
       "31 6539        No diabetes Aab-         24.60     0     Male   Hispanic/Latino\n",
       "32 6455        No diabetes Aab-          6.76     0     Male   Caucasian      \n",
       "33 6560        No diabetes Aab-         34.36     0     Female Caucasian      \n",
       "34 6392        No diabetes Aab-         14.10     0     Male   African Am     \n",
       "   BMI   HbA1C GRS1..European.Hispanic GRS.High.Low...0.28 Bullshit.Percentiles\n",
       "1  32.08 5.5   0.187200                low                 0.0200              \n",
       "2  31.90 5.4   0.282700                high                0.6000              \n",
       "3  23.80       0.244000                low                 0.1300              \n",
       "4  19.50       0.160670                low                 0.0100              \n",
       "5  29.60 6     0.174830                low                 0.0150              \n",
       "6  22.20 5.5   0.268833                low                 0.3500              \n",
       "7  17.90 5.6   0.220333                low                 0.0400              \n",
       "8  51.40 5.8   0.344830                high                0.9900              \n",
       "9  30.80 5.3   0.268330                low                 0.3450              \n",
       "10 24.17 5.7   0.266800                low                 0.3200              \n",
       "11 20.80 5.5   0.263500                low                 0.3000              \n",
       "12 15.20 4.9   0.205500                low                 0.0250              \n",
       "13 46.10 5.7   0.225000                low                 0.0450              \n",
       "14 26.60 5.6   0.156167                low                 0.0010              \n",
       "15 23.90 5.9   0.212667                low                 0.0300              \n",
       "16 32.90 5.8   0.257500                low                 0.2500              \n",
       "17 24.80 8.4   0.333667                high                0.9800              \n",
       "18 27.80 4.4   0.268833                low                 0.3500              \n",
       "19 18.20 5.6   0.253330                low                 0.2400              \n",
       "20 22.00 5.1   0.220167                low                 0.0400              \n",
       "21 12.00 n/a   0.210000                low                 0.0250              \n",
       "22 24.10       0.236167                low                 0.0500              \n",
       "23 17.00 5.1   0.291000                high                0.6000              \n",
       "24 17.40 5.3   0.211667                low                 0.0275              \n",
       "25 16.60 5.2   0.318667                high                0.8500              \n",
       "26 20.80 n/a   0.131833                low                 0.0001              \n",
       "27 28.80 5.5   0.196167                low                 0.0250              \n",
       "28 25.80 5.4   0.198667                low                 0.0250              \n",
       "29 29.00 5.3   0.260833                low                 0.2800              \n",
       "30 23.80 5.7   0.324500                high                0.9000              \n",
       "31 19.30 5.7   0.276833                low                 0.4500              \n",
       "32 17.69 5.3   0.254833                low                 0.2400              \n",
       "33 26.60 5.9   0.300000                high                0.7500              \n",
       "34 23.60 5.7   0.224833                low                 0.0450              "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Load sample metadata to attach\n",
    "sample_info <- read.table('/nfs/lab/projects/npod_aab/data/012824_metadata.tsv', \n",
    "           sep='\\t', header=T)\n",
    "\n",
    "dim(sample_info)\n",
    "\n",
    "#sample_info <- read.table('/nfs/lab/projects/npod_aab/data/012824_metadata.tsv', \n",
    "#           sep='\\t', header=T)\n",
    "\n",
    "sample_info\n",
    "#dataset_sample_info <- filter(sample_info, Dataset==dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create Filtered RDS Object "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Loading h5\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Genome matrix has multiple modalities, returning a list of matrices for this genome\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Creating RNA Object\"\n",
      "[1] \"Creating ATAC Object\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Computing hash\n",
      "\n"
     ]
    }
   ],
   "source": [
    "outdir <- paste0(\"/nfs/lab/projects/mega_pancreas/data/cellranger/\",dataset,\"/\", modality, \"/\", \"intermediates\", \"/\")\n",
    "\n",
    "#a filtered RDS object - and some tables prepareped to feed into AMULET\n",
    "\n",
    "# Load in 10x and calculate QC metrics\n",
    "for (samp in samples) {\n",
    "    sample_info_sub <- filter(sample_info, donor_demux==samp)\n",
    "    if (nrow(sample_info_sub) > 1) {\n",
    "        print(\"WARNING: Sample meta matched multiple rows.\")\n",
    "    print(samp)\n",
    "    }\n",
    "    \n",
    "    \n",
    "    ### Load in 10x data\n",
    "    print(\"Loading h5\")\n",
    "    wd <- paste0(samp_dir, samp, \"/outs/\")\n",
    "    inputdata.10x <- Read10X_h5(file.path(wd, 'raw_feature_bc_matrix.h5'))\n",
    "    \n",
    "    adata <- load_multiome_object(inputdata.10x, wd)\n",
    "    for (fact in colnames(sample_info_sub)) {\n",
    "        adata[[fact]] <- sample_info_sub[[fact]]\n",
    "    }\n",
    "    \n",
    "    amulet_barcodes <- data.frame(barcode=colnames(adata))\n",
    "\n",
    "    \n",
    "    ### Print some details and add sample metadata\n",
    "    print(paste0(samp, \": \", length(colnames(adata[[\"RNA\"]])),\" total BCs\"))\n",
    "    \n",
    "    print(paste0(samp, \" metadata dimensions: \",  stringr::str_flatten(dim(adata@meta.data), collapse='x')))\n",
    "    print(paste0(samp, \" Seurat summary:\"))\n",
    "    print(adata)\n",
    "    gc()\n",
    "    \n",
    "    \n",
    "    ### Subset\n",
    "    adata <- subset(\n",
    "      x = adata,\n",
    "      subset = nFeature_RNA >= 500\n",
    "    )\n",
    "    print(paste(length(colnames(adata[[\"RNA\"]])),\"total BCs after RNA UMI threshold\"))\n",
    "    \n",
    "    adata <- subset(\n",
    "      x = adata,\n",
    "      subset = percent.mt <= 10\n",
    "    )\n",
    "    print(paste(length(colnames(adata[[\"RNA\"]])),\"total BCs after RNA MT threshold\"))\n",
    "    \n",
    "    adata <- subset(\n",
    "      x = adata,\n",
    "      subset = atac_fragments >= 1000\n",
    "    )\n",
    "    print(paste(length(colnames(adata[[\"RNA\"]])),\"total BCs after ATAC fragment threshold\"))\n",
    "    \n",
    "    adata <- subset(\n",
    "      x = adata,\n",
    "      subset = TSS.enrichment >= 3\n",
    "    )\n",
    "    print(paste(length(colnames(adata[[\"RNA\"]])),\"total BCs after ATAC TSS enrichment threshold\"))\n",
    "    \n",
    "    #adata <- subset(\n",
    "    #  x = adata,\n",
    "    #  subset = pct_reads_in_peaks >= 15\n",
    "    #)\n",
    "    #print(paste(length(colnames(adata[[\"RNA\"]])),\"total BCs after ATAC FRiP threshold\"))\n",
    "    \n",
    "    print(paste(\"RNA median genes per cell:\",median(adata[[]][,'nFeature_RNA'])))\n",
    "    print(paste(\"RNA median counts per cell:\",median(adata[[]][,'nCount_RNA'])))\n",
    "    print(paste(\"ATAC median peaks per cell:\",median(adata[[]][,'nFeature_ATAC'])))\n",
    "    print(paste(\"ATAC median fragments per cell:\",median(adata[[]][,'atac_fragments'])))\n",
    "    \n",
    "    adata <- multiome_cluster_single_samp(adata)\n",
    "    \n",
    "    saveRDS(adata, paste0(outdir, samp,\"_\",modality,\".RDS\"))\n",
    "    \n",
    "    amulet_table <- dplyr::left_join(amulet_barcodes, data.frame(barcode=colnames(adata), \n",
    "                                             cell_id=colnames(adata), \n",
    "                                             is__cell_barcode=TRUE))\n",
    "    amulet_table$is__cell_barcode[is.na(amulet_table$is__cell_barcode)] <- FALSE\n",
    "    amulet_table$is__cell_barcode <- as.numeric(amulet_table$is__cell_barcode)\n",
    "    amulet_table$cell_id[is.na(amulet_table$cell_id)] <- 'None'\n",
    "    \n",
    "    write.table(amulet_table, paste0(outdir, samp,  \"_AMULET_barcode_table.csv\"), \n",
    "                sep=',', col.names=T, row.names=T, quote=F)\n",
    "    \n",
    "    gc()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write AMULET Script"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Multiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'MM_1349_1_2_MM_1345_1_2'</li><li>'MM_1350_1_2_MM_1346_1_2'</li><li>'MM_1351_1_2_MM_1347_1_2'</li><li>'MM_1352_1_2_MM_1348_1_2'</li><li>'QY_2312_1_2_QY_2311_1_2'</li><li>'QY_2314_1_2_QY_2313_1_2'</li><li>'QY_2316_1_2_QY_2315_1_2'</li><li>'QY_2318_1_2_QY_2317_1_2'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'MM\\_1349\\_1\\_2\\_MM\\_1345\\_1\\_2'\n",
       "\\item 'MM\\_1350\\_1\\_2\\_MM\\_1346\\_1\\_2'\n",
       "\\item 'MM\\_1351\\_1\\_2\\_MM\\_1347\\_1\\_2'\n",
       "\\item 'MM\\_1352\\_1\\_2\\_MM\\_1348\\_1\\_2'\n",
       "\\item 'QY\\_2312\\_1\\_2\\_QY\\_2311\\_1\\_2'\n",
       "\\item 'QY\\_2314\\_1\\_2\\_QY\\_2313\\_1\\_2'\n",
       "\\item 'QY\\_2316\\_1\\_2\\_QY\\_2315\\_1\\_2'\n",
       "\\item 'QY\\_2318\\_1\\_2\\_QY\\_2317\\_1\\_2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'MM_1349_1_2_MM_1345_1_2'\n",
       "2. 'MM_1350_1_2_MM_1346_1_2'\n",
       "3. 'MM_1351_1_2_MM_1347_1_2'\n",
       "4. 'MM_1352_1_2_MM_1348_1_2'\n",
       "5. 'QY_2312_1_2_QY_2311_1_2'\n",
       "6. 'QY_2314_1_2_QY_2313_1_2'\n",
       "7. 'QY_2316_1_2_QY_2315_1_2'\n",
       "8. 'QY_2318_1_2_QY_2317_1_2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"MM_1349_1_2_MM_1345_1_2\" \"MM_1350_1_2_MM_1346_1_2\"\n",
       "[3] \"MM_1351_1_2_MM_1347_1_2\" \"MM_1352_1_2_MM_1348_1_2\"\n",
       "[5] \"QY_2312_1_2_QY_2311_1_2\" \"QY_2314_1_2_QY_2313_1_2\"\n",
       "[7] \"QY_2316_1_2_QY_2315_1_2\" \"QY_2318_1_2_QY_2317_1_2\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Set some values and load sample list\n",
    "dataset <- 'nPOD'\n",
    "modality <- 'Pooled'\n",
    "\n",
    "samp_dir <- paste0(\"/nfs/lab/projects/mega_pancreas/data/cellranger/\",dataset,\"/\", modality, \"/\")\n",
    "\n",
    "samples <- list.dirs(samp_dir, full.names = FALSE, recursive = FALSE)\n",
    "samples = samples[samples != \"intermediates\"]\n",
    "\n",
    "samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li><span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1349_1_2_MM_1345_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/MM_1349_1_2_MM_1345_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM_1349_1_2_MM_1345_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1349_1_2_MM_1345_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span></li><li><span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1350_1_2_MM_1346_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/MM_1350_1_2_MM_1346_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM_1350_1_2_MM_1346_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1350_1_2_MM_1346_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span></li><li><span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1351_1_2_MM_1347_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/MM_1351_1_2_MM_1347_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM_1351_1_2_MM_1347_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1351_1_2_MM_1347_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span></li><li><span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1352_1_2_MM_1348_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/MM_1352_1_2_MM_1348_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM_1352_1_2_MM_1348_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1352_1_2_MM_1348_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span></li><li><span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2312_1_2_QY_2311_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/QY_2312_1_2_QY_2311_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY_2312_1_2_QY_2311_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2312_1_2_QY_2311_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span></li><li><span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2314_1_2_QY_2313_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/QY_2314_1_2_QY_2313_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY_2314_1_2_QY_2313_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2314_1_2_QY_2313_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span></li><li><span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2316_1_2_QY_2315_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/QY_2316_1_2_QY_2315_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY_2316_1_2_QY_2315_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2316_1_2_QY_2315_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span></li><li><span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2318_1_2_QY_2317_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/QY_2318_1_2_QY_2317_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY_2318_1_2_QY_2317_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2318_1_2_QY_2317_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span></li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'mkdir /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/MM\\_1349\\_1\\_2\\_MM\\_1345\\_1\\_2 -p \\&\\&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/MM\\_1349\\_1\\_2\\_MM\\_1345\\_1\\_2/outs/atac\\_fragments.tsv.gz  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM\\_1349\\_1\\_2\\_MM\\_1345\\_1\\_2\\_AMULET\\_barcode\\_table.csv  /nfs/lab/welison/packages/AMULET/human\\_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist\\_repeats\\_segdups\\_rmsk\\_hg38.bed  /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/MM\\_1349\\_1\\_2\\_MM\\_1345\\_1\\_2/ /nfs/lab/welison/packages/AMULET/ \\&'\n",
       "\\item 'mkdir /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/MM\\_1350\\_1\\_2\\_MM\\_1346\\_1\\_2 -p \\&\\&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/MM\\_1350\\_1\\_2\\_MM\\_1346\\_1\\_2/outs/atac\\_fragments.tsv.gz  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM\\_1350\\_1\\_2\\_MM\\_1346\\_1\\_2\\_AMULET\\_barcode\\_table.csv  /nfs/lab/welison/packages/AMULET/human\\_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist\\_repeats\\_segdups\\_rmsk\\_hg38.bed  /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/MM\\_1350\\_1\\_2\\_MM\\_1346\\_1\\_2/ /nfs/lab/welison/packages/AMULET/ \\&'\n",
       "\\item 'mkdir /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/MM\\_1351\\_1\\_2\\_MM\\_1347\\_1\\_2 -p \\&\\&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/MM\\_1351\\_1\\_2\\_MM\\_1347\\_1\\_2/outs/atac\\_fragments.tsv.gz  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM\\_1351\\_1\\_2\\_MM\\_1347\\_1\\_2\\_AMULET\\_barcode\\_table.csv  /nfs/lab/welison/packages/AMULET/human\\_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist\\_repeats\\_segdups\\_rmsk\\_hg38.bed  /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/MM\\_1351\\_1\\_2\\_MM\\_1347\\_1\\_2/ /nfs/lab/welison/packages/AMULET/ \\&'\n",
       "\\item 'mkdir /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/MM\\_1352\\_1\\_2\\_MM\\_1348\\_1\\_2 -p \\&\\&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/MM\\_1352\\_1\\_2\\_MM\\_1348\\_1\\_2/outs/atac\\_fragments.tsv.gz  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM\\_1352\\_1\\_2\\_MM\\_1348\\_1\\_2\\_AMULET\\_barcode\\_table.csv  /nfs/lab/welison/packages/AMULET/human\\_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist\\_repeats\\_segdups\\_rmsk\\_hg38.bed  /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/MM\\_1352\\_1\\_2\\_MM\\_1348\\_1\\_2/ /nfs/lab/welison/packages/AMULET/ \\&'\n",
       "\\item 'mkdir /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/QY\\_2312\\_1\\_2\\_QY\\_2311\\_1\\_2 -p \\&\\&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/QY\\_2312\\_1\\_2\\_QY\\_2311\\_1\\_2/outs/atac\\_fragments.tsv.gz  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY\\_2312\\_1\\_2\\_QY\\_2311\\_1\\_2\\_AMULET\\_barcode\\_table.csv  /nfs/lab/welison/packages/AMULET/human\\_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist\\_repeats\\_segdups\\_rmsk\\_hg38.bed  /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/QY\\_2312\\_1\\_2\\_QY\\_2311\\_1\\_2/ /nfs/lab/welison/packages/AMULET/ \\&'\n",
       "\\item 'mkdir /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/QY\\_2314\\_1\\_2\\_QY\\_2313\\_1\\_2 -p \\&\\&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/QY\\_2314\\_1\\_2\\_QY\\_2313\\_1\\_2/outs/atac\\_fragments.tsv.gz  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY\\_2314\\_1\\_2\\_QY\\_2313\\_1\\_2\\_AMULET\\_barcode\\_table.csv  /nfs/lab/welison/packages/AMULET/human\\_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist\\_repeats\\_segdups\\_rmsk\\_hg38.bed  /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/QY\\_2314\\_1\\_2\\_QY\\_2313\\_1\\_2/ /nfs/lab/welison/packages/AMULET/ \\&'\n",
       "\\item 'mkdir /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/QY\\_2316\\_1\\_2\\_QY\\_2315\\_1\\_2 -p \\&\\&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/QY\\_2316\\_1\\_2\\_QY\\_2315\\_1\\_2/outs/atac\\_fragments.tsv.gz  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY\\_2316\\_1\\_2\\_QY\\_2315\\_1\\_2\\_AMULET\\_barcode\\_table.csv  /nfs/lab/welison/packages/AMULET/human\\_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist\\_repeats\\_segdups\\_rmsk\\_hg38.bed  /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/QY\\_2316\\_1\\_2\\_QY\\_2315\\_1\\_2/ /nfs/lab/welison/packages/AMULET/ \\&'\n",
       "\\item 'mkdir /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/QY\\_2318\\_1\\_2\\_QY\\_2317\\_1\\_2 -p \\&\\&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/QY\\_2318\\_1\\_2\\_QY\\_2317\\_1\\_2/outs/atac\\_fragments.tsv.gz  /nfs/lab/projects/mega\\_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY\\_2318\\_1\\_2\\_QY\\_2317\\_1\\_2\\_AMULET\\_barcode\\_table.csv  /nfs/lab/welison/packages/AMULET/human\\_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist\\_repeats\\_segdups\\_rmsk\\_hg38.bed  /nfs/lab/welison/mega\\_pancreas/intermediates/Single\\_Sample\\_Processing\\_Intermediates/nPOD/231201\\_AMULET/QY\\_2318\\_1\\_2\\_QY\\_2317\\_1\\_2/ /nfs/lab/welison/packages/AMULET/ \\&'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. <span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1349_1_2_MM_1345_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/MM_1349_1_2_MM_1345_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM_1349_1_2_MM_1345_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1349_1_2_MM_1345_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span>\n",
       "2. <span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1350_1_2_MM_1346_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/MM_1350_1_2_MM_1346_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM_1350_1_2_MM_1346_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1350_1_2_MM_1346_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span>\n",
       "3. <span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1351_1_2_MM_1347_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/MM_1351_1_2_MM_1347_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM_1351_1_2_MM_1347_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1351_1_2_MM_1347_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span>\n",
       "4. <span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1352_1_2_MM_1348_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/MM_1352_1_2_MM_1348_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM_1352_1_2_MM_1348_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1352_1_2_MM_1348_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span>\n",
       "5. <span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2312_1_2_QY_2311_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/QY_2312_1_2_QY_2311_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY_2312_1_2_QY_2311_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2312_1_2_QY_2311_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span>\n",
       "6. <span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2314_1_2_QY_2313_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/QY_2314_1_2_QY_2313_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY_2314_1_2_QY_2313_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2314_1_2_QY_2313_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span>\n",
       "7. <span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2316_1_2_QY_2315_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/QY_2316_1_2_QY_2315_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY_2316_1_2_QY_2315_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2316_1_2_QY_2315_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span>\n",
       "8. <span style=white-space:pre-wrap>'mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2318_1_2_QY_2317_1_2 -p &amp;&amp;  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/QY_2318_1_2_QY_2317_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY_2318_1_2_QY_2317_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2318_1_2_QY_2317_1_2/ /nfs/lab/welison/packages/AMULET/ &amp;'</span>\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1349_1_2_MM_1345_1_2 -p &&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/MM_1349_1_2_MM_1345_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM_1349_1_2_MM_1345_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1349_1_2_MM_1345_1_2/ /nfs/lab/welison/packages/AMULET/ &\"\n",
       "[2] \"mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1350_1_2_MM_1346_1_2 -p &&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/MM_1350_1_2_MM_1346_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM_1350_1_2_MM_1346_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1350_1_2_MM_1346_1_2/ /nfs/lab/welison/packages/AMULET/ &\"\n",
       "[3] \"mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1351_1_2_MM_1347_1_2 -p &&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/MM_1351_1_2_MM_1347_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM_1351_1_2_MM_1347_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1351_1_2_MM_1347_1_2/ /nfs/lab/welison/packages/AMULET/ &\"\n",
       "[4] \"mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1352_1_2_MM_1348_1_2 -p &&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/MM_1352_1_2_MM_1348_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/MM_1352_1_2_MM_1348_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/MM_1352_1_2_MM_1348_1_2/ /nfs/lab/welison/packages/AMULET/ &\"\n",
       "[5] \"mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2312_1_2_QY_2311_1_2 -p &&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/QY_2312_1_2_QY_2311_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY_2312_1_2_QY_2311_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2312_1_2_QY_2311_1_2/ /nfs/lab/welison/packages/AMULET/ &\"\n",
       "[6] \"mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2314_1_2_QY_2313_1_2 -p &&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/QY_2314_1_2_QY_2313_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY_2314_1_2_QY_2313_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2314_1_2_QY_2313_1_2/ /nfs/lab/welison/packages/AMULET/ &\"\n",
       "[7] \"mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2316_1_2_QY_2315_1_2 -p &&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/QY_2316_1_2_QY_2315_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY_2316_1_2_QY_2315_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2316_1_2_QY_2315_1_2/ /nfs/lab/welison/packages/AMULET/ &\"\n",
       "[8] \"mkdir /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2318_1_2_QY_2317_1_2 -p &&  /nfs/lab/welison/packages/AMULET/AMULET.sh --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/QY_2318_1_2_QY_2317_1_2/outs/atac_fragments.tsv.gz  /nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/QY_2318_1_2_QY_2317_1_2_AMULET_barcode_table.csv  /nfs/lab/welison/packages/AMULET/human_autosomes.txt  /nfs/lab/welison/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed  /nfs/lab/welison/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/QY_2318_1_2_QY_2317_1_2/ /nfs/lab/welison/packages/AMULET/ &\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Amulet_script_dir <- '/nfs/lab/packages/AMULET/AMULET.sh'\n",
    "Amulet_out_dir_base <- paste0('/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/', dataset, '/231201_AMULET/')\n",
    "fragment_base_path <- paste0('/nfs/lab/projects/mega_pancreas/data/cellranger/', dataset, '/', modality, '/')\n",
    "Amulet_table_path <- paste0('/nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/')\n",
    "autosome_file <- '/nfs/lab/packages/AMULET/human_autosomes.txt'\n",
    "restriction_file <- '/nfs/lab/packages/AMULET/RestrictionRepeatLists/restrictionlist_repeats_segdups_rmsk_hg38.bed'\n",
    "Amulet_package_path <- '/nfs/lab/packages/AMULET/'\n",
    "\n",
    "amulet_commands <- c(paste0('mkdir ', Amulet_out_dir_base, samples, ' -p && \\ ', Amulet_script_dir,' --forcesorted --bambc CB --bcidx 1 --cellidx 2 --iscellidx 3 \\ ',fragment_base_path, samples, '/outs/atac_fragments.tsv.gz \\ ',Amulet_table_path, samples, '_AMULET_barcode_table.csv \\ ',autosome_file,' \\ ',restriction_file,' \\ ',Amulet_out_dir_base, samples, '/ ', Amulet_package_path, ' &'))\n",
    "\n",
    "amulet_commands"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "system(paste0('mkdir /nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/', dataset, \n",
    "           '/231201_AMULET/ -p'))\n",
    "\n",
    "write.table(data.frame(amulet_commands), \n",
    "    paste0('/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/', dataset, \n",
    "           '/231201_AMULET/231207_AH_',dataset,'_',modality,'_AMULET_script.sh'),\n",
    "           quote=F, row.names=F, col.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "#conda env you need to load to then run these commands in a tmux window \n",
    "#/home/.conda/envs/mamba/envs/amulet_2\n",
    "#in tmux attach -t amulet\n",
    "#in (base) ahowell@ophelia:/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET$\n",
    "#activate conda env\n",
    "#conda activate /home/.conda/envs/mamba/envs/amulet_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make Initial Clustering Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker.genes <- c('INS','GCG','MKI67','SST','PPY','GHRL', #Beta, alpha, delta, gamma, epsilon, cycling\n",
    "                  'CTRB2','PRSS2','CPA1','REG1A', #Acinar\n",
    "                  'CFTR','MUC5B', #Ductal + muc5b+\n",
    "                  'COL6A3','PDGFRB','IGA1','SPARCL1','SPARC','COL1A1', #2x general stellate, 2x quiescent, 2x activated\n",
    "                  'PLVAP','VWF','LYVE1','PROX1','PMP22', #Endothelial x2, lymph endo 2x, schwann\n",
    "                  'C1QA','C1QB','C1QC', # Macrophage\n",
    "                  'CD3D','CD4','CD8B','CXCR4') # T cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Multiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'MM_1349_1_2_MM_1345_1_2'</li><li>'MM_1350_1_2_MM_1346_1_2'</li><li>'MM_1351_1_2_MM_1347_1_2'</li><li>'MM_1352_1_2_MM_1348_1_2'</li><li>'QY_2312_1_2_QY_2311_1_2'</li><li>'QY_2314_1_2_QY_2313_1_2'</li><li>'QY_2316_1_2_QY_2315_1_2'</li><li>'QY_2318_1_2_QY_2317_1_2'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'MM\\_1349\\_1\\_2\\_MM\\_1345\\_1\\_2'\n",
       "\\item 'MM\\_1350\\_1\\_2\\_MM\\_1346\\_1\\_2'\n",
       "\\item 'MM\\_1351\\_1\\_2\\_MM\\_1347\\_1\\_2'\n",
       "\\item 'MM\\_1352\\_1\\_2\\_MM\\_1348\\_1\\_2'\n",
       "\\item 'QY\\_2312\\_1\\_2\\_QY\\_2311\\_1\\_2'\n",
       "\\item 'QY\\_2314\\_1\\_2\\_QY\\_2313\\_1\\_2'\n",
       "\\item 'QY\\_2316\\_1\\_2\\_QY\\_2315\\_1\\_2'\n",
       "\\item 'QY\\_2318\\_1\\_2\\_QY\\_2317\\_1\\_2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'MM_1349_1_2_MM_1345_1_2'\n",
       "2. 'MM_1350_1_2_MM_1346_1_2'\n",
       "3. 'MM_1351_1_2_MM_1347_1_2'\n",
       "4. 'MM_1352_1_2_MM_1348_1_2'\n",
       "5. 'QY_2312_1_2_QY_2311_1_2'\n",
       "6. 'QY_2314_1_2_QY_2313_1_2'\n",
       "7. 'QY_2316_1_2_QY_2315_1_2'\n",
       "8. 'QY_2318_1_2_QY_2317_1_2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"MM_1349_1_2_MM_1345_1_2\" \"MM_1350_1_2_MM_1346_1_2\"\n",
       "[3] \"MM_1351_1_2_MM_1347_1_2\" \"MM_1352_1_2_MM_1348_1_2\"\n",
       "[5] \"QY_2312_1_2_QY_2311_1_2\" \"QY_2314_1_2_QY_2313_1_2\"\n",
       "[7] \"QY_2316_1_2_QY_2315_1_2\" \"QY_2318_1_2_QY_2317_1_2\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Generate Plots and save in a nested set of lists. \n",
    "#Do this for each sample and cleanup each time to reduce memory.\n",
    "\n",
    "dataset <- 'nPOD'\n",
    "modality <- 'Pooled'\n",
    "\n",
    "samp_dir <- paste0(\"/nfs/lab/projects/mega_pancreas/data/cellranger/\",dataset,\"/\", modality, \"/\")\n",
    "samples <- list.dirs(samp_dir, full.names = FALSE, recursive = FALSE)\n",
    "samples = samples[samples != \"intermediates\"]\n",
    "\n",
    "#says outdir its not misleading\n",
    "outdir <- paste0(\"/nfs/lab/projects/mega_pancreas/data/cellranger/\",dataset,\"/\", modality, \"/\", \"intermediates\", \"/\")\n",
    "pdf_dir <- (paste0('/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/', dataset,'/231121_WE_SSP_Intermediates/Raw_Plots/'))\n",
    "\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 6 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>used</th><th scope=col>(Mb)</th><th scope=col>gc trigger</th><th scope=col>(Mb)</th><th scope=col>max used</th><th scope=col>(Mb)</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>Ncells</th><td>  15421090</td><td> 823.6</td><td>  43642744</td><td> 2330.8</td><td>  59109191</td><td> 3156.8</td></tr>\n",
       "\t<tr><th scope=row>Vcells</th><td>1163911862</td><td>8880.0</td><td>4823368108</td><td>36799.4</td><td>4345218349</td><td>33151.4</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 6 of type dbl\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & used & (Mb) & gc trigger & (Mb) & max used & (Mb)\\\\\n",
       "\\hline\n",
       "\tNcells &   15421090 &  823.6 &   43642744 &  2330.8 &   59109191 &  3156.8\\\\\n",
       "\tVcells & 1163911862 & 8880.0 & 4823368108 & 36799.4 & 4345218349 & 33151.4\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 6 of type dbl\n",
       "\n",
       "| <!--/--> | used | (Mb) | gc trigger | (Mb) | max used | (Mb) |\n",
       "|---|---|---|---|---|---|---|\n",
       "| Ncells |   15421090 |  823.6 |   43642744 |  2330.8 |   59109191 |  3156.8 |\n",
       "| Vcells | 1163911862 | 8880.0 | 4823368108 | 36799.4 | 4345218349 | 33151.4 |\n",
       "\n"
      ],
      "text/plain": [
       "       used       (Mb)   gc trigger (Mb)    max used   (Mb)   \n",
       "Ncells   15421090  823.6   43642744  2330.8   59109191  3156.8\n",
       "Vcells 1163911862 8880.0 4823368108 36799.4 4345218349 33151.4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Generate Plots and save in a nested set of lists. \n",
    "#Do this for each sample and cleanup each time to reduce memory.\n",
    "plots <- list()\n",
    "\n",
    "#for (samp in samples[1]) {\n",
    "for (samp in samples) {\n",
    "    #Load in adata and cleanup\n",
    "    adata <- readRDS(paste0(outdir, samp,\"_\",modality,\".RDS\"))\n",
    "    gc()\n",
    "    \n",
    "    plots[[samp]] <- multiome_single_samp_plots(adata, samp, marker.genes)\n",
    "    \n",
    "    print_single_sample_pdf_multiome(samp, pdf_dir, plots, '_Initial_QC_Plots.pdf')\n",
    "}\n",
    " \n",
    "print_pdfs_for_each_graph_type_multiome(plots, samples, pdf_dir, modality)\n",
    "\n",
    "plots <- NA\n",
    "gc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create AMULET filtered RDS Object "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'MM_1349_1_2_MM_1345_1_2'</li><li>'MM_1350_1_2_MM_1346_1_2'</li><li>'MM_1351_1_2_MM_1347_1_2'</li><li>'MM_1352_1_2_MM_1348_1_2'</li><li>'QY_2312_1_2_QY_2311_1_2'</li><li>'QY_2314_1_2_QY_2313_1_2'</li><li>'QY_2316_1_2_QY_2315_1_2'</li><li>'QY_2318_1_2_QY_2317_1_2'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'MM\\_1349\\_1\\_2\\_MM\\_1345\\_1\\_2'\n",
       "\\item 'MM\\_1350\\_1\\_2\\_MM\\_1346\\_1\\_2'\n",
       "\\item 'MM\\_1351\\_1\\_2\\_MM\\_1347\\_1\\_2'\n",
       "\\item 'MM\\_1352\\_1\\_2\\_MM\\_1348\\_1\\_2'\n",
       "\\item 'QY\\_2312\\_1\\_2\\_QY\\_2311\\_1\\_2'\n",
       "\\item 'QY\\_2314\\_1\\_2\\_QY\\_2313\\_1\\_2'\n",
       "\\item 'QY\\_2316\\_1\\_2\\_QY\\_2315\\_1\\_2'\n",
       "\\item 'QY\\_2318\\_1\\_2\\_QY\\_2317\\_1\\_2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'MM_1349_1_2_MM_1345_1_2'\n",
       "2. 'MM_1350_1_2_MM_1346_1_2'\n",
       "3. 'MM_1351_1_2_MM_1347_1_2'\n",
       "4. 'MM_1352_1_2_MM_1348_1_2'\n",
       "5. 'QY_2312_1_2_QY_2311_1_2'\n",
       "6. 'QY_2314_1_2_QY_2313_1_2'\n",
       "7. 'QY_2316_1_2_QY_2315_1_2'\n",
       "8. 'QY_2318_1_2_QY_2317_1_2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"MM_1349_1_2_MM_1345_1_2\" \"MM_1350_1_2_MM_1346_1_2\"\n",
       "[3] \"MM_1351_1_2_MM_1347_1_2\" \"MM_1352_1_2_MM_1348_1_2\"\n",
       "[5] \"QY_2312_1_2_QY_2311_1_2\" \"QY_2314_1_2_QY_2313_1_2\"\n",
       "[7] \"QY_2316_1_2_QY_2315_1_2\" \"QY_2318_1_2_QY_2317_1_2\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#read in initial RDS, filter using the AMULET identified MultipletBarcodes.txt file\n",
    "#get out AMULET plots and AMULET filtered RDS\n",
    "dataset <- 'nPOD'\n",
    "modality <- 'Pooled'\n",
    "\n",
    "#this directory is just for getting a list of sample names\n",
    "samp_dir <- paste0(\"/nfs/lab/projects/mega_pancreas/data/cellranger/\",dataset,\"/\", modality, \"/\")\n",
    "samples <- list.dirs(samp_dir, full.names = FALSE, recursive = FALSE)\n",
    "samples = samples[samples != \"intermediates\"]\n",
    "samples\n",
    "\n",
    "indir <- paste0(\"/nfs/lab/projects/mega_pancreas/data/cellranger/nPOD/Pooled/intermediates/\")\n",
    "outdir <- paste0(\"/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/\",dataset,'/231121_WE_SSP_Intermediates/Amulet_adatas/')\n",
    "amulet_dir <- paste0(\"/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231201_AMULET/\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Normalizing layer: counts\n",
      "\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message:\n",
      "\"Different features in new layer data than already exists for scale.data\"\n",
      "PC_ 1 \n",
      "Positive:  RIMS2, KCNMB2, KCNB2, NRXN1, SLC30A8, NOL4, CACNA1A, CACNA1D, CACNA1C, FMN2 \n",
      "\t   ABCC8, ST18, LRRTM4, NLGN1, RGS7, UNC80, RIMBP2, SOX5, CACNA2D1, DACH1 \n",
      "\t   HMGCLL1, CADPS, CNTN1, STXBP5L, SNAP25, RFX6, TRPM3, PCSK2, ROBO2, DPP6 \n",
      "Negative:  TRHDE, CEL, CUZD1, SEL1L, CATSPERB, CPA1, PNLIP, PRSS2, ZNF804B, CTRC \n",
      "\t   PEX5L, AC010967.1, REG1B, NTN4, GPT2, MIR217HG, AASS, PLA2G1B, BTG2, PWRN1 \n",
      "\t   BNIP3, ANPEP, AL159156.1, ERRFI1, AMY2B, AMY2A, AC013652.1, SYBU, KANK4, GAREM1 \n",
      "PC_ 2 \n",
      "Positive:  RIMS2, PDE3B, KCNMB2, SLC30A8, KCNB2, NRXN1, CACNA1A, ABCC8, NOL4, DPP6 \n",
      "\t   RGS7, ST18, FMN2, MTUS2, LRRTM4, HS6ST3, KCNMA1, PWRN1, MYT1L, RIMBP2 \n",
      "\t   UNC80, RFX6, CEL, TRHDE, STXBP5L, PAM, TMEM178B, ELAVL4, PCSK2, CUZD1 \n",
      "Negative:  BICC1, PKHD1, CFTR, SLC4A4, THSD4, PDE3A, MAML2, NRP1, SCTR, ANXA4 \n",
      "\t   FGFR2, AC000061.1, ADAMTS9-AS2, FRAS1, COL18A1, KCNJ16, RBMS3, DCDC2, GLIS3, WWTR1 \n",
      "\t   CFAP221, PLD1, ABCC3, ACSM3, AKAP7, PDGFD, GUCY1A2, ADAMTS9, LINC00472, KCNJ15 \n",
      "PC_ 3 \n",
      "Positive:  PKHD1, CFTR, SLC4A4, SCTR, FRAS1, ANXA4, DCDC2, THSD4, FGFR2, KCNJ16 \n",
      "\t   AC000061.1, CFAP221, CADPS, CCDC198, ACSM3, POU6F2, CASR, FAM189A1, CHST9, RALYL \n",
      "\t   KCNJ15, ABCC3, ERBB4, ESRRG, AC104035.1, PDE3A, AKAP7, SLC17A4, ANXA3, GLIS3 \n",
      "Negative:  LAMA2, COL6A3, PDGFRA, BNC2, ZFPM2, FBXL7, COL1A2, CRISPLD2, COL6A2, NID1 \n",
      "\t   FGF7, FBN1, C7, COL3A1, DCN, LHFPL6, NEGR1, SVEP1, ABCA8, ADH1B \n",
      "\t   DCLK1, COL4A1, CDH11, IGFBP4, NR2F1-AS1, TIMP3, MRC2, LAMA4, COL12A1, SMOC2 \n",
      "PC_ 4 \n",
      "Positive:  PRKG1, LAMA2, PDGFRA, SVEP1, COL6A3, GAREM1, NEGR1, CRISPLD2, TRHDE, FGF7 \n",
      "\t   COL1A2, C7, CATSPERB, CUZD1, DCN, COL6A2, ERRFI1, SLIT2, DCLK1, NR2F1-AS1 \n",
      "\t   ADH1B, COL5A2, COL3A1, PRICKLE2, PWRN1, SRPX2, FLRT2, NID1, SMOC2, SLC7A2 \n",
      "Negative:  DOCK2, PTPRC, TBXAS1, AOAH, PIK3R5, FYB1, INPP5D, MYO1F, MS4A6A, PRKCB \n",
      "\t   ARHGAP15, CD163, FLI1, LCP2, MSR1, ATP8B4, ELMO1, PECAM1, PTPRE, IKZF1 \n",
      "\t   ZEB2, DOCK10, STAB1, C22orf34, SLC11A1, SLC1A3, TGFBI, MEF2C, TFEC, ITGA4 \n",
      "PC_ 5 \n",
      "Positive:  ZNF385D, DACH2, TRPM3, LRRTM3, INS, LMNTD1, UNC5D, SYT16, LINC01099, PPM1E \n",
      "\t   NRG1, G6PC2, KIRREL3, SLCO5A1, AP000829.1, ZNF385D-AS2, KIAA1211, PLCH2, HDAC9, NRG3 \n",
      "\t   KCNH7, VAT1L, LINC01098, KCNH1, BMP5, KCTD8, BEST3, RASD1, PCDH7, CAPN13 \n",
      "Negative:  AL591501.1, FAP, LINC02388, GCG, PLPPR1, AL138694.1, DSCAM, TTR, SNTG1, GPC6 \n",
      "\t   TENM2, SPOCK3, KIAA1549L, PTPRT, BTBD11, SLC35F4, KCNIP1, FSTL5, AC090531.1, DCC \n",
      "\t   NPSR1-AS1, SYNDIG1, HECW2, PAPPA2, STK32B, AC020637.1, AC006369.1, PPP2R2B, POU6F2, KCNJ3 \n",
      "\n",
      "16:36:06 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:36:06 Read 5755 rows and found 50 numeric columns\n",
      "\n",
      "16:36:06 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:36:06 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:36:07 Writing NN index file to temp file /tmp/Rtmp95L5vt/file32037727f4c1e2\n",
      "\n",
      "16:36:07 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "16:36:08 Annoy recall = 100%\n",
      "\n",
      "16:36:11 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:36:16 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:36:16 Commencing optimization for 500 epochs, with 248378 positive edges\n",
      "\n",
      "16:36:25 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "16:37:54 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:37:54 Read 5755 rows and found 49 numeric columns\n",
      "\n",
      "16:37:54 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:37:54 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:37:55 Writing NN index file to temp file /tmp/Rtmp95L5vt/file32037743d2652b\n",
      "\n",
      "16:37:55 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "16:37:56 Annoy recall = 100%\n",
      "\n",
      "16:37:59 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:38:03 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:38:03 Commencing optimization for 500 epochs, with 239768 positive edges\n",
      "\n",
      "16:38:13 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "16:38:31 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:38:33 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "16:38:38 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:38:38 Commencing optimization for 500 epochs, with 173046 positive edges\n",
      "\n",
      "16:38:48 Optimization finished\n",
      "\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Normalizing layer: counts\n",
      "\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message:\n",
      "\"Different features in new layer data than already exists for scale.data\"\n",
      "PC_ 1 \n",
      "Positive:  KCNB2, RIMS2, SLC30A8, KCNMB2, NOL4, NRXN1, CACNA1A, ABCC8, CACNA1C, FMN2 \n",
      "\t   CACNA1D, ST18, CNTN1, UNC80, RIMBP2, RGS7, LRRTM4, RFX6, NLGN1, MYT1L \n",
      "\t   ROBO2, SNAP25, HMGCLL1, ELAVL4, DPP6, PCSK2, CACNA2D1, SOX5, TRPM3, STXBP5L \n",
      "Negative:  TC2N, PRSS1, TRHDE, CUZD1, CATSPERB, CTRB2, PNLIP, GULP1, ZNF804B, PRSS2 \n",
      "\t   CTRC, NTN4, GPT2, AC010967.1, REG1B, YBX3, MIR217HG, AASS, EIF2AK3, BTG2 \n",
      "\t   NEDD4L, ANPEP, PLA2G1B, AL159156.1, GAREM1, AMY2A, AC013652.1, ERRFI1, FAM107B, LINC02532 \n",
      "PC_ 2 \n",
      "Positive:  PWRN1, TRHDE, PRSS1, CUZD1, PAM, RIMS2, KCNB2, KCNMA1, SLC30A8, KCNMB2 \n",
      "\t   CACNA1A, MTUS2, ABCC8, NOL4, TMEM178B, NRXN1, DPP6, RGS7, ST18, MYT1L \n",
      "\t   RFX6, FMN2, MEG3, RIMBP2, ELAVL4, PCSK2, MAGI2, TC2N, LRRTM4, SYT14 \n",
      "Negative:  BICC1, PKHD1, CFTR, SLC4A4, THSD4, PDE3A, MAML2, NRP1, SCTR, ANXA4 \n",
      "\t   AC000061.1, FGFR2, ADAMTS9-AS2, FRAS1, COL18A1, DCDC2, PLD1, AKAP7, GLIS3, AUTS2 \n",
      "\t   WWTR1, CFAP221, KCNJ16, RBMS3, CCDC198, ADAMTS9, ABCC3, ACSM3, BCL2, LINC00472 \n",
      "PC_ 3 \n",
      "Positive:  PKHD1, CFTR, SLC4A4, SCTR, ANXA4, FRAS1, DCDC2, THSD4, FGFR2, CCDC198 \n",
      "\t   AC000061.1, KCNJ16, CFAP221, CADPS, POU6F2, ACSM3, CHST9, FAM189A1, CASR, ABCC3 \n",
      "\t   ERBB4, KCNJ15, ESRRG, WWC1, SLC17A4, AC104035.1, AKAP7, RALYL, ATP1B1, SLC28A3 \n",
      "Negative:  LAMA2, COL6A3, BNC2, FBXL7, COL1A2, ZFPM2, FBN1, COL3A1, NID1, COL6A2 \n",
      "\t   CRISPLD2, PDGFRA, DCN, LHFPL6, C7, ABCA8, FGF7, DCLK1, SVEP1, IGFBP4 \n",
      "\t   COL4A1, ADH1B, NR2F1-AS1, EBF1, ZEB2, NEGR1, LAMA4, SYNPO2, ZFPM2-AS1, CDH11 \n",
      "PC_ 4 \n",
      "Positive:  DOCK2, PTPRC, TBXAS1, PIK3R5, FYB1, AOAH, CD163, SLA, LCP2, MS4A6A \n",
      "\t   MSR1, PREX1, TFEC, INPP5D, MYO1F, FLI1, PRKCB, ARHGAP15, MS4A7, ATP8B4 \n",
      "\t   MRC1, KYNU, SLC1A3, IKZF1, HCK, CLEC7A, PTPRE, STAB1, DOCK11, ELMO1 \n",
      "Negative:  LAMA2, COL6A3, COL1A2, SVEP1, NEGR1, C7, COL3A1, FBN1, ABCA8, DCLK1 \n",
      "\t   PDGFRA, FGF7, DCN, NID1, COL4A1, COL6A2, COL5A2, COL6A1, ADH1B, CRISPLD2 \n",
      "\t   NR2F1-AS1, AC079298.3, PRRX1, SYNPO2, DNM3OS, GLI3, PRR16, CDH11, SRPX, SLIT2 \n",
      "PC_ 5 \n",
      "Positive:  DACH2, TRPM3, ZNF385D, LRRTM3, INS, SYT16, LMNTD1, LINC01099, UNC5D, PPM1E \n",
      "\t   NRG1, HDAC9, ZNF385D-AS2, AP000829.1, KIRREL3, VAT1L, KIAA1211, C1orf127, KCNH1, LINC01098 \n",
      "\t   SLCO5A1, NRG3, G6PC2, GLRA1, KCNH7, BMP5, PLCH2, VWDE, GRIN2A, KCTD8 \n",
      "Negative:  FAP, AL591501.1, LINC02388, PLPPR1, GCG, AL138694.1, DSCAM, KIAA1549L, SPOCK3, AC090531.1 \n",
      "\t   GPC6, BTBD11, SLC35F4, SNTG1, PTPRT, TENM2, TTR, KCNIP1, AC020637.1, FSTL5 \n",
      "\t   SYNDIG1, NPSR1-AS1, STK32B, MYO16, TMEM132D, PPP2R2B, MMP16, KCNJ3, SEMA3E, EYA1 \n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "16:41:21 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:41:21 Read 5137 rows and found 50 numeric columns\n",
      "\n",
      "16:41:21 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:41:21 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:41:22 Writing NN index file to temp file /tmp/Rtmp95L5vt/file320377453ebb98\n",
      "\n",
      "16:41:22 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "16:41:23 Annoy recall = 100%\n",
      "\n",
      "16:41:26 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:41:32 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:41:32 Commencing optimization for 500 epochs, with 220848 positive edges\n",
      "\n",
      "16:41:43 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "16:43:09 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:43:09 Read 5137 rows and found 49 numeric columns\n",
      "\n",
      "16:43:09 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:43:09 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:43:09 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203776083238b\n",
      "\n",
      "16:43:09 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "16:43:11 Annoy recall = 100%\n",
      "\n",
      "16:43:14 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:43:20 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:43:20 Commencing optimization for 500 epochs, with 215926 positive edges\n",
      "\n",
      "16:43:30 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "16:43:46 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:43:50 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "16:43:55 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:43:55 Commencing optimization for 500 epochs, with 155996 positive edges\n",
      "\n",
      "16:44:06 Optimization finished\n",
      "\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Normalizing layer: counts\n",
      "\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message:\n",
      "\"Different features in new layer data than already exists for scale.data\"\n",
      "PC_ 1 \n",
      "Positive:  RIMS2, NRXN1, SLC30A8, KCNMB2, KCNB2, NOL4, CACNA1A, ABCC8, FMN2, CACNA1C \n",
      "\t   ST18, LRRTM4, CACNA1D, UNC80, ELAVL4, CNTN1, SNAP25, NLGN1, RGS7, KCNJ6 \n",
      "\t   DPP6, TRPM3, RFX6, CACNA2D1, RIMBP2, HMGCLL1, SOX5, MYT1L, PCSK2, ROBO2 \n",
      "Negative:  PRSS1, CTRB1, TC2N, CUZD1, CPA1, SEL1L, CTRB2, CLPS, CATSPERB, PNLIP \n",
      "\t   PRSS2, CTRC, ZNF804B, NTN4, REG1B, AC010967.1, PEX5L, GPT2, MIR217HG, PHACTR1 \n",
      "\t   ANPEP, AASS, PLA2G1B, AMY2A, LINC02532, AL159156.1, AMY2B, AC013652.1, KANK4, IFRD1 \n",
      "PC_ 2 \n",
      "Positive:  PRSS1, CTRB1, CUZD1, PWRN1, SEL1L, CPA1, PDE3B, FKBP5, CLPS, PEX5L \n",
      "\t   HS6ST3, PLXNA2, PNLIP, CTRB2, CTRC, MEG3, ZNF804B, RIMS2, TC2N, ABCC8 \n",
      "\t   SLC30A8, KCNMB2, NOL4, KCNB2, TMEM178B, PRSS2, NRXN1, CACNA1A, KCNMA1, MTUS2 \n",
      "Negative:  BICC1, PKHD1, CFTR, SLC4A4, THSD4, PDE3A, SCTR, NRP1, ANXA4, MAML2 \n",
      "\t   FGFR2, FRAS1, COL18A1, ADAMTS9-AS2, GLIS3, ABCC3, AC000061.1, DCDC2, CFAP221, KCNJ16 \n",
      "\t   ACSM3, AUTS2, CCDC198, PLD1, PDGFD, KCNJ15, AKAP7, GUCY1A2, BCL2, ANXA3 \n",
      "PC_ 3 \n",
      "Positive:  PKHD1, SLC4A4, CFTR, SCTR, ANXA4, FRAS1, THSD4, DCDC2, FGFR2, KCNJ16 \n",
      "\t   ACSM3, CASR, CFAP221, CADPS, CCDC198, FAM189A1, AC000061.1, ABCC3, RALYL, CHST9 \n",
      "\t   ERBB4, ESRRG, POU6F2, KCNJ15, WWC1, FLRT2, AC104035.1, GLIS3, ROCR, SLC17A4 \n",
      "Negative:  LAMA2, ZFPM2, FBXL7, COL6A3, PDGFRA, COL3A1, BNC2, NID1, COL6A2, ZEB2 \n",
      "\t   C7, COL1A2, TIMP3, DCN, CRISPLD2, FBN1, A2M, ABCA8, LHFPL6, DNM3OS \n",
      "\t   SVEP1, COL4A1, LAMA4, PDGFRB, EBF1, ADAMTS12, DCLK1, NEGR1, NR2F1-AS1, SPARCL1 \n",
      "PC_ 4 \n",
      "Positive:  LAMA2, ZFPM2, PDGFRA, COL6A3, COL3A1, COL6A2, C7, FBXL7, NID1, NEGR1 \n",
      "\t   SVEP1, FBN1, DCN, SLIT2, COL1A2, NR2F1-AS1, COL4A1, DNM3OS, PDGFRB, FGF7 \n",
      "\t   COL4A2, ABCA8, ADAMTS12, FBLN1, ADH1B, EBF1, COL5A2, SMOC2, CRISPLD2, GLI3 \n",
      "Negative:  DOCK2, CD163, MSR1, PTPRC, ATP8B4, TBXAS1, PIK3R5, MS4A7, MYO1F, LCP2 \n",
      "\t   STAB1, MRC1, AOAH, FLI1, MS4A6A, ARHGAP15, FYB1, INPP5D, PTPRE, F13A1 \n",
      "\t   FGD2, TGFBI, HCK, PREX1, SLA, CD163L1, SLC1A3, ELMO1, SLCO2B1, ST8SIA4 \n",
      "PC_ 5 \n",
      "Positive:  ZNF385D, TRPM3, INS, DACH2, LRRTM3, SYT16, LMNTD1, LINC01099, NRG1, HDAC9 \n",
      "\t   UNC5D, PPM1E, PLCH2, KIRREL3, C1orf127, VAT1L, SLCO5A1, KIAA1211, G6PC2, LINC01098 \n",
      "\t   ANKRD34C-AS1, AP000829.1, DGKG, NRG3, KCNH7, KLHL1, PCDH7, ASB9, KCNH1, KCTD8 \n",
      "Negative:  AL591501.1, LINC02388, FAP, GCG, PLPPR1, DSCAM, TTR, GPC6, TENM2, KIAA1549L \n",
      "\t   SLC35F4, KCNIP1, SNTG1, HECW2, BTBD11, PTPRT, AL138694.1, FSTL5, SPOCK3, NPSR1-AS1 \n",
      "\t   PPP2R2B, ARX, KCNJ3, MYO16, PAPPA2, STK32B, AC090531.1, TM4SF4, PDZD2, AC020637.1 \n",
      "\n",
      "16:46:32 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:46:32 Read 4433 rows and found 50 numeric columns\n",
      "\n",
      "16:46:32 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:46:32 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:46:32 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203771caec04\n",
      "\n",
      "16:46:32 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "16:46:34 Annoy recall = 100%\n",
      "\n",
      "16:46:37 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:46:41 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:46:42 Commencing optimization for 500 epochs, with 189514 positive edges\n",
      "\n",
      "16:46:50 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "16:48:06 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:48:06 Read 4433 rows and found 49 numeric columns\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "16:48:06 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:48:06 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:48:07 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203777b6e82a0\n",
      "\n",
      "16:48:07 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "16:48:08 Annoy recall = 100%\n",
      "\n",
      "16:48:11 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:48:15 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:48:15 Commencing optimization for 500 epochs, with 183862 positive edges\n",
      "\n",
      "16:48:24 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "16:48:37 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:48:41 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "16:48:45 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:48:45 Commencing optimization for 500 epochs, with 132722 positive edges\n",
      "\n",
      "16:48:54 Optimization finished\n",
      "\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Normalizing layer: counts\n",
      "\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message:\n",
      "\"Different features in new layer data than already exists for scale.data\"\n",
      "PC_ 1 \n",
      "Positive:  CUZD1, CTRB2, PNLIP, ZNF804B, NTN4, PRSS2, AC010967.1, PEX5L, REG1B, EIF2AK3 \n",
      "\t   PRKG1, MIR217HG, AASS, ANPEP, AL159156.1, AMY2B, AC013652.1, AMY2A, CYP3A5, ZNF710 \n",
      "\t   AC002463.1, PWRN1, CLDN10-AS1, MSC-AS1, LMO3, MYOM1, ERRFI1, CA12, DDIT4, PAG1 \n",
      "Negative:  RIMS2, KCNB2, SLC30A8, NRXN1, KCNMB2, NOL4, CACNA1A, ABCC8, FMN2, ST18 \n",
      "\t   RGS7, CACNA1D, CACNA1C, UNC80, LRRTM4, CNTN1, RIMBP2, ROBO2, NLGN1, DPP6 \n",
      "\t   MYT1L, SNAP25, RFX6, HMGCLL1, KCNJ6, ELAVL4, PCSK2, CACNA2D1, STXBP5L, TRPM3 \n",
      "PC_ 2 \n",
      "Positive:  PWRN1, PDE3B, CUZD1, PEX5L, MEG3, PNLIP, ZNF804B, KCNMA1, CTRB2, EIF2AK3 \n",
      "\t   TMEM178B, MEG8, AC010967.1, RIMS2, ABCC8, MTUS2, KCNB2, SLC30A8, NOL4, DPP6 \n",
      "\t   KCNMB2, MIR217HG, NRXN1, SLC7A2, FMN2, RGS7, CACNA1A, REG1B, LINC01091, MYT1L \n",
      "Negative:  BICC1, PKHD1, CFTR, SLC4A4, THSD4, PDE3A, NRP1, MAML2, SCTR, ANXA4 \n",
      "\t   FGFR2, ADAMTS9-AS2, FRAS1, AC000061.1, CFAP221, GLIS3, RBMS3, DCDC2, KCNJ16, AUTS2 \n",
      "\t   ABCC3, ADAMTS9, PLD1, WWTR1, AKAP7, GUCY1A2, CRIM1, BCL2, AP002989.1, KCNJ15 \n",
      "PC_ 3 \n",
      "Positive:  LAMA2, FBXL7, BNC2, COL6A3, NID1, ZFPM2, ZEB2, PDGFRA, COL6A2, CRISPLD2 \n",
      "\t   COL1A2, LHFPL6, DCN, TIMP3, COL3A1, IRAK3, COL4A1, SPARCL1, C7, EBF1 \n",
      "\t   SLC9A9, A2M, SVEP1, NR2F1-AS1, LAMA4, ADH1B, FBN1, ABCA8, LRP1, PDGFRB \n",
      "Negative:  PKHD1, CFTR, SLC4A4, SCTR, ANXA4, THSD4, FRAS1, DCDC2, FGFR2, KCNJ16 \n",
      "\t   CFAP221, AC000061.1, ABCC3, ACSM3, POU6F2, FAM189A1, CADPS, KCNJ15, CASR, ERBB4 \n",
      "\t   BICC1, SLC17A4, AC104035.1, GLIS3, RALYL, ESRRG, CHST9, ROCR, PDE3A, DZIP1 \n",
      "PC_ 4 \n",
      "Positive:  LAMA2, NID1, COL6A3, PDGFRA, SVEP1, CRISPLD2, DCN, COL6A2, COL1A2, C7 \n",
      "\t   ZFPM2, FBXL7, COL4A1, COL3A1, NR2F1-AS1, PRKG1, NEGR1, ABCA8, EBF1, FGF7 \n",
      "\t   COL4A2, ADAMTS12, PDGFRB, ADH1B, COL5A2, FBN1, FBLN1, GLI3, DCLK1, COL5A1 \n",
      "Negative:  DOCK2, PTPRC, PIK3R5, FYB1, TBXAS1, CD163, AOAH, MSR1, MS4A6A, MYO1F \n",
      "\t   LCP2, MS4A7, ARHGAP15, INPP5D, STAB1, MRC1, ATP8B4, PRKCB, SLC1A3, IKZF1 \n",
      "\t   FGD2, SLC11A1, SLA, APBB1IP, FLI1, CSF2RA, ST8SIA4, LCP1, DOCK11, PTPRE \n",
      "PC_ 5 \n",
      "Positive:  AL591501.1, FAP, LINC02388, PLPPR1, GCG, AL138694.1, DSCAM, SPOCK3, KIAA1549L, GPC6 \n",
      "\t   KCNIP1, TENM2, PTPRT, AC090531.1, FSTL5, SNTG1, TTR, BTBD11, SLC35F4, SYNDIG1 \n",
      "\t   STK32B, MYO16, PAPPA2, AC020637.1, NPSR1-AS1, DCC, PPP2R2B, TM4SF4, AJ009632.2, ADAMTS18 \n",
      "Negative:  ZNF385D, INS, LRRTM3, TRPM3, DACH2, LMNTD1, SYT16, LINC01099, NRG1, PPM1E \n",
      "\t   C1orf127, KIAA1211, UNC5D, KIRREL3, BMP5, VAT1L, ZNF385D-AS2, NRG3, G6PC2, HDAC9 \n",
      "\t   SLCO5A1, LINC01098, PLCH2, PCDH7, AP000829.1, GLRA1, TFCP2L1, ANKRD34C-AS1, TMEM108, CAPN13 \n",
      "\n",
      "16:51:12 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:51:12 Read 7283 rows and found 50 numeric columns\n",
      "\n",
      "16:51:12 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:51:12 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:51:13 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203771f472f3b\n",
      "\n",
      "16:51:13 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "16:51:15 Annoy recall = 100%\n",
      "\n",
      "16:51:17 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:51:22 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:51:22 Commencing optimization for 500 epochs, with 316776 positive edges\n",
      "\n",
      "16:51:32 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "16:53:14 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:53:14 Read 7283 rows and found 49 numeric columns\n",
      "\n",
      "16:53:14 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:53:14 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:53:15 Writing NN index file to temp file /tmp/Rtmp95L5vt/file32037713d29f41\n",
      "\n",
      "16:53:15 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "16:53:17 Annoy recall = 100%\n",
      "\n",
      "16:53:20 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:53:24 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:53:24 Commencing optimization for 500 epochs, with 310300 positive edges\n",
      "\n",
      "16:53:35 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "16:53:55 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:53:58 Commencing smooth kNN distance calibration using 1 thread\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " with target n_neighbors = 20\n",
      "\n",
      "16:54:02 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:54:03 Commencing optimization for 500 epochs, with 223810 positive edges\n",
      "\n",
      "16:54:14 Optimization finished\n",
      "\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Normalizing layer: counts\n",
      "\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message:\n",
      "\"Different features in new layer data than already exists for scale.data\"\n",
      "PC_ 1 \n",
      "Positive:  CACNA1C, NRXN1, KCNMB2, SOX5, RIMS2, SLC30A8, NOL4, KCNB2, CACNA2D1, CACNA1A \n",
      "\t   RGS7, ST18, NLGN1, CACNA1D, FMN2, CNTN1, LRRTM4, HMGCLL1, ROBO2, ABCC8 \n",
      "\t   UNC80, RIMBP2, SNAP25, DPP6, NEGR1, RFX6, PCSK2, KCNJ6, DACH1, PLCB4 \n",
      "Negative:  CELA3A, REG1A, PRSS1, CTRB1, CELA3B, CEL, CLPS, CPA1, CATSPERB, PNLIP \n",
      "\t   CTRC, CTRB2, ZNF804B, PRSS2, NTN4, AC010967.1, REG1B, MIR217HG, EIF2AK3, PWRN1 \n",
      "\t   PLA2G1B, ANPEP, AL159156.1, AMY2A, AMY2B, AASS, AC013652.1, MYOM1, KANK4, AC002463.1 \n",
      "PC_ 2 \n",
      "Positive:  RIMS2, SLC30A8, NRXN1, NOL4, KCNB2, KCNMB2, CACNA1A, RGS7, FMN2, ABCC8 \n",
      "\t   ST18, DPP6, KCNMB2-AS1, LRRTM4, KCNJ6, RFX6, RIMBP2, ELAVL4, PCSK2, LINGO2 \n",
      "\t   SYT14, MTUS2, MYT1L, UNC80, SLC4A10, SNAP25, STXBP5L, TMEM132D, G6PC2, ROBO2 \n",
      "Negative:  RBMS3, COL6A3, COL4A2, COL4A1, LAMA2, EBF1, BNC2, COL1A2, COL6A2, ZFPM2 \n",
      "\t   FBXL7, PDGFRA, SVEP1, C7, DCN, NID1, BICC1, COL3A1, IGFBP7, CRISPLD2 \n",
      "\t   NRP1, CALD1, MAML2, PLXDC2, IGFBP4, FBN1, LRP1, FBLN1, SRPX, THBS1 \n",
      "PC_ 3 \n",
      "Positive:  PKHD1, CFTR, SLC4A4, SCTR, THSD4, ANXA4, BICC1, AC000061.1, FRAS1, DCDC2 \n",
      "\t   FGFR2, PDE3A, KCNJ16, GLIS3, AKAP7, CFAP221, ACSM3, KCNJ15, ERBB4, POU6F2 \n",
      "\t   FAM189A1, GUCY1A2, AC104035.1, PDGFD, PLD1, SLC17A4, RALYL, CADPS, LINC00472, CASR \n",
      "Negative:  COL6A3, LAMA2, COL1A2, FBXL7, COL6A2, C7, BNC2, DCN, PDGFRA, ZFPM2 \n",
      "\t   NID1, COL3A1, CRISPLD2, FBN1, SRPX, FBLN1, LHFPL6, IGFBP4, ZEB1, FGF7 \n",
      "\t   ADH1B, PID1, NEGR1, ABCA8, DCLK1, SMOC2, ABCA9, NAV3, COL5A1, PDGFRB \n",
      "PC_ 4 \n",
      "Positive:  COL6A3, LAMA2, C7, PDGFRA, COL1A2, COL6A2, DCN, NID1, SVEP1, ZFPM2 \n",
      "\t   COL3A1, NEGR1, FBXL7, COL4A1, FBN1, FBLN1, CRISPLD2, SRPX, FGF7, DCLK1 \n",
      "\t   COL6A1, ADH1B, ABCA8, NR2F1-AS1, COL4A2, PDGFRB, EBF1, ADAMTS12, SLIT2, CDH11 \n",
      "Negative:  DOCK2, PTPRC, TBXAS1, PIK3R5, FYB1, CD163, AOAH, MSR1, MS4A6A, ARHGAP15 \n",
      "\t   MS4A7, LCP2, MYO1F, CLEC7A, TFEC, INPP5D, ATP8B4, MRC1, SLA, PRKCB \n",
      "\t   SLC1A3, HCK, FGD2, FLI1, PREX1, ZEB2, MS4A4A, STAB1, ST8SIA4, CYBB \n",
      "PC_ 5 \n",
      "Positive:  ZNF385D, DACH2, TRPM3, LRRTM3, ZNF385D-AS2, LINC01099, NRG1, SYT16, LMNTD1, INS \n",
      "\t   UNC5D, LINC01098, KIAA1211, KIRREL3, PPM1E, GPM6A, SLCO5A1, G6PC2, BMP5, HDAC9 \n",
      "\t   VAT1L, AP000829.1, KCTD8, CTNNA3, KCNH7, NRG3, PLCH2, TMEM108, RASD1, KLHL1 \n",
      "Negative:  AL591501.1, GCG, LINC02388, FAP, PLPPR1, AC020637.1, TTR, DSCAM, AL138694.1, FSTL5 \n",
      "\t   AC090531.1, SPOCK3, SNTG1, GPC6, KIAA1549L, PTPRT, TENM2, SLC35F4, BTBD11, KCNIP1 \n",
      "\t   STK32B, HECW2, MYO16, NPSR1-AS1, SYNDIG1, KCNJ3, PPP2R2B, ARX, PAPPA2, MMP16 \n",
      "\n",
      "16:57:34 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:57:34 Read 18086 rows and found 50 numeric columns\n",
      "\n",
      "16:57:34 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "16:57:34 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "16:57:36 Writing NN index file to temp file /tmp/Rtmp95L5vt/file320377b2f8ecd\n",
      "\n",
      "16:57:36 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "16:57:42 Annoy recall = 100%\n",
      "\n",
      "16:57:45 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "16:57:50 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "16:57:51 Commencing optimization for 200 epochs, with 803586 positive edges\n",
      "\n",
      "16:58:02 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "17:01:38 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:01:38 Read 18086 rows and found 49 numeric columns\n",
      "\n",
      "17:01:38 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:01:38 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "17:01:41 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203776fdc4c65\n",
      "\n",
      "17:01:41 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "17:01:47 Annoy recall = 100%\n",
      "\n",
      "17:01:51 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "17:01:57 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "17:01:57 Commencing optimization for 200 epochs, with 785124 positive edges\n",
      "\n",
      "17:02:08 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "17:03:08 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:03:11 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "17:03:17 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "17:03:18 Commencing optimization for 200 epochs, with 567470 positive edges\n",
      "\n",
      "17:03:29 Optimization finished\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.4 GiB\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Normalizing layer: counts\n",
      "\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message:\n",
      "\"Different features in new layer data than already exists for scale.data\"\n",
      "PC_ 1 \n",
      "Positive:  CACNA1C, NRXN1, KCNMB2, RIMS2, SLC30A8, KCNB2, NOL4, SOX5, CACNA1A, CACNA2D1 \n",
      "\t   LRRTM4, RGS7, NLGN1, ST18, CNTN1, FMN2, HMGCLL1, CACNA1D, ABCC8, RIMBP2 \n",
      "\t   UNC80, ROBO2, RFX6, DPP6, PCSK2, KCNJ6, NEGR1, KCNMB2-AS1, DACH1, ELAVL4 \n",
      "Negative:  REG1A, CELA3A, PRSS1, CTRB1, CEL, CLPS, CPA1, CATSPERB, CTRC, PNLIP \n",
      "\t   CTRB2, ZNF804B, PRSS2, AC010967.1, NTN4, REG1B, MIR217HG, EIF2AK3, PWRN1, PLA2G1B \n",
      "\t   ANPEP, PLXNA2, AL159156.1, AMY2A, AMY2B, AASS, AC013652.1, MYOM1, CYP3A5, CLPSL1 \n",
      "PC_ 2 \n",
      "Positive:  RIMS2, SLC30A8, NRXN1, KCNB2, NOL4, KCNMB2, CACNA1A, RGS7, ABCC8, LRRTM4 \n",
      "\t   FMN2, RIMBP2, DPP6, RFX6, ST18, KCNMB2-AS1, KCNJ6, PCSK2, MTUS2, UNC80 \n",
      "\t   ELAVL4, LINGO2, SYT14, MYT1L, TMEM132D, SLC4A10, CNTN1, STXBP5L, GRIA2, ROBO2 \n",
      "Negative:  RBMS3, COL4A2, COL6A3, COL4A1, LAMA2, BNC2, COL6A2, EBF1, FBXL7, NID1 \n",
      "\t   ZFPM2, PDGFRA, COL1A2, DCN, CRISPLD2, SVEP1, IGFBP7, C7, BICC1, COL3A1 \n",
      "\t   NRP1, CALD1, MAML2, IGFBP4, PLXDC2, TIMP3, SRPX, MRC2, LRP1, FBN1 \n",
      "PC_ 3 \n",
      "Positive:  PKHD1, CFTR, SLC4A4, SCTR, THSD4, BICC1, ANXA4, AC000061.1, FRAS1, PDE3A \n",
      "\t   DCDC2, FGFR2, KCNJ16, GLIS3, AKAP7, CFAP221, ACSM3, KCNJ15, ERBB4, POU6F2 \n",
      "\t   GUCY1A2, PDGFD, PLD1, FAM189A1, SLC17A4, RALYL, AC104035.1, CADPS, LINC00472, CASR \n",
      "Negative:  COL6A3, LAMA2, COL6A2, DCN, C7, PDGFRA, FBXL7, BNC2, COL1A2, NID1 \n",
      "\t   ZFPM2, COL3A1, CRISPLD2, IGFBP4, FBN1, SRPX, ZEB1, FGF7, NEGR1, PID1 \n",
      "\t   MRC2, ABCA8, REG1A, PRSS1, LHFPL6, FBLN1, CELA3A, TIMP3, ADH1B, PDGFRB \n",
      "PC_ 4 \n",
      "Positive:  DOCK2, PTPRC, TBXAS1, CD163, MSR1, PIK3R5, MS4A6A, AOAH, FYB1, ARHGAP15 \n",
      "\t   MS4A7, TFEC, MYO1F, ATP8B4, PRKCB, CLEC7A, SLC1A3, LCP2, INPP5D, KYNU \n",
      "\t   HCLS1, SLA, FGD2, MRC1, HCK, NCKAP1L, ZEB2, SLC11A1, CYBB, APBB1IP \n",
      "Negative:  LAMA2, COL6A3, PDGFRA, DCN, C7, COL1A2, NID1, COL6A2, SVEP1, COL3A1 \n",
      "\t   ZFPM2, CRISPLD2, SRPX, COL4A1, COL6A1, FGF7, FBN1, NEGR1, FBLN1, ADH1B \n",
      "\t   FBXL7, ABCA8, PDGFRB, COL5A2, COL4A2, DCLK1, COL1A1, SMOC2, ADAMTS12, NR2F1-AS1 \n",
      "PC_ 5 \n",
      "Positive:  AL591501.1, GCG, LINC02388, FAP, PLPPR1, AC020637.1, DSCAM, AL138694.1, TTR, SPOCK3 \n",
      "\t   FSTL5, AC090531.1, SNTG1, GPC6, PTPRT, KIAA1549L, TENM2, SLC35F4, BTBD11, KCNIP1 \n",
      "\t   STK32B, NPSR1-AS1, HECW2, MYO16, SYNDIG1, SEMA3E, PAPPA2, MMP16, DCC, PCSK2 \n",
      "Negative:  ZNF385D, DACH2, LRRTM3, TRPM3, ZNF385D-AS2, SYT16, NRG1, UNC5D, LINC01099, LMNTD1 \n",
      "\t   INS, PPM1E, LINC01098, KIAA1211, HDAC9, GPM6A, KIRREL3, G6PC2, BMP5, SLCO5A1 \n",
      "\t   VAT1L, PLCH2, CTNNA3, NRG3, IAPP, ANKRD34C-AS1, KCTD8, TMEM108, AP000829.1, KCNH7 \n",
      "\n",
      "17:11:28 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:11:29 Read 18258 rows and found 50 numeric columns\n",
      "\n",
      "17:11:29 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:11:29 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "17:11:31 Writing NN index file to temp file /tmp/Rtmp95L5vt/file320377516de84\n",
      "\n",
      "17:11:31 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "17:11:38 Annoy recall = 100%\n",
      "\n",
      "17:11:42 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "17:11:46 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "17:11:48 Commencing optimization for 200 epochs, with 812150 positive edges\n",
      "\n",
      "17:12:01 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "17:16:12 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:16:12 Read 18258 rows and found 49 numeric columns\n",
      "\n",
      "17:16:12 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:16:12 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "17:16:15 Writing NN index file to temp file /tmp/Rtmp95L5vt/file32037726d5ef68\n",
      "\n",
      "17:16:15 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "17:16:23 Annoy recall = 100%\n",
      "\n",
      "17:16:26 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "17:16:31 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "17:16:31 Commencing optimization for 200 epochs, with 793352 positive edges\n",
      "\n",
      "17:16:43 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "17:17:52 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:17:54 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "17:17:59 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "17:17:59 Commencing optimization for 200 epochs, with 572504 positive edges\n",
      "\n",
      "17:18:13 Optimization finished\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.5 GiB\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Normalizing layer: counts\n",
      "\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message:\n",
      "\"Different features in new layer data than already exists for scale.data\"\n",
      "PC_ 1 \n",
      "Positive:  CACNA1C, NRXN1, SOX5, KCNMB2, RIMS2, SLC30A8, KCNB2, NOL4, CACNA2D1, CACNA1A \n",
      "\t   NLGN1, CNTN1, CACNA1D, RGS7, ST18, LRRTM4, FMN2, HMGCLL1, ROBO2, UNC80 \n",
      "\t   DPP6, ABCC8, RFX6, NEGR1, DACH1, NOVA1, PCSK2, KCNMB2-AS1, RIMBP2, SNAP25 \n",
      "Negative:  REG1A, CELA3A, PRSS1, CTRB1, CELA3B, CEL, CLPS, CPA1, CATSPERB, CTRC \n",
      "\t   CTRB2, PNLIP, PRSS2, ZNF804B, AC010967.1, REG1B, NTN4, MIR217HG, EIF2AK3, PWRN1 \n",
      "\t   PLA2G1B, ANPEP, AMY2B, AL159156.1, AMY2A, AC013652.1, AASS, MYOM1, DDIT4, AC002463.1 \n",
      "PC_ 2 \n",
      "Positive:  RIMS2, NRXN1, SLC30A8, NOL4, KCNMB2, KCNB2, CACNA1A, RGS7, ABCC8, FMN2 \n",
      "\t   ST18, KCNMB2-AS1, DPP6, RFX6, LRRTM4, PCSK2, KCNJ6, MTUS2, RIMBP2, ELAVL4 \n",
      "\t   UNC80, SYT14, MYT1L, LINGO2, SLC4A10, CNTN1, TMEM132D, SNTG1, STXBP5L, G6PC2 \n",
      "Negative:  RBMS3, COL6A3, COL4A2, COL4A1, LAMA2, EBF1, FBXL7, COL1A2, PDGFRA, COL6A2 \n",
      "\t   ZFPM2, NID1, BNC2, COL3A1, DCN, CRISPLD2, SVEP1, C7, IGFBP7, CALD1 \n",
      "\t   FBLN1, IGFBP4, BICC1, SRPX, THBS1, NRP1, FBN1, COL1A1, TIMP3, PDGFRB \n",
      "PC_ 3 \n",
      "Positive:  LAMA2, COL6A3, COL1A2, FBXL7, PDGFRA, COL6A2, ZFPM2, DCN, C7, COL3A1 \n",
      "\t   NID1, CRISPLD2, BNC2, NEGR1, FBN1, ZEB1, FBLN1, SRPX, IGFBP4, PID1 \n",
      "\t   NAV3, LHFPL6, FGF7, ABCA8, PRSS1, CELA3A, ADH1B, MRC2, PDGFRB, REG1A \n",
      "Negative:  PKHD1, CFTR, SLC4A4, SCTR, THSD4, ANXA4, BICC1, AC000061.1, FRAS1, DCDC2 \n",
      "\t   FGFR2, PDE3A, KCNJ16, GLIS3, CFAP221, AKAP7, ACSM3, ERBB4, POU6F2, KCNJ15 \n",
      "\t   GUCY1A2, AC104035.1, FAM189A1, PLD1, PDGFD, SLC17A4, CADPS, RALYL, LINC00472, CASR \n",
      "PC_ 4 \n",
      "Positive:  DOCK2, PTPRC, TBXAS1, PIK3R5, CD163, MS4A6A, FYB1, AOAH, MSR1, ARHGAP15 \n",
      "\t   MYO1F, LCP2, INPP5D, SLC1A3, CLEC7A, MS4A7, SLA, ZEB2, PRKCB, FLI1 \n",
      "\t   ATP8B4, PREX1, TFEC, HCK, MRC1, CYBB, KYNU, HCLS1, FGD2, APBB1IP \n",
      "Negative:  LAMA2, COL6A3, PDGFRA, COL1A2, DCN, C7, COL6A2, NID1, COL3A1, SVEP1 \n",
      "\t   FBLN1, NEGR1, CRISPLD2, ZFPM2, SRPX, COL6A1, COL5A2, ADH1B, ABCA8, COL4A1 \n",
      "\t   FGF7, NR2F1-AS1, DCLK1, PDGFRB, FBN1, FBXL7, COL1A1, ADAMTS12, C1R, COL4A2 \n",
      "PC_ 5 \n",
      "Positive:  ZNF385D, TRPM3, DACH2, LRRTM3, SYT16, ZNF385D-AS2, UNC5D, INS, NRG1, LMNTD1 \n",
      "\t   LINC01099, PPM1E, GPM6A, LINC01098, HDAC9, KIRREL3, VAT1L, G6PC2, PLCH2, SLCO5A1 \n",
      "\t   BMP5, KIAA1211, KCTD8, NRG3, KCNH7, CTNNA3, AP000829.1, RASD1, KCNH1, VWDE \n",
      "Negative:  AL591501.1, GCG, FAP, LINC02388, PLPPR1, AC020637.1, SPOCK3, TTR, DSCAM, FSTL5 \n",
      "\t   AL138694.1, AC090531.1, TENM2, SLC35F4, SNTG1, GPC6, KIAA1549L, PTPRT, NPSR1-AS1, BTBD11 \n",
      "\t   SYNDIG1, IRX2, KCNIP1, ADAMTS18, STK32B, MYO16, KCNJ3, HECW2, PAPPA2, MMP16 \n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "17:26:27 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:26:27 Read 18512 rows and found 50 numeric columns\n",
      "\n",
      "17:26:27 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:26:27 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "17:26:29 Writing NN index file to temp file /tmp/Rtmp95L5vt/file32037770ce8b71\n",
      "\n",
      "17:26:29 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "17:26:36 Annoy recall = 100%\n",
      "\n",
      "17:26:38 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "17:26:43 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "17:26:44 Commencing optimization for 200 epochs, with 822124 positive edges\n",
      "\n",
      "17:26:56 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "17:30:34 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:30:34 Read 18512 rows and found 49 numeric columns\n",
      "\n",
      "17:30:34 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:30:34 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "17:30:37 Writing NN index file to temp file /tmp/Rtmp95L5vt/file320377690c4813\n",
      "\n",
      "17:30:37 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "17:30:45 Annoy recall = 100%\n",
      "\n",
      "17:30:48 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "17:30:53 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "17:30:54 Commencing optimization for 200 epochs, with 807128 positive edges\n",
      "\n",
      "17:31:06 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "17:32:13 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:32:16 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "17:32:20 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "17:32:21 Commencing optimization for 200 epochs, with 580734 positive edges\n",
      "\n",
      "17:32:34 Optimization finished\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.6 GiB\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Normalizing layer: counts\n",
      "\n",
      "Finding variable features for layer counts\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "Warning message:\n",
      "\"Different features in new layer data than already exists for scale.data\"\n",
      "PC_ 1 \n",
      "Positive:  REG1A, CELA3A, PRSS1, CTRB1, CEL, CPA1, CATSPERB, PNLIP, CTRC, CTRB2 \n",
      "\t   ZNF804B, PRSS2, NTN4, REG1B, AC010967.1, MIR217HG, EIF2AK3, PLA2G1B, PWRN1, ANPEP \n",
      "\t   AL159156.1, PLXNA2, AMY2B, AASS, AMY2A, MYOM1, AC013652.1, CYP3A5, DDIT4, AC002463.1 \n",
      "Negative:  CACNA1C, NRXN1, SLC30A8, KCNMB2, RIMS2, KCNB2, NOL4, SOX5, CACNA1A, CACNA2D1 \n",
      "\t   LRRTM4, RGS7, CNTN1, NLGN1, FMN2, ST18, CACNA1D, HMGCLL1, ABCC8, ROBO2 \n",
      "\t   RFX6, DPP6, RIMBP2, PCSK2, UNC80, SNAP25, KCNJ6, NEGR1, KCNMB2-AS1, TMEM132D \n",
      "PC_ 2 \n",
      "Positive:  COL4A1, COL4A2, RBMS3, COL6A3, LAMA2, ZFPM2, COL6A2, FBXL7, EBF1, NID1 \n",
      "\t   COL1A2, BNC2, SVEP1, DCN, CRISPLD2, PDGFRA, COL3A1, C7, IGFBP7, IGFBP4 \n",
      "\t   FBN1, CALD1, SRPX, PLXDC2, FBLN1, FGF7, TIMP3, ABCA8, NR2F1-AS1, MAML2 \n",
      "Negative:  RIMS2, NRXN1, SLC30A8, NOL4, KCNB2, KCNMB2, CACNA1A, RGS7, ABCC8, FMN2 \n",
      "\t   LRRTM4, RFX6, ST18, RIMBP2, DPP6, PCSK2, KCNJ6, KCNMB2-AS1, MTUS2, MYT1L \n",
      "\t   ELAVL4, UNC80, TMEM132D, SYT14, LINGO2, SLC4A10, STXBP5L, CNTN1, G6PC2, SNAP25 \n",
      "PC_ 3 \n",
      "Positive:  LAMA2, COL6A3, ZFPM2, COL6A2, COL1A2, DCN, FBXL7, NID1, C7, BNC2 \n",
      "\t   PDGFRA, CELA3A, REG1A, COL3A1, CRISPLD2, PRSS1, FBN1, PID1, CTRB1, ZEB1 \n",
      "\t   NEGR1, FGF7, SRPX, LHFPL6, IGFBP4, ABCA8, CEL, FBLN1, NAV3, SMOC2 \n",
      "Negative:  PKHD1, CFTR, SLC4A4, SCTR, THSD4, BICC1, ANXA4, FRAS1, AC000061.1, DCDC2 \n",
      "\t   PDE3A, FGFR2, GLIS3, KCNJ16, AKAP7, CFAP221, ACSM3, KCNJ15, ERBB4, POU6F2 \n",
      "\t   PLD1, AC104035.1, PDGFD, GUCY1A2, FAM189A1, SLC17A4, RALYL, CADPS, LINC00472, CASR \n",
      "PC_ 4 \n",
      "Positive:  LAMA2, COL6A3, PDGFRA, COL1A2, C7, DCN, PRKG1, SVEP1, COL6A2, NID1 \n",
      "\t   COL3A1, ZFPM2, FGF7, ABCA8, FBN1, COL6A1, FBLN1, NEGR1, COL4A1, SRPX \n",
      "\t   NR2F1-AS1, ADH1B, CRISPLD2, FBXL7, DCLK1, SMOC2, DNM3OS, SLIT2, PTPRG, COL5A2 \n",
      "Negative:  TBXAS1, DOCK2, PTPRC, PIK3R5, CD163, MS4A6A, FYB1, AOAH, MSR1, ARHGAP15 \n",
      "\t   MS4A7, SLA, MYO1F, LCP2, TFEC, ATP8B4, INPP5D, CLEC7A, KYNU, SLC1A3 \n",
      "\t   PRKCB, ZEB2, APBB1IP, FGD2, SLC11A1, MRC1, PREX1, FLI1, CYBB, IKZF1 \n",
      "PC_ 5 \n",
      "Positive:  AL591501.1, FAP, LINC02388, GCG, PLPPR1, AC020637.1, DSCAM, SPOCK3, TTR, AL138694.1 \n",
      "\t   FSTL5, SNTG1, AC090531.1, KIAA1549L, SLC35F4, GPC6, TENM2, PTPRT, BTBD11, NPSR1-AS1 \n",
      "\t   KCNIP1, MYO16, PAPPA2, STK32B, SYNDIG1, KCNJ3, SEMA3E, MMP16, ADAMTS18, HECW2 \n",
      "Negative:  ZNF385D, DACH2, LRRTM3, TRPM3, ZNF385D-AS2, LINC01099, NRG1, LMNTD1, INS, SYT16 \n",
      "\t   PPM1E, UNC5D, KIRREL3, LINC01098, KIAA1211, G6PC2, VAT1L, GPM6A, HDAC9, SLCO5A1 \n",
      "\t   CTNNA3, AP000829.1, BMP5, KCTD8, KCNH7, PLCH2, NRG3, IAPP, RASD1, KCNH1 \n",
      "\n",
      "17:40:35 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:40:35 Read 19128 rows and found 50 numeric columns\n",
      "\n",
      "17:40:35 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:40:35 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "17:40:38 Writing NN index file to temp file /tmp/Rtmp95L5vt/file32037731936eef\n",
      "\n",
      "17:40:38 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "17:40:45 Annoy recall = 100%\n",
      "\n",
      "17:40:48 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "17:40:53 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "17:40:54 Commencing optimization for 200 epochs, with 852924 positive edges\n",
      "\n",
      "17:41:07 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "17:44:52 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:44:52 Read 19128 rows and found 49 numeric columns\n",
      "\n",
      "17:44:52 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:44:52 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "17:44:54 Writing NN index file to temp file /tmp/Rtmp95L5vt/file32037729305ed3\n",
      "\n",
      "17:44:54 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "17:45:01 Annoy recall = 100%\n",
      "\n",
      "17:45:04 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "17:45:08 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "17:45:09 Commencing optimization for 200 epochs, with 830408 positive edges\n",
      "\n",
      "17:45:22 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "17:46:28 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "17:46:30 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "17:46:35 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "17:46:36 Commencing optimization for 200 epochs, with 599484 positive edges\n",
      "\n",
      "17:46:49 Optimization finished\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.7 GiB\"\n"
     ]
    }
   ],
   "source": [
    "for (samp in samples) {\n",
    "    pdf(paste0(amulet_dir, samp, '/', samp, \"_amulet_plots.pdf\"))\n",
    "    adata <- readRDS(paste0(indir, samp,\"_\",modality,\".RDS\"))\n",
    "    gc()\n",
    "    \n",
    "    multiplets <- read.table(paste0(amulet_dir, samp, '/MultipletBarcodes_01.txt'))\n",
    "    \n",
    "    adata[[]]$amulet_doublet <- FALSE\n",
    "    adata[[]][multiplets$V1,]$amulet_doublet <- TRUE\n",
    "    \n",
    "    #plot UMAPs of WNN\n",
    "    p1 <- DimPlot(adata, reduction='umap.wnn', group.by='seurat_clusters', label=TRUE, label.size=6, repel=TRUE)\n",
    "    p1 <- p1 + xlab('UMAP 1') + ylab('UMAP 2') + ggtitle(paste0('WNN - ', samp))\n",
    "    p1 <- p1 + NoLegend() & theme(plot.title=element_text(hjust=0.5))\n",
    "    print(p1)\n",
    "    \n",
    "    #plot UMAPs of WNN\n",
    "    p1 <- DimPlot(adata, reduction='umap.wnn', group.by='amulet_doublet', label=TRUE, label.size=6, repel=TRUE)\n",
    "    p1 <- p1 + xlab('UMAP 1') + ylab('UMAP 2') + ggtitle(paste0('WNN - ', samp))\n",
    "    p1 <- p1 + NoLegend() & theme(plot.title=element_text(hjust=0.5)) \n",
    "    p1 <- p1 + annotate('text',x=Inf, y=Inf, hjust=1, vjust=1, label=paste0(\"# Doublets: \",sum(adata[[]]$amulet_doublet)))  + \n",
    "         annotate('text',x=Inf, y=Inf, hjust=1, vjust=3, \n",
    "             label=paste0(\"% Doublets: \", format(sum(adata[[]]$amulet_doublet) * 100 / nrow(adata[[]]), digits=2, nsmall=2), \"%\"))\n",
    "    print(p1)\n",
    "    \n",
    "    #plot UMAPs of ATAC\n",
    "    p1 <- DimPlot(adata, reduction='umap.atac', group.by='seurat_clusters', label=TRUE, label.size=6, repel=TRUE)\n",
    "    p1 <- p1 + xlab('UMAP 1') + ylab('UMAP 2') + ggtitle(paste0('ATAC - ', samp))\n",
    "    p1 <- p1 + NoLegend() & theme(plot.title=element_text(hjust=0.5))\n",
    "    print(p1)\n",
    "    \n",
    "    #plot UMAPs of ATAC\n",
    "    p1 <- DimPlot(adata, reduction='umap.atac', group.by='amulet_doublet', label=TRUE, label.size=6, repel=TRUE)\n",
    "    p1 <- p1 + xlab('UMAP 1') + ylab('UMAP 2') + ggtitle(paste0('ATAC - ', samp))\n",
    "    p1 <- p1 + NoLegend() & theme(plot.title=element_text(hjust=0.5)) \n",
    "    p1 <- p1 + annotate('text',x=Inf, y=Inf, hjust=1, vjust=1, label=paste0(\"# Doublets: \",sum(adata[[]]$amulet_doublet)))  + \n",
    "         annotate('text',x=Inf, y=Inf, hjust=1, vjust=3, \n",
    "             label=paste0(\"% Doublets: \", format(sum(adata[[]]$amulet_doublet) * 100 / nrow(adata[[]]), digits=2, nsmall=2), \"%\"))\n",
    "    print(p1)\n",
    "    \n",
    "    \n",
    "    p1 <- VlnPlot(adata, features='nFeature_ATAC', group.by='amulet_doublet', pt.size=0, log=TRUE) + \n",
    "        geom_boxplot(width=.6, fill='white', alpha=.6, pt.size=0) + \n",
    "        geom_hline(yintercept=median(adata$nFeature_ATAC), linetype='dashed', lw=2) + \n",
    "        ggtitle(paste0('ATAC Features -  ', samp))\n",
    "    \n",
    "    p2 <- VlnPlot(adata, features='nCount_ATAC', group.by='amulet_doublet', pt.size=0, log=TRUE) + \n",
    "        geom_boxplot(width=.6, fill='white', alpha=.6, pt.size=0) + \n",
    "        geom_hline(yintercept=median(adata$nCount_ATAC), linetype='dashed', lw=2) + \n",
    "        ggtitle(paste0('ATAC Counts -  ', samp))\n",
    "    \n",
    "    p3 <- VlnPlot(adata, features='nFeature_RNA', group.by='amulet_doublet', pt.size=0, log=TRUE) + \n",
    "        geom_boxplot(width=.6, fill='white', alpha=.6, pt.size=0) + \n",
    "        geom_hline(yintercept=median(adata$nFeature_RNA), linetype='dashed', lw=2) + \n",
    "        ggtitle(paste0('RNA Features -  ', samp))\n",
    "    \n",
    "    p4 <- VlnPlot(adata, features='nCount_RNA', group.by='amulet_doublet', pt.size=0, log=TRUE) + \n",
    "        geom_boxplot(width=.6, fill='white', alpha=.6, pt.size=0) + \n",
    "        geom_hline(yintercept=median(adata$nCount_RNA), linetype='dashed', lw=2) + \n",
    "        ggtitle(paste0('RNA Counts -  ', samp))\n",
    "    print(p1/p2)\n",
    "    print(p3/p4)\n",
    "    \n",
    "    adata <- subset(\n",
    "      x = adata,\n",
    "      subset = amulet_doublet==FALSE\n",
    "    )\n",
    "    \n",
    "    adata <- multiome_cluster_single_samp(adata)\n",
    "    \n",
    "    #plot UMAPs of WNN\n",
    "    p1 <- DimPlot(adata, reduction='umap.wnn', group.by='seurat_clusters', label=TRUE, label.size=6, repel=TRUE)\n",
    "    p1 <- p1 + xlab('UMAP 1') + ylab('UMAP 2') + ggtitle(paste0('WNN - ', samp))\n",
    "    p1 <- p1 + NoLegend() & theme(plot.title=element_text(hjust=0.5))\n",
    "    print(p1)\n",
    "    \n",
    "    #plot UMAPs of ATAC\n",
    "    p1 <- DimPlot(adata, reduction='umap.atac', group.by='seurat_clusters', label=TRUE, label.size=6, repel=TRUE)\n",
    "    p1 <- p1 + xlab('UMAP 1') + ylab('UMAP 2') + ggtitle(paste0('ATAC - ', samp))\n",
    "    p1 <- p1 + NoLegend() & theme(plot.title=element_text(hjust=0.5))\n",
    "    print(p1)\n",
    "    \n",
    "    saveRDS(adata, paste0(outdir, samp,\"_\",modality,\".RDS\"))\n",
    "    \n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove Background Noise from RNA Counts with DecontX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "library(celda)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker.genes <- c('INS','GCG','MKI67','SST','PPY','GHRL', #Beta, alpha, delta, gamma, epsilon, cycling\n",
    "                  'CTRB2','PRSS2','CPA1','REG1A', #Acinar\n",
    "                  'CFTR','MUC5B', #Ductal + muc5b+\n",
    "                  'COL6A3','PDGFRB','IGA1','SPARCL1','SPARC','COL1A1', #2x general stellate, 2x quiescent, 2x activated\n",
    "                  'PLVAP','VWF','LYVE1','PROX1','PMP22', #Endothelial x2, lymph endo 2x, schwann\n",
    "                  'C1QA','C1QB','C1QC', # Macrophage\n",
    "                  'CD3D','CD4','CD8B','CXCR4') # T cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Multiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'MM_1349_1_2_MM_1345_1_2'</li><li>'MM_1350_1_2_MM_1346_1_2'</li><li>'MM_1351_1_2_MM_1347_1_2'</li><li>'MM_1352_1_2_MM_1348_1_2'</li><li>'QY_2312_1_2_QY_2311_1_2'</li><li>'QY_2314_1_2_QY_2313_1_2'</li><li>'QY_2316_1_2_QY_2315_1_2'</li><li>'QY_2318_1_2_QY_2317_1_2'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'MM\\_1349\\_1\\_2\\_MM\\_1345\\_1\\_2'\n",
       "\\item 'MM\\_1350\\_1\\_2\\_MM\\_1346\\_1\\_2'\n",
       "\\item 'MM\\_1351\\_1\\_2\\_MM\\_1347\\_1\\_2'\n",
       "\\item 'MM\\_1352\\_1\\_2\\_MM\\_1348\\_1\\_2'\n",
       "\\item 'QY\\_2312\\_1\\_2\\_QY\\_2311\\_1\\_2'\n",
       "\\item 'QY\\_2314\\_1\\_2\\_QY\\_2313\\_1\\_2'\n",
       "\\item 'QY\\_2316\\_1\\_2\\_QY\\_2315\\_1\\_2'\n",
       "\\item 'QY\\_2318\\_1\\_2\\_QY\\_2317\\_1\\_2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'MM_1349_1_2_MM_1345_1_2'\n",
       "2. 'MM_1350_1_2_MM_1346_1_2'\n",
       "3. 'MM_1351_1_2_MM_1347_1_2'\n",
       "4. 'MM_1352_1_2_MM_1348_1_2'\n",
       "5. 'QY_2312_1_2_QY_2311_1_2'\n",
       "6. 'QY_2314_1_2_QY_2313_1_2'\n",
       "7. 'QY_2316_1_2_QY_2315_1_2'\n",
       "8. 'QY_2318_1_2_QY_2317_1_2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"MM_1349_1_2_MM_1345_1_2\" \"MM_1350_1_2_MM_1346_1_2\"\n",
       "[3] \"MM_1351_1_2_MM_1347_1_2\" \"MM_1352_1_2_MM_1348_1_2\"\n",
       "[5] \"QY_2312_1_2_QY_2311_1_2\" \"QY_2314_1_2_QY_2313_1_2\"\n",
       "[7] \"QY_2316_1_2_QY_2315_1_2\" \"QY_2318_1_2_QY_2317_1_2\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Set some values and load sample list\n",
    "dataset <- 'nPOD'\n",
    "modality <- 'Pooled'\n",
    "\n",
    "#sample isnt where youre pulling samples from its just to get the file names \n",
    "samp_dir <- paste0(\"/nfs/lab/projects/mega_pancreas/data/cellranger/\",dataset,\"/\", modality, \"/\")\n",
    "samples <- list.dirs(samp_dir, full.names = FALSE, recursive = FALSE)\n",
    "samples = samples[samples != \"intermediates\"]\n",
    "samples\n",
    "\n",
    "####MAKE SURE YOU ARE USING THE AMULET FILTERED RDS NOT THE RAW RDS\n",
    "#here this is fine youre saving it by name to another directory \n",
    "indir <- paste0(\"/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/nPOD/231121_WE_SSP_Intermediates/Amulet_adatas/\")\n",
    "outdir <- paste0(\"/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/\", dataset,\"/231121_WE_SSP_Intermediates/DecontX_adatas/\")\n",
    "pdf_dir <- paste0('/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/', dataset,'/231121_WE_SSP_Intermediates/DecontX_Plots/')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------\n",
      "\n",
      "Starting DecontX\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Fri Dec 13 11:20:13 2024 .. Analyzing all cells\n",
      "\n",
      "Fri Dec 13 11:20:13 2024 .... Generating UMAP and estimating cell types\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Fri Dec 13 11:20:46 2024 .... Estimating contamination\n",
      "\n",
      "Fri Dec 13 11:20:48 2024 ...... Completed iteration: 10 | converge: 0.02407\n",
      "\n",
      "Fri Dec 13 11:20:49 2024 ...... Completed iteration: 20 | converge: 0.01138\n",
      "\n",
      "Fri Dec 13 11:20:50 2024 ...... Completed iteration: 30 | converge: 0.006882\n",
      "\n",
      "Fri Dec 13 11:20:52 2024 ...... Completed iteration: 40 | converge: 0.004842\n",
      "\n",
      "Fri Dec 13 11:20:53 2024 ...... Completed iteration: 50 | converge: 0.003421\n",
      "\n",
      "Fri Dec 13 11:20:54 2024 ...... Completed iteration: 60 | converge: 0.002542\n",
      "\n",
      "Fri Dec 13 11:20:56 2024 ...... Completed iteration: 70 | converge: 0.001859\n",
      "\n",
      "Fri Dec 13 11:20:57 2024 ...... Completed iteration: 80 | converge: 0.001287\n",
      "\n",
      "Fri Dec 13 11:20:59 2024 ...... Completed iteration: 90 | converge: 0.001094\n",
      "\n",
      "Fri Dec 13 11:21:00 2024 ...... Completed iteration: 100 | converge: 0.001096\n",
      "\n",
      "Fri Dec 13 11:21:01 2024 ...... Completed iteration: 106 | converge: 0.0009941\n",
      "\n",
      "Fri Dec 13 11:21:01 2024 .. Calculating final decontaminated matrix\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Completed DecontX. Total time: 49.43444 secs\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.6 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.6 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message:\n",
      "\"Key 'rna_' taken, using 'rnaraw_' instead\"\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  CEL, TRHDE, CUZD1, SEL1L, CPA1, CATSPERB, LRIG1, CTRB2, CLPS, PNLIP \n",
      "\t   PRSS2, SLC16A10, GULP1, REG1B, CTRC, PEX5L, ZNF804B, AC010967.1, NTN4, GPT2 \n",
      "\t   MIR217HG, EIF2AK3, AASS, PWRN1, BTG2, PLA2G1B, AMY2B, ANPEP, PLXNA2, BNIP3 \n",
      "Negative:  CACNA1D, RIMS2, KCNMB2, DPP6, NRXN1, CACNA1C, AGBL4, CACNA1A, KCNB2, SLC30A8 \n",
      "\t   NOL4, RGS7, DACH1, CADPS, FMN2, NLGN1, UNC80, ABCC8, ST18, STXBP5L \n",
      "\t   SOX5, CACNA2D1, LRRTM4, MAPK10, MYT1L, MTUS2, MEIS2, CADM1, PLCB4, RIMBP2 \n",
      "PC_ 2 \n",
      "Positive:  RIMS2, KCNMB2, SLC30A8, CACNA1A, NRXN1, KCNB2, PDE3B, DPP6, RGS7, ABCC8 \n",
      "\t   NOL4, MTUS2, FMN2, ST18, LRRTM4, MYT1L, MAGI2, RIMBP2, KCNMA1, UNC80 \n",
      "\t   PAM, STXBP5L, KCNMB2-AS1, TMEM178B, PCSK2, KCNJ6, RFX6, SNAP25, TRPM3, SYT14 \n",
      "Negative:  BICC1, PKHD1, CFTR, SLC4A4, ANXA4, THSD4, PDE3A, DCDC2, SCTR, FRAS1 \n",
      "\t   ABCC3, ACSM3, FGFR2, PDGFD, AKAP7, NRP1, WWTR1, PLD1, AC000061.1, COL18A1 \n",
      "\t   MAML2, CFAP221, KCNJ16, GUCY1A2, ADAMTS9-AS2, LINC00472, RBMS3, GLIS3, ANXA2, EPB41L4A \n",
      "PC_ 3 \n",
      "Positive:  SLC4A4, PKHD1, CFTR, FRAS1, SCTR, ANXA4, DCDC2, CADPS, ACSM3, THSD4 \n",
      "\t   FGFR2, CFAP221, KCNJ16, FAM189A1, CHST9, ESRRG, CCDC198, ERBB4, ABCC3, POU6F2 \n",
      "\t   CASR, AC000061.1, RALYL, ATP1B1, KIFC3, KCNJ15, AKAP7, PDE8B, LINC00511, FLRT2 \n",
      "Negative:  LAMA2, COL6A3, ZFPM2, BNC2, PDGFRA, FBXL7, FBN1, CRISPLD2, COL1A2, C7 \n",
      "\t   COL6A2, FGF7, NID1, LHFPL6, COL3A1, DCN, DCLK1, SVEP1, NR2F1-AS1, ABCA8 \n",
      "\t   NEGR1, COL4A1, THBS1, IGFBP4, CDH11, MRC2, GFPT2, ADH1B, COL12A1, TIMP3 \n",
      "PC_ 4 \n",
      "Positive:  PRKG1, LAMA2, COL6A3, PDGFRA, SVEP1, C7, FGF7, NEGR1, DCN, CRISPLD2 \n",
      "\t   COL1A2, DCLK1, COL5A2, COL3A1, COL6A2, NR2F1-AS1, GAREM1, COL6A1, FLRT2, SLIT2 \n",
      "\t   ABCA9, NID1, SMOC2, ADH1B, PRICKLE2, TRHDE, ROR2, ABCA8, CATSPERB, SOX6 \n",
      "Negative:  DOCK2, PTPRC, TBXAS1, AOAH, ARHGAP15, FYB1, INPP5D, PIK3R5, MYO1F, PREX1 \n",
      "\t   PRKCB, MS4A6A, FLI1, CD163, LCP2, ELMO1, MSR1, ATP8B4, ZEB2, FGD2 \n",
      "\t   PTPRE, IKZF1, PECAM1, SLC1A3, DOCK10, MCTP1, STAB1, C22orf34, CELF2, MEF2C \n",
      "PC_ 5 \n",
      "Positive:  ZNF385D, INS, DACH2, TRPM3, LRRTM3, UNC5D, LMNTD1, SYT16, PPM1E, LINC01099 \n",
      "\t   NRG1, SLCO5A1, AP000829.1, KIRREL3, G6PC2, KIAA1211, ZNF385D-AS2, HDAC9, PLCH2, NRG3 \n",
      "\t   VAT1L, LINC01098, KCNH1, HADH, KCTD8, KCNH7, BMP5, RASD1, TMEM108, BEST3 \n",
      "Negative:  AL591501.1, FAP, GCG, LINC02388, PLPPR1, GPC6, AL138694.1, DSCAM, TENM2, TTR \n",
      "\t   SNTG1, KIAA1549L, SPOCK3, PTPRT, BTBD11, KCNIP1, SLC35F4, FSTL5, NPSR1-AS1, AC090531.1 \n",
      "\t   DCC, HECW2, PAPPA2, STK32B, SYNDIG1, AC020637.1, MYO16, PPP2R2B, ARX, GALNT13 \n",
      "\n",
      "11:21:24 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:21:24 Read 5755 rows and found 50 numeric columns\n",
      "\n",
      "11:21:24 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:21:24 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "11:21:26 Writing NN index file to temp file /tmp/Rtmp95L5vt/file32037714dfd1c1\n",
      "\n",
      "11:21:26 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "11:21:27 Annoy recall = 100%\n",
      "\n",
      "11:21:31 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "11:21:36 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "11:21:36 Commencing optimization for 500 epochs, with 237430 positive edges\n",
      "\n",
      "11:21:45 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "11:23:10 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:23:10 Read 5755 rows and found 49 numeric columns\n",
      "\n",
      "11:23:10 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:23:10 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "11:23:11 Writing NN index file to temp file /tmp/Rtmp95L5vt/file32037772f93f68\n",
      "\n",
      "11:23:11 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "11:23:12 Annoy recall = 100%\n",
      "\n",
      "11:23:16 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "11:23:20 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "11:23:20 Commencing optimization for 500 epochs, with 239768 positive edges\n",
      "\n",
      "11:23:28 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "11:23:43 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:23:45 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "11:23:49 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "11:23:49 Commencing optimization for 500 epochs, with 171030 positive edges\n",
      "\n",
      "11:23:58 Optimization finished\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Starting DecontX\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Fri Dec 13 11:26:03 2024 .. Analyzing all cells\n",
      "\n",
      "Fri Dec 13 11:26:03 2024 .... Generating UMAP and estimating cell types\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Fri Dec 13 11:26:24 2024 .... Estimating contamination\n",
      "\n",
      "Fri Dec 13 11:26:26 2024 ...... Completed iteration: 10 | converge: 0.03343\n",
      "\n",
      "Fri Dec 13 11:26:27 2024 ...... Completed iteration: 20 | converge: 0.01386\n",
      "\n",
      "Fri Dec 13 11:26:29 2024 ...... Completed iteration: 30 | converge: 0.007862\n",
      "\n",
      "Fri Dec 13 11:26:30 2024 ...... Completed iteration: 40 | converge: 0.005211\n",
      "\n",
      "Fri Dec 13 11:26:32 2024 ...... Completed iteration: 50 | converge: 0.003707\n",
      "\n",
      "Fri Dec 13 11:26:33 2024 ...... Completed iteration: 60 | converge: 0.002718\n",
      "\n",
      "Fri Dec 13 11:26:35 2024 ...... Completed iteration: 70 | converge: 0.001881\n",
      "\n",
      "Fri Dec 13 11:26:36 2024 ...... Completed iteration: 80 | converge: 0.001404\n",
      "\n",
      "Fri Dec 13 11:26:37 2024 ...... Completed iteration: 90 | converge: 0.001165\n",
      "\n",
      "Fri Dec 13 11:26:39 2024 ...... Completed iteration: 100 | converge: 0.001012\n",
      "\n",
      "Fri Dec 13 11:26:39 2024 ...... Completed iteration: 101 | converge: 0.0009984\n",
      "\n",
      "Fri Dec 13 11:26:39 2024 .. Calculating final decontaminated matrix\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Completed DecontX. Total time: 37.99273 secs\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.4 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.4 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message:\n",
      "\"Key 'rna_' taken, using 'rnaraw_' instead\"\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  PRSS1, TRHDE, CUZD1, CPA1, CATSPERB, CTRB2, PNLIP, GULP1, PRSS2, CLPS \n",
      "\t   CTRC, REG1B, ZNF804B, PEX5L, NTN4, GPT2, AC010967.1, AASS, EIF2AK3, MIR217HG \n",
      "\t   BTG2, BNIP3, PLA2G1B, ANPEP, AC013652.1, GAREM1, AC002463.1, AMY2B, AMY2A, AL159156.1 \n",
      "Negative:  RIMS2, DPP6, NRXN1, CACNA1D, KCNMB2, KCNB2, CACNA1C, SLC30A8, CACNA1A, NOL4 \n",
      "\t   UNC80, MTUS2, ST18, ABCC8, AGBL4, RGS7, FMN2, STXBP5L, LRRTM4, CNTN1 \n",
      "\t   MYT1L, NLGN1, RIMBP2, PLCB4, CACNA2D1, ROBO2, SOX5, DACH1, RFX6, PCSK2 \n",
      "PC_ 2 \n",
      "Positive:  BICC1, SLC4A4, PKHD1, CFTR, ANXA4, THSD4, PDE3A, MAML2, DCDC2, NRP1 \n",
      "\t   FRAS1, SCTR, AKAP7, RBMS3, ACSM3, PLD1, WWTR1, FGFR2, AC000061.1, ADAMTS9-AS2 \n",
      "\t   ABCC3, COL18A1, PDGFD, AUTS2, CFAP221, LINC00472, GLIS3, GUCY1A2, KCNJ16, IGF2BP2 \n",
      "Negative:  KCNMA1, RIMS2, PAM, SLC30A8, KCNMB2, CACNA1A, FKBP5, KCNB2, TMEM178B, NRXN1 \n",
      "\t   MTUS2, NOL4, ABCC8, RGS7, PWRN1, ST18, FMN2, MYT1L, DPP6, RIMBP2 \n",
      "\t   LRRTM4, RFX6, MAGI2, PCSK2, PRSS1, TRHDE, UNC80, MEG3, LINGO2, KCNJ6 \n",
      "PC_ 3 \n",
      "Positive:  LAMA2, COL6A3, ZFPM2, BNC2, FBXL7, FBN1, COL1A2, PDGFRA, NID1, COL3A1 \n",
      "\t   C7, DCN, CRISPLD2, COL6A2, LHFPL6, FGF7, DCLK1, ABCA8, NR2F1-AS1, SVEP1 \n",
      "\t   COL4A1, ADH1B, IGFBP4, EBF1, PRR16, SYNPO2, COL6A1, NEGR1, LAMA4, ZEB1 \n",
      "Negative:  SLC4A4, PKHD1, CFTR, ANXA4, DCDC2, FRAS1, SCTR, THSD4, CCDC198, FGFR2 \n",
      "\t   ACSM3, CFAP221, FAM189A1, KCNJ16, AC000061.1, ERBB4, ABCC3, CADPS, CHST9, POU6F2 \n",
      "\t   WWC1, ESRRG, CASR, AKAP7, ATP1B1, KCNJ15, IGF2BP2, RALYL, LINC00511, AC104035.1 \n",
      "PC_ 4 \n",
      "Positive:  PTPRC, DOCK2, TBXAS1, PIK3R5, FYB1, AOAH, LCP2, SLA, CD163, MS4A6A \n",
      "\t   MSR1, INPP5D, ARHGAP15, FLI1, PREX1, TFEC, MYO1F, ATP8B4, PRKCB, KYNU \n",
      "\t   SLC1A3, MS4A7, FGD2, NCKAP1L, MRC1, ELMO1, PTPRE, IKZF1, ZEB2, DOCK11 \n",
      "Negative:  LAMA2, COL6A3, COL1A2, C7, SVEP1, NEGR1, COL3A1, DCLK1, FGF7, PDGFRA \n",
      "\t   FBN1, COL5A2, DCN, ABCA8, COL6A1, NID1, COL4A1, NR2F1-AS1, COL6A2, PTPRG \n",
      "\t   ADH1B, GAREM1, TSHZ2, SLIT2, SYNPO2, SMOC2, RBPMS, CRISPLD2, KAZN, GLI3 \n",
      "PC_ 5 \n",
      "Positive:  ZNF385D, TRPM3, DACH2, LRRTM3, INS, SYT16, LMNTD1, UNC5D, PPM1E, LINC01099 \n",
      "\t   NRG1, AP000829.1, HDAC9, KIRREL3, ZNF385D-AS2, VAT1L, LINC01098, KIAA1211, SLCO5A1, KCNH1 \n",
      "\t   C1orf127, NRG3, G6PC2, KCNH7, KCTD8, GRIN2A, VWDE, GLRA1, BMP5, CAPN13 \n",
      "Negative:  FAP, AL591501.1, LINC02388, GCG, PLPPR1, AL138694.1, DSCAM, SPOCK3, KIAA1549L, GPC6 \n",
      "\t   BTBD11, AC090531.1, SLC35F4, TENM2, PTPRT, SNTG1, KCNIP1, AC020637.1, TTR, FSTL5 \n",
      "\t   SYNDIG1, NPSR1-AS1, STK32B, MYO16, SEMA3E, TMEM132D, PPP2R2B, PAPPA2, KCNJ3, ADAMTS18 \n",
      "\n",
      "11:27:01 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:27:01 Read 5137 rows and found 50 numeric columns\n",
      "\n",
      "11:27:01 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:27:01 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "11:27:02 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203771961ec19\n",
      "\n",
      "11:27:02 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "11:27:04 Annoy recall = 100%\n",
      "\n",
      "11:27:06 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "11:27:10 Found 2 connected components, \n",
      "falling back to 'spca' initialization with init_sdev = 1\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:27:10 Using 'irlba' for PCA\n",
      "\n",
      "11:27:10 PCA: 2 components explained 39.75% variance\n",
      "\n",
      "11:27:10 Scaling init to sdev = 1\n",
      "\n",
      "11:27:10 Commencing optimization for 500 epochs, with 211586 positive edges\n",
      "\n",
      "11:27:18 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "11:28:35 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:28:35 Read 5137 rows and found 49 numeric columns\n",
      "\n",
      "11:28:35 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:28:35 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "11:28:36 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203771e30d4b4\n",
      "\n",
      "11:28:36 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "11:28:37 Annoy recall = 100%\n",
      "\n",
      "11:28:40 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "11:28:44 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "11:28:44 Commencing optimization for 500 epochs, with 215926 positive edges\n",
      "\n",
      "11:28:52 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "11:29:05 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:29:07 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "11:29:11 Found 2 connected components, \n",
      "falling back to random initialization\n",
      "\n",
      "11:29:11 Initializing from uniform random\n",
      "\n",
      "11:29:11 Commencing optimization for 500 epochs, with 153650 positive edges\n",
      "\n",
      "11:29:19 Optimization finished\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Starting DecontX\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Fri Dec 13 11:31:15 2024 .. Analyzing all cells\n",
      "\n",
      "Fri Dec 13 11:31:15 2024 .... Generating UMAP and estimating cell types\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Fri Dec 13 11:31:35 2024 .... Estimating contamination\n",
      "\n",
      "Fri Dec 13 11:31:36 2024 ...... Completed iteration: 10 | converge: 0.01903\n",
      "\n",
      "Fri Dec 13 11:31:37 2024 ...... Completed iteration: 20 | converge: 0.007524\n",
      "\n",
      "Fri Dec 13 11:31:38 2024 ...... Completed iteration: 30 | converge: 0.003976\n",
      "\n",
      "Fri Dec 13 11:31:39 2024 ...... Completed iteration: 40 | converge: 0.001837\n",
      "\n",
      "Fri Dec 13 11:31:40 2024 ...... Completed iteration: 50 | converge: 0.001073\n",
      "\n",
      "Fri Dec 13 11:31:40 2024 ...... Completed iteration: 51 | converge: 0.0009903\n",
      "\n",
      "Fri Dec 13 11:31:40 2024 .. Calculating final decontaminated matrix\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Completed DecontX. Total time: 26.37931 secs\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.2 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 1.2 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message:\n",
      "\"Key 'rna_' taken, using 'rnaraw_' instead\"\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  PRSS1, CTRB1, SPINK1, CEL, TC2N, CUZD1, CPA1, TRHDE, SEL1L, CTRB2 \n",
      "\t   CLPS, PRSS2, CATSPERB, PNLIP, CTRC, GULP1, REG1B, ZNF804B, NTN4, AC010967.1 \n",
      "\t   PEX5L, GPT2, PHACTR1, MIR217HG, AASS, PLA2G1B, AMY2B, ANPEP, AC013652.1, AMY2A \n",
      "Negative:  MAGI2, RIMS2, NRXN1, DPP6, CACNA1D, KCNMB2, SLC30A8, SPAG16, AGBL4, CACNA1A \n",
      "\t   KCNB2, CACNA1C, NOL4, ABCC8, FMN2, RGS7, MEIS2, MAPK10, ST18, UNC80 \n",
      "\t   MTUS2, CADPS, LRRTM4, FGF14, STXBP5L, NLGN1, CACNA2D1, DACH1, PLCXD3, MYT1L \n",
      "PC_ 2 \n",
      "Positive:  RIMS2, NRXN1, PRSS1, PDE3B, SLC30A8, KCNMB2, KCNB2, CACNA1A, ABCC8, NOL4 \n",
      "\t   RGS7, CTRB1, MTUS2, HS6ST3, ST18, SPINK1, CEL, FMN2, DPP6, LRRTM4 \n",
      "\t   MEG3, MYT1L, KCNMA1, TRPM3, PWRN1, TMEM178B, RIMBP2, ELAVL4, KCNJ6, MAGI2 \n",
      "Negative:  BICC1, SLC4A4, PKHD1, CFTR, ANXA4, THSD4, ABCC3, SCTR, FRAS1, ACSM3 \n",
      "\t   PDE3A, DCDC2, NRP1, FGFR2, AKAP7, COL18A1, ADAMTS9-AS2, CFAP221, PLD1, MAML2 \n",
      "\t   AC000061.1, WWTR1, PDGFD, KCNJ16, EPB41L4A, IGF2BP2, GUCY1A2, GLIS3, RBMS3, AUTS2 \n",
      "PC_ 3 \n",
      "Positive:  LAMA2, ZFPM2, FBXL7, COL6A3, C7, PDGFRA, NID1, BNC2, COL3A1, COL6A2 \n",
      "\t   TIMP3, FBN1, COL1A2, ZEB2, A2M, CRISPLD2, DCN, LHFPL6, COL4A1, ABCA8 \n",
      "\t   SVEP1, LAMA4, NR2F1-AS1, EBF1, PDGFRB, ADAMTS12, SMOC2, FGF7, DCLK1, MRC2 \n",
      "Negative:  SLC4A4, PKHD1, CFTR, SCTR, ANXA4, FRAS1, DCDC2, ACSM3, THSD4, FGFR2 \n",
      "\t   CFAP221, ABCC3, KCNJ16, CASR, CCDC198, FAM189A1, AC000061.1, ERBB4, CADPS, ESRRG \n",
      "\t   CHST9, RALYL, WWC1, KCNJ15, IGF2BP2, POU6F2, FLRT2, AC104035.1, LINC00511, KIFC3 \n",
      "PC_ 4 \n",
      "Positive:  GPC6, LAMA2, GLI3, LINC02388, FAP, AL591501.1, ZFPM2, GCG, C7, SNED1 \n",
      "\t   COL6A3, PDGFRA, COL3A1, FBN1, SLIT2, SVEP1, COL6A2, TENM2, KCNIP1, DCN \n",
      "\t   NR2F1-AS1, PLPPR1, COL1A2, COL5A2, NID1, DSCAM, PTPRG, NEGR1, TTR, PTPRT \n",
      "Negative:  DOCK2, PTPRC, ATP8B4, CD163, TBXAS1, MSR1, MYO1F, ELMO1, LCP2, STAB1 \n",
      "\t   FLI1, PIK3R5, ARHGAP15, AOAH, PTPRE, MS4A7, INPP5D, FYB1, MRC1, PREX1 \n",
      "\t   MS4A6A, HCK, F13A1, SLA, FGD2, TGFBI, CD163L1, SLCO2B1, SLC9A9, SLC1A3 \n",
      "PC_ 5 \n",
      "Positive:  ZNF385D, INS, TRPM3, DACH2, LRRTM3, SYT16, LMNTD1, LINC01099, NRG1, PPM1E \n",
      "\t   SLCO5A1, KIRREL3, UNC5D, PLCH2, HDAC9, VAT1L, DGKG, KIAA1211, G6PC2, PFKFB2 \n",
      "\t   C1orf127, LINC01098, AP000829.1, ANKRD34C-AS1, LRFN2, ASB9, NRG3, TMEM108, TMEM150C, KLHL1 \n",
      "Negative:  AL591501.1, GCG, LINC02388, FAP, PLPPR1, GPC6, TENM2, DSCAM, TTR, BTBD11 \n",
      "\t   SLC35F4, KIAA1549L, HECW2, SNTG1, KCNIP1, AL138694.1, FSTL5, SPOCK3, PTPRT, PPP2R2B \n",
      "\t   NPSR1-AS1, MYO16, STK32B, KCNJ3, PAPPA2, DCC, PDZD2, ARX, AC090531.1, TM4SF4 \n",
      "\n",
      "11:31:57 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:31:57 Read 4433 rows and found 50 numeric columns\n",
      "\n",
      "11:31:57 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:31:58 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "11:31:58 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203771bc2c0e6\n",
      "\n",
      "11:31:58 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "11:31:59 Annoy recall = 100%\n",
      "\n",
      "11:32:02 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "11:32:06 Found 2 connected components, \n",
      "falling back to 'spca' initialization with init_sdev = 1\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:32:06 Using 'irlba' for PCA\n",
      "\n",
      "11:32:06 PCA: 2 components explained 43.5% variance\n",
      "\n",
      "11:32:06 Scaling init to sdev = 1\n",
      "\n",
      "11:32:06 Commencing optimization for 500 epochs, with 181366 positive edges\n",
      "\n",
      "11:32:13 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "11:33:21 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:33:21 Read 4433 rows and found 49 numeric columns\n",
      "\n",
      "11:33:21 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:33:21 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "11:33:22 Writing NN index file to temp file /tmp/Rtmp95L5vt/file32037711ba8707\n",
      "\n",
      "11:33:22 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "11:33:23 Annoy recall = 100%\n",
      "\n",
      "11:33:25 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "11:33:29 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "11:33:29 Commencing optimization for 500 epochs, with 183862 positive edges\n",
      "\n",
      "11:33:37 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "11:33:47 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:33:50 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "11:33:54 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "11:33:54 Commencing optimization for 500 epochs, with 131006 positive edges\n",
      "\n",
      "11:34:01 Optimization finished\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Starting DecontX\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Fri Dec 13 11:35:44 2024 .. Analyzing all cells\n",
      "\n",
      "Fri Dec 13 11:35:45 2024 .... Generating UMAP and estimating cell types\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Fri Dec 13 11:36:08 2024 .... Estimating contamination\n",
      "\n",
      "Fri Dec 13 11:36:11 2024 ...... Completed iteration: 10 | converge: 0.0403\n",
      "\n",
      "Fri Dec 13 11:36:13 2024 ...... Completed iteration: 20 | converge: 0.01158\n",
      "\n",
      "Fri Dec 13 11:36:15 2024 ...... Completed iteration: 30 | converge: 0.006224\n",
      "\n",
      "Fri Dec 13 11:36:17 2024 ...... Completed iteration: 40 | converge: 0.004403\n",
      "\n",
      "Fri Dec 13 11:36:19 2024 ...... Completed iteration: 50 | converge: 0.003373\n",
      "\n",
      "Fri Dec 13 11:36:21 2024 ...... Completed iteration: 60 | converge: 0.002496\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Dec 13 11:36:23 2024 ...... Completed iteration: 70 | converge: 0.001931\n",
      "\n",
      "Fri Dec 13 11:36:25 2024 ...... Completed iteration: 80 | converge: 0.001394\n",
      "\n",
      "Fri Dec 13 11:36:27 2024 ...... Completed iteration: 90 | converge: 0.001024\n",
      "\n",
      "Fri Dec 13 11:36:27 2024 ...... Completed iteration: 91 | converge: 0.0009905\n",
      "\n",
      "Fri Dec 13 11:36:27 2024 .. Calculating final decontaminated matrix\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Completed DecontX. Total time: 44.96881 secs\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.0 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.0 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message:\n",
      "\"Key 'rna_' taken, using 'rnaraw_' instead\"\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  TRHDE, SEL1L, CUZD1, CATSPERB, GULP1, PNLIP, CTRB2, SLC16A10, CTRC, PRSS2 \n",
      "\t   UST, ZNF804B, NTN4, REG1B, AC010967.1, PEX5L, EIF2AK3, PRKG1, GPT2, AASS \n",
      "\t   MIR217HG, GAREM1, IFRD1, ANPEP, AC013652.1, AMY2B, CYP3A5, AL159156.1, AMY2A, AC002463.1 \n",
      "Negative:  RIMS2, DPP6, NRXN1, KCNMB2, KCNB2, SLC30A8, CACNA1A, NOL4, CACNA1C, CACNA1D \n",
      "\t   ABCC8, RGS7, FMN2, ST18, UNC80, MTUS2, STXBP5L, MYT1L, PLCB4, AGBL4 \n",
      "\t   LRRTM4, RIMBP2, CNTN1, NLGN1, ROBO2, CACNA2D1, SOX5, SNAP25, RTN1, KCNJ6 \n",
      "PC_ 2 \n",
      "Positive:  PDE3B, TRHDE, HS6ST3, FKBP5, PEX5L, PLXNA2, PWRN1, SEL1L, MEG3, KCNMA1 \n",
      "\t   SLC16A10, CUZD1, PNLIP, GULP1, SYBU, MEG8, TMEM178B, UST, EIF2AK3, ZNF804B \n",
      "\t   MAGI2, LINC01091, CTRC, AC010967.1, SLC7A2, ATRNL1, RIMS2, SLC30A8, GRIK2, MYOM1 \n",
      "Negative:  BICC1, PKHD1, CFTR, SLC4A4, ANXA4, THSD4, PDE3A, SCTR, FRAS1, NRP1 \n",
      "\t   DCDC2, ABCC3, MAML2, FGFR2, PLD1, ACSM3, AKAP7, GLIS3, COL18A1, CFAP221 \n",
      "\t   ADAMTS9-AS2, RBMS3, AC000061.1, AUTS2, WWTR1, PDGFD, KCNJ16, GUCY1A2, LINC00472, CCDC198 \n",
      "PC_ 3 \n",
      "Positive:  SLC4A4, PKHD1, CFTR, SCTR, FRAS1, THSD4, ANXA4, DCDC2, FGFR2, CFAP221 \n",
      "\t   KCNJ16, CCDC198, ACSM3, CADPS, FAM189A1, ABCC3, ESRRG, POU6F2, FLRT2, ERBB4 \n",
      "\t   AC000061.1, CASR, CHST9, KCNJ15, AC104035.1, LINC00511, RALYL, GLIS3, SLC17A4, SLC28A3 \n",
      "Negative:  LAMA2, ZFPM2, FBXL7, BNC2, ZEB2, COL6A3, NID1, PDGFRA, LHFPL6, COL6A2 \n",
      "\t   CRISPLD2, IRAK3, COL1A2, DCN, SPARCL1, TIMP3, SLC9A9, COL4A1, COL3A1, A2M \n",
      "\t   C7, LRP1, EBF1, NR2F1-AS1, THBS1, CPED1, SVEP1, FBN1, IFI16, CELF2 \n",
      "PC_ 4 \n",
      "Positive:  DOCK2, PTPRC, PIK3R5, TBXAS1, FYB1, AOAH, CD163, ARHGAP15, MYO1F, MSR1 \n",
      "\t   MS4A6A, LCP2, INPP5D, ATP8B4, MS4A7, STAB1, SLC1A3, PRKCB, SLA, SLC11A1 \n",
      "\t   IKZF1, PREX1, MRC1, FGD2, APBB1IP, FLI1, LCP1, CSF2RA, DOCK11, ELMO1 \n",
      "Negative:  LAMA2, NID1, ZFPM2, SVEP1, COL6A3, PDGFRA, CRISPLD2, DCN, FBXL7, C7 \n",
      "\t   COL6A2, COL1A2, NR2F1-AS1, COL4A1, COL3A1, EBF1, NEGR1, ABCA8, FGF7, FBN1 \n",
      "\t   COL4A2, COL5A2, PRKG1, ADAMTS12, ADH1B, COL5A1, ZFPM2-AS1, GLI3, PDGFRB, DCLK1 \n",
      "PC_ 5 \n",
      "Positive:  AL591501.1, FAP, LINC02388, PLPPR1, GCG, AL138694.1, DSCAM, GPC6, SPOCK3, TENM2 \n",
      "\t   KCNIP1, PTPRT, KIAA1549L, AC090531.1, BTBD11, FSTL5, TTR, SLC35F4, SNTG1, MYO16 \n",
      "\t   STK32B, SYNDIG1, AC020637.1, IRX2, NPSR1-AS1, PAPPA2, DCC, PPP2R2B, SEMA3E, AJ009632.2 \n",
      "Negative:  ZNF385D, INS, LRRTM3, DACH2, TRPM3, LMNTD1, SYT16, LINC01099, NRG1, KIRREL3 \n",
      "\t   PPM1E, KIAA1211, UNC5D, C1orf127, VAT1L, BMP5, SLCO5A1, G6PC2, ZNF385D-AS2, NRG3 \n",
      "\t   LINC01098, HDAC9, AP000829.1, PLCH2, PFKFB2, PCDH7, GLRA1, TFCP2L1, HADH, TMEM108 \n",
      "\n",
      "11:37:01 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:37:01 Read 7283 rows and found 50 numeric columns\n",
      "\n",
      "11:37:01 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:37:01 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "11:37:01 Writing NN index file to temp file /tmp/Rtmp95L5vt/file320377dae8456\n",
      "\n",
      "11:37:01 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "11:37:03 Annoy recall = 100%\n",
      "\n",
      "11:37:06 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "11:37:10 Found 2 connected components, \n",
      "falling back to 'spca' initialization with init_sdev = 1\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:37:10 Using 'irlba' for PCA\n",
      "\n",
      "11:37:10 PCA: 2 components explained 42.82% variance\n",
      "\n",
      "11:37:10 Scaling init to sdev = 1\n",
      "\n",
      "11:37:10 Commencing optimization for 500 epochs, with 302816 positive edges\n",
      "\n",
      "11:37:20 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "11:39:00 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:39:00 Read 7283 rows and found 49 numeric columns\n",
      "\n",
      "11:39:00 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:39:00 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "11:39:00 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203775cdcd828\n",
      "\n",
      "11:39:00 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "11:39:02 Annoy recall = 100%\n",
      "\n",
      "11:39:06 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "11:39:10 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "11:39:10 Commencing optimization for 500 epochs, with 310300 positive edges\n",
      "\n",
      "11:39:20 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "11:39:39 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:39:41 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "11:39:46 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "11:39:46 Commencing optimization for 500 epochs, with 219960 positive edges\n",
      "\n",
      "11:39:57 Optimization finished\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Starting DecontX\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Fri Dec 13 11:42:53 2024 .. Analyzing all cells\n",
      "\n",
      "Fri Dec 13 11:42:53 2024 .... Generating UMAP and estimating cell types\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Fri Dec 13 11:43:30 2024 .... Estimating contamination\n",
      "\n",
      "Fri Dec 13 11:43:36 2024 ...... Completed iteration: 10 | converge: 0.03658\n",
      "\n",
      "Fri Dec 13 11:43:41 2024 ...... Completed iteration: 20 | converge: 0.01509\n",
      "\n",
      "Fri Dec 13 11:43:46 2024 ...... Completed iteration: 30 | converge: 0.007613\n",
      "\n",
      "Fri Dec 13 11:43:51 2024 ...... Completed iteration: 40 | converge: 0.005374\n",
      "\n",
      "Fri Dec 13 11:43:56 2024 ...... Completed iteration: 50 | converge: 0.003891\n",
      "\n",
      "Fri Dec 13 11:44:02 2024 ...... Completed iteration: 60 | converge: 0.002606\n",
      "\n",
      "Fri Dec 13 11:44:07 2024 ...... Completed iteration: 70 | converge: 0.002055\n",
      "\n",
      "Fri Dec 13 11:44:12 2024 ...... Completed iteration: 80 | converge: 0.001591\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Dec 13 11:44:17 2024 ...... Completed iteration: 90 | converge: 0.001424\n",
      "\n",
      "Fri Dec 13 11:44:22 2024 ...... Completed iteration: 100 | converge: 0.00113\n",
      "\n",
      "Fri Dec 13 11:44:25 2024 ...... Completed iteration: 106 | converge: 0.0009827\n",
      "\n",
      "Fri Dec 13 11:44:25 2024 .. Calculating final decontaminated matrix\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Completed DecontX. Total time: 1.616995 mins\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 4.9 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 4.9 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message:\n",
      "\"Key 'rna_' taken, using 'rnaraw_' instead\"\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  CACNA1C, SOX5, RIMS2, DPP6, KCNMB2, CACNA2D1, CACNA1D, NRXN1, SLC30A8, KCNB2 \n",
      "\t   NLGN1, NOL4, KCNMB2-AS1, CACNA1A, RGS7, ST18, NOVA1, FGF14, FMN2, PLCB4 \n",
      "\t   NEGR1, LRRTM4, CNTN1, DACH1, UNC80, HMGCLL1, ROBO2, ABCC8, AKT3, STXBP5L \n",
      "Negative:  REG1A, CELA3A, PRSS1, CTRB1, CELA3B, CEL, CLPS, CPA1, PNLIP, CTRB2 \n",
      "\t   CATSPERB, CTRC, PRSS2, PEX5L, REG1B, ZNF804B, AC010967.1, NTN4, PWRN1, EIF2AK3 \n",
      "\t   MIR217HG, PLA2G1B, AMY2B, AMY2A, BTG2, PLXNA2, ANPEP, AASS, AC013652.1, AL159156.1 \n",
      "PC_ 2 \n",
      "Positive:  RIMS2, SLC30A8, NRXN1, KCNMB2, KCNB2, NOL4, DPP6, CACNA1A, RGS7, KCNMB2-AS1 \n",
      "\t   FMN2, ST18, ABCC8, MTUS2, LRRTM4, MYT1L, RIMBP2, KCNJ6, LINGO2, RFX6 \n",
      "\t   STXBP5L, UNC80, SYT14, PCSK2, ELAVL4, SNAP25, TMEM132D, INS, ROBO2, TRPM3 \n",
      "Negative:  RBMS3, BICC1, NRP1, CFTR, PKHD1, SLC4A4, ANXA4, MAML2, COL4A2, THSD4 \n",
      "\t   PDE3A, WWTR1, ADAMTS9-AS2, COL4A1, PTPRG, EBF1, AKAP7, CALD1, SCTR, LAMA2 \n",
      "\t   PDGFD, PLD1, AC000061.1, IGFBP7, COL6A3, DCDC2, ADAMTS9, RUNX1, FRAS1, SVEP1 \n",
      "PC_ 3 \n",
      "Positive:  LAMA2, COL6A3, C7, ZFPM2, COL1A2, DCN, BNC2, COL6A2, FBXL7, PDGFRA \n",
      "\t   NID1, COL3A1, FBN1, CRISPLD2, FGF7, SRPX, FBLN1, LHFPL6, IGFBP4, DCLK1 \n",
      "\t   ABCA9, SMOC2, ADH1B, THBS1, COL5A1, ZEB1, PID1, MIR99AHG, LRP1, ABCA8 \n",
      "Negative:  SLC4A4, CFTR, PKHD1, ANXA4, SCTR, DCDC2, FRAS1, THSD4, AC000061.1, AKAP7 \n",
      "\t   GLIS3, BICC1, FGFR2, ACSM3, KCNJ16, PDE3A, ERBB4, CFAP221, FAM189A1, ABCC3 \n",
      "\t   GUCY1A2, POU6F2, PDGFD, CADPS, PLD1, ATP1B1, KCNJ15, RALYL, CHST9, LINC00472 \n",
      "PC_ 4 \n",
      "Positive:  DOCK2, TBXAS1, PTPRC, PIK3R5, FYB1, CD163, AOAH, MSR1, MS4A6A, ARHGAP15 \n",
      "\t   MS4A7, MYO1F, LCP2, ATP8B4, TFEC, SLC1A3, INPP5D, CLEC7A, MRC1, HCLS1 \n",
      "\t   SLA, FGD2, HCK, PRKCB, CYBB, SLC11A1, ZEB2, PREX1, STAB1, FLI1 \n",
      "Negative:  LAMA2, C7, COL6A3, PDGFRA, DCN, COL6A2, COL1A2, COL3A1, SVEP1, ZFPM2 \n",
      "\t   NID1, FBN1, NEGR1, SRPX, FBXL7, FGF7, FBLN1, COL4A1, COL6A1, CRISPLD2 \n",
      "\t   DCLK1, SLIT2, COL5A2, NR2F1-AS1, SMOC2, ADH1B, COL4A2, PTPRG, ABCA8, EBF1 \n",
      "PC_ 5 \n",
      "Positive:  AL591501.1, GCG, LINC02388, FAP, PLPPR1, AC020637.1, TTR, AL138694.1, DSCAM, FSTL5 \n",
      "\t   AC090531.1, SPOCK3, GPC6, TENM2, SNTG1, PTPRT, KIAA1549L, SLC35F4, BTBD11, KCNIP1 \n",
      "\t   MYO16, HECW2, STK32B, NPSR1-AS1, KCNJ3, SYNDIG1, PAPPA2, PPP2R2B, ARX, MMP16 \n",
      "Negative:  ZNF385D, DACH2, LRRTM3, TRPM3, INS, ZNF385D-AS2, NRG1, LINC01099, LMNTD1, SYT16 \n",
      "\t   UNC5D, PPM1E, LINC01098, KIAA1211, GPM6A, KIRREL3, SLCO5A1, G6PC2, BMP5, KCTD8 \n",
      "\t   HDAC9, AP000829.1, VAT1L, CTNNA3, NRG3, TMEM108, KCNH7, PLCH2, KLHL1, RASD1 \n",
      "\n",
      "11:45:34 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:45:34 Read 18086 rows and found 50 numeric columns\n",
      "\n",
      "11:45:34 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:45:34 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "11:45:36 Writing NN index file to temp file /tmp/Rtmp95L5vt/file32037755461e5\n",
      "\n",
      "11:45:36 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "11:45:41 Annoy recall = 100%\n",
      "\n",
      "11:45:44 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "11:45:49 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "11:45:53 Commencing optimization for 200 epochs, with 769982 positive edges\n",
      "\n",
      "11:46:04 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "11:49:26 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:49:26 Read 18086 rows and found 49 numeric columns\n",
      "\n",
      "11:49:26 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:49:26 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "11:49:28 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203777eac5b5d\n",
      "\n",
      "11:49:28 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "11:49:34 Annoy recall = 100%\n",
      "\n",
      "11:49:37 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "11:49:41 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "11:49:42 Commencing optimization for 200 epochs, with 785124 positive edges\n",
      "\n",
      "11:49:52 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "11:50:45 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "11:50:48 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "11:50:52 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "11:50:54 Commencing optimization for 200 epochs, with 558186 positive edges\n",
      "\n",
      "11:51:05 Optimization finished\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.4 GiB\"\n",
      "--------------------------------------------------\n",
      "\n",
      "Starting DecontX\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Fri Dec 13 11:58:04 2024 .. Analyzing all cells\n",
      "\n",
      "Fri Dec 13 11:58:04 2024 .... Generating UMAP and estimating cell types\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Fri Dec 13 11:58:41 2024 .... Estimating contamination\n",
      "\n",
      "Fri Dec 13 11:58:47 2024 ...... Completed iteration: 10 | converge: 0.0294\n",
      "\n",
      "Fri Dec 13 11:58:51 2024 ...... Completed iteration: 20 | converge: 0.01227\n",
      "\n",
      "Fri Dec 13 11:58:56 2024 ...... Completed iteration: 30 | converge: 0.009146\n",
      "\n",
      "Fri Dec 13 11:59:01 2024 ...... Completed iteration: 40 | converge: 0.007065\n",
      "\n",
      "Fri Dec 13 11:59:06 2024 ...... Completed iteration: 50 | converge: 0.005034\n",
      "\n",
      "Fri Dec 13 11:59:10 2024 ...... Completed iteration: 60 | converge: 0.003833\n",
      "\n",
      "Fri Dec 13 11:59:15 2024 ...... Completed iteration: 70 | converge: 0.002536\n",
      "\n",
      "Fri Dec 13 11:59:20 2024 ...... Completed iteration: 80 | converge: 0.001935\n",
      "\n",
      "Fri Dec 13 11:59:24 2024 ...... Completed iteration: 90 | converge: 0.001559\n",
      "\n",
      "Fri Dec 13 11:59:29 2024 ...... Completed iteration: 100 | converge: 0.001284\n",
      "\n",
      "Fri Dec 13 11:59:34 2024 ...... Completed iteration: 110 | converge: 0.001074\n",
      "\n",
      "Fri Dec 13 11:59:36 2024 ...... Completed iteration: 114 | converge: 0.0009992\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fri Dec 13 11:59:36 2024 .. Calculating final decontaminated matrix\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Completed DecontX. Total time: 1.609091 mins\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 5.0 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 5.0 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message:\n",
      "\"Key 'rna_' taken, using 'rnaraw_' instead\"\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  CACNA1C, SOX5, CACNA2D1, NEGR1, AKT3, NOVA1, CACNA1D, DPP6, RIMS2, FGF14 \n",
      "\t   NRXN1, NLGN1, KCNMB2, SLC30A8, NOL4, KCNB2, ZEB1, PLCB4, CACNA1A, DACH1 \n",
      "\t   MAML2, KCNMB2-AS1, CNTN1, LRRTM4, ST18, RGS7, MEIS2, PDE10A, CADPS, FMN2 \n",
      "Negative:  REG1A, CELA3A, PRSS1, CTRB1, CELA3B, CEL, CUZD1, CLPS, CPA1, CTRB2 \n",
      "\t   PNLIP, CTRC, CATSPERB, PRSS2, PEX5L, ZNF804B, REG1B, EIF2AK3, AC010967.1, NTN4 \n",
      "\t   GPT2, PWRN1, MIR217HG, PLXNA2, PLA2G1B, IFRD1, ANPEP, AMY2B, BTG2, AMY2A \n",
      "PC_ 2 \n",
      "Positive:  RBMS3, BICC1, COL4A2, NRP1, COL4A1, LAMA2, PTPRG, COL6A3, CFTR, WWTR1 \n",
      "\t   SLC4A4, EBF1, PKHD1, ANXA4, C7, CALD1, IGFBP7, SVEP1, ZFPM2, THSD4 \n",
      "\t   LAMC1, PDE3A, COL6A2, MAML2, PDGFRA, NID1, COL1A2, PDGFD, BNC2, DCN \n",
      "Negative:  RIMS2, NRXN1, SLC30A8, KCNMB2, KCNB2, NOL4, CACNA1A, RGS7, DPP6, KCNMB2-AS1 \n",
      "\t   ABCC8, ST18, FMN2, LRRTM4, MTUS2, RIMBP2, UNC80, MYT1L, LINGO2, RFX6 \n",
      "\t   KCNJ6, PCSK2, SYT14, STXBP5L, ELAVL4, TMEM132D, CNTN1, ROBO2, SLC4A10, HMGCLL1 \n",
      "PC_ 3 \n",
      "Positive:  SLC4A4, CFTR, PKHD1, ANXA4, SCTR, FRAS1, DCDC2, THSD4, AC000061.1, BICC1 \n",
      "\t   AKAP7, GLIS3, ACSM3, FGFR2, PDE3A, KCNJ16, ERBB4, ABCC3, CFAP221, PDGFD \n",
      "\t   FAM189A1, PLD1, GUCY1A2, ATP1B1, POU6F2, CADPS, KCNJ15, RALYL, LINC00472, SEMA6A \n",
      "Negative:  LAMA2, COL6A3, C7, COL6A2, ZFPM2, DCN, COL1A2, PDGFRA, NID1, FBXL7 \n",
      "\t   COL3A1, CRISPLD2, BNC2, FBN1, FGF7, IGFBP4, SRPX, ZEB1, MRC2, NEGR1 \n",
      "\t   FBLN1, ABCA8, LHFPL6, TIMP3, ADH1B, PID1, COL6A1, SMOC2, PDGFRB, PRR16 \n",
      "PC_ 4 \n",
      "Positive:  DOCK2, TBXAS1, PTPRC, PIK3R5, CD163, MSR1, AOAH, FYB1, MS4A6A, ARHGAP15 \n",
      "\t   TFEC, ATP8B4, MS4A7, MYO1F, PRKCB, INPP5D, SLC1A3, LCP2, HCLS1, ZEB2 \n",
      "\t   CLEC7A, FGD2, KYNU, FLI1, NCKAP1L, SLA, MRC1, CYBB, APBB1IP, PREX1 \n",
      "Negative:  LAMA2, DCN, PDGFRA, C7, COL6A3, SVEP1, COL1A2, NID1, COL6A1, PTPRG \n",
      "\t   COL6A2, COL3A1, NEGR1, ZFPM2, COL5A2, FGF7, FBN1, SLIT2, SMOC2, CRISPLD2 \n",
      "\t   SRPX, COL4A1, DCLK1, TSHZ2, ABCA9, ADH1B, NR2F1-AS1, FBLN1, GLI3, KAZN \n",
      "PC_ 5 \n",
      "Positive:  ZNF385D, DACH2, TRPM3, LRRTM3, INS, SYT16, ZNF385D-AS2, NRG1, PPM1E, UNC5D \n",
      "\t   LMNTD1, LINC01099, KIAA1211, LINC01098, GPM6A, SLCO5A1, HDAC9, KIRREL3, G6PC2, BMP5 \n",
      "\t   VAT1L, PLCH2, KCTD8, NRG3, CTNNA3, IAPP, AP000829.1, ANKRD34C-AS1, TMEM108, KCNH7 \n",
      "Negative:  AL591501.1, GCG, LINC02388, FAP, PLPPR1, AC020637.1, TTR, AL138694.1, DSCAM, FSTL5 \n",
      "\t   SPOCK3, GPC6, AC090531.1, TENM2, SNTG1, PTPRT, SLC35F4, KIAA1549L, BTBD11, KCNIP1 \n",
      "\t   HECW2, NPSR1-AS1, STK32B, MYO16, SEMA3E, PAPPA2, SYNDIG1, AL592043.1, MMP16, KCNJ3 \n",
      "\n",
      "12:00:40 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:00:40 Read 18258 rows and found 50 numeric columns\n",
      "\n",
      "12:00:40 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:00:40 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "12:00:42 Writing NN index file to temp file /tmp/Rtmp95L5vt/file320377e5ad969\n",
      "\n",
      "12:00:42 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "12:00:47 Annoy recall = 100%\n",
      "\n",
      "12:00:50 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "12:00:55 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "12:03:16 Commencing optimization for 200 epochs, with 776528 positive edges\n",
      "\n",
      "12:03:26 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "12:06:50 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:06:50 Read 18258 rows and found 49 numeric columns\n",
      "\n",
      "12:06:50 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:06:50 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "12:06:52 Writing NN index file to temp file /tmp/Rtmp95L5vt/file32037725ad3b2d\n",
      "\n",
      "12:06:52 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "12:06:58 Annoy recall = 100%\n",
      "\n",
      "12:07:01 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "12:07:05 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "12:07:06 Commencing optimization for 200 epochs, with 793352 positive edges\n",
      "\n",
      "12:07:16 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "12:08:10 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:08:12 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "12:08:17 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "12:08:18 Commencing optimization for 200 epochs, with 563622 positive edges\n",
      "\n",
      "12:08:29 Optimization finished\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.5 GiB\"\n",
      "--------------------------------------------------\n",
      "\n",
      "Starting DecontX\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Fri Dec 13 12:16:08 2024 .. Analyzing all cells\n",
      "\n",
      "Fri Dec 13 12:16:08 2024 .... Generating UMAP and estimating cell types\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Fri Dec 13 12:16:51 2024 .... Estimating contamination\n",
      "\n",
      "Fri Dec 13 12:16:57 2024 ...... Completed iteration: 10 | converge: 0.03631\n",
      "\n",
      "Fri Dec 13 12:17:03 2024 ...... Completed iteration: 20 | converge: 0.0155\n",
      "\n",
      "Fri Dec 13 12:17:08 2024 ...... Completed iteration: 30 | converge: 0.008032\n",
      "\n",
      "Fri Dec 13 12:17:13 2024 ...... Completed iteration: 40 | converge: 0.005146\n",
      "\n",
      "Fri Dec 13 12:17:18 2024 ...... Completed iteration: 50 | converge: 0.003597\n",
      "\n",
      "Fri Dec 13 12:17:23 2024 ...... Completed iteration: 60 | converge: 0.002521\n",
      "\n",
      "Fri Dec 13 12:17:28 2024 ...... Completed iteration: 70 | converge: 0.001862\n",
      "\n",
      "Fri Dec 13 12:17:34 2024 ...... Completed iteration: 80 | converge: 0.00147\n",
      "\n",
      "Fri Dec 13 12:17:39 2024 ...... Completed iteration: 90 | converge: 0.001149\n",
      "\n",
      "Fri Dec 13 12:17:41 2024 ...... Completed iteration: 94 | converge: 0.0009764\n",
      "\n",
      "Fri Dec 13 12:17:41 2024 .. Calculating final decontaminated matrix\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Completed DecontX. Total time: 1.642491 mins\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 5.0 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 5.0 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message:\n",
      "\"Key 'rna_' taken, using 'rnaraw_' instead\"\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  REG1A, CELA3A, PRSS1, CTRB1, CELA3B, CEL, CLPS, CPA1, CATSPERB, CTRB2 \n",
      "\t   PNLIP, CTRC, PRSS2, REG1B, SLC16A10, ZNF804B, PEX5L, AC010967.1, EIF2AK3, NTN4 \n",
      "\t   PLA2G1B, MIR217HG, GPT2, BTG2, AASS, PWRN1, AMY2B, AMY2A, ANPEP, PLXNA2 \n",
      "Negative:  RIMS2, NRXN1, KCNMB2, DPP6, SLC30A8, CACNA1C, KCNB2, NOL4, CACNA1D, CACNA1A \n",
      "\t   KCNMB2-AS1, ST18, RGS7, UNC80, LRRTM4, FMN2, SOX5, NLGN1, CACNA2D1, CNTN1 \n",
      "\t   ABCC8, DACH1, HMGCLL1, ROBO2, MYT1L, STXBP5L, FGF14, PLCB4, MTUS2, PCSK2 \n",
      "PC_ 2 \n",
      "Positive:  BICC1, CFTR, SLC4A4, PKHD1, ANXA4, THSD4, PDE3A, SCTR, DCDC2, NRP1 \n",
      "\t   MAML2, AKAP7, FRAS1, AC000061.1, ACSM3, PDGFD, PLD1, FGFR2, WWTR1, GLIS3 \n",
      "\t   ABCC3, GUCY1A2, RBMS3, ADAMTS9-AS2, AUTS2, LINC00472, KCNJ16, CFAP221, ANXA2, EPB41L4A \n",
      "Negative:  RIMS2, HS6ST3, NRXN1, PDE3B, KCNMB2, SLC30A8, KCNB2, NOL4, DPP6, CACNA1A \n",
      "\t   KCNMA1, RGS7, KCNMB2-AS1, MTUS2, ST18, FMN2, ABCC8, LRRTM4, MAGI2, TMEM178B \n",
      "\t   MYT1L, CELA3A, PRSS1, REG1A, PCSK2, RIMBP2, RFX6, LINGO2, SYT14, CNTN1 \n",
      "PC_ 3 \n",
      "Positive:  LAMA2, COL6A3, FBXL7, ZFPM2, COL6A2, C7, COL1A2, BNC2, DCN, NID1 \n",
      "\t   PDGFRA, CRISPLD2, COL3A1, FBN1, COL6A1, LHFPL6, IGFBP4, FBLN1, COL4A1, LRP1 \n",
      "\t   THBS1, SRPX, FGF7, NEGR1, DCLK1, COL1A1, SVEP1, TIMP3, MRC2, ADH1B \n",
      "Negative:  SLC4A4, CFTR, PKHD1, ANXA4, SCTR, FRAS1, DCDC2, THSD4, ERBB4, AC000061.1 \n",
      "\t   ESRRG, FGFR2, FAM189A1, CADPS, ACSM3, KCNJ16, CFAP221, CHST9, AKAP7, POU6F2 \n",
      "\t   ABCC3, GLIS3, RALYL, CASR, ATP1B1, PDE3A, CNTN4, LINC00511, PDGFD, KCNJ15 \n",
      "PC_ 4 \n",
      "Positive:  TBXAS1, PTPRC, DOCK2, MS4A6A, PIK3R5, AOAH, FYB1, CD163, MSR1, ARHGAP15 \n",
      "\t   MYO1F, SLC1A3, SLA, INPP5D, LCP2, MS4A7, ATP8B4, FLI1, CLEC7A, MRC1 \n",
      "\t   PRKCB, TFEC, HCLS1, HCK, PREX1, ZEB2, CYBB, FGD2, SLC11A1, APBB1IP \n",
      "Negative:  LAMA2, COL6A3, DCN, COL1A2, PDGFRA, COL6A2, C7, NID1, SVEP1, COL3A1 \n",
      "\t   COL6A1, CRISPLD2, FBLN1, NEGR1, COL5A2, ZFPM2, SRPX, COL4A1, DCLK1, NR2F1-AS1 \n",
      "\t   ADH1B, FBXL7, FGF7, ABCA8, COL4A2, COL1A1, FBN1, PDGFRB, C1R, SMOC2 \n",
      "PC_ 5 \n",
      "Positive:  AL591501.1, GCG, LINC02388, FAP, PLPPR1, AC020637.1, SPOCK3, TTR, DSCAM, FSTL5 \n",
      "\t   AL138694.1, GPC6, TENM2, AC090531.1, PTPRT, SLC35F4, KIAA1549L, SNTG1, BTBD11, NPSR1-AS1 \n",
      "\t   SYNDIG1, KCNIP1, MYO16, STK32B, KCNJ3, ADAMTS18, IRX2, HECW2, PAPPA2, AL591378.1 \n",
      "Negative:  ZNF385D, TRPM3, DACH2, LRRTM3, INS, SYT16, ZNF385D-AS2, NRG1, UNC5D, LMNTD1 \n",
      "\t   PPM1E, LINC01099, GPM6A, HDAC9, LINC01098, SLCO5A1, KIRREL3, VAT1L, KIAA1211, KCTD8 \n",
      "\t   BMP5, G6PC2, PLCH2, AP000829.1, KCNH7, NRG3, CTNNA3, KCNH1, VWDE, RASD1 \n",
      "\n",
      "12:18:54 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:18:54 Read 18512 rows and found 50 numeric columns\n",
      "\n",
      "12:18:54 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:18:54 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "12:18:56 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203771e5d82a6\n",
      "\n",
      "12:18:56 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "12:19:01 Annoy recall = 100%\n",
      "\n",
      "12:19:05 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "12:19:09 Found 2 connected components, \n",
      "falling back to 'spca' initialization with init_sdev = 1\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:19:09 Using 'irlba' for PCA\n",
      "\n",
      "12:19:09 PCA: 2 components explained 39.41% variance\n",
      "\n",
      "12:19:09 Scaling init to sdev = 1\n",
      "\n",
      "12:19:10 Commencing optimization for 200 epochs, with 785508 positive edges\n",
      "\n",
      "12:19:21 Optimization finished\n",
      "\n",
      "Performing TF-IDF normalization\n",
      "\n",
      "Running SVD\n",
      "\n",
      "Scaling cell embeddings\n",
      "\n",
      "12:23:04 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:23:04 Read 18512 rows and found 49 numeric columns\n",
      "\n",
      "12:23:04 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:23:04 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "12:23:06 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203775979585f\n",
      "\n",
      "12:23:06 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "12:23:12 Annoy recall = 100%\n",
      "\n",
      "12:23:16 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "12:23:20 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "12:23:21 Commencing optimization for 200 epochs, with 807128 positive edges\n",
      "\n",
      "12:23:32 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "12:24:32 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:24:35 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "12:24:39 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "12:24:41 Commencing optimization for 200 epochs, with 571318 positive edges\n",
      "\n",
      "12:24:53 Optimization finished\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.6 GiB\"\n",
      "--------------------------------------------------\n",
      "\n",
      "Starting DecontX\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Fri Dec 13 12:32:46 2024 .. Analyzing all cells\n",
      "\n",
      "Fri Dec 13 12:32:46 2024 .... Generating UMAP and estimating cell types\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "Fri Dec 13 12:33:35 2024 .... Estimating contamination\n",
      "\n",
      "Fri Dec 13 12:33:42 2024 ...... Completed iteration: 10 | converge: 0.04035\n",
      "\n",
      "Fri Dec 13 12:33:48 2024 ...... Completed iteration: 20 | converge: 0.01849\n",
      "\n",
      "Fri Dec 13 12:33:53 2024 ...... Completed iteration: 30 | converge: 0.00967\n",
      "\n",
      "Fri Dec 13 12:33:58 2024 ...... Completed iteration: 40 | converge: 0.005293\n",
      "\n",
      "Fri Dec 13 12:34:04 2024 ...... Completed iteration: 50 | converge: 0.003998\n",
      "\n",
      "Fri Dec 13 12:34:10 2024 ...... Completed iteration: 60 | converge: 0.003906\n",
      "\n",
      "Fri Dec 13 12:34:15 2024 ...... Completed iteration: 70 | converge: 0.002502\n",
      "\n",
      "Fri Dec 13 12:34:21 2024 ...... Completed iteration: 80 | converge: 0.001764\n",
      "\n",
      "Fri Dec 13 12:34:27 2024 ...... Completed iteration: 90 | converge: 0.001316\n",
      "\n",
      "Fri Dec 13 12:34:32 2024 ...... Completed iteration: 100 | converge: 0.001107\n",
      "\n",
      "Fri Dec 13 12:34:37 2024 ...... Completed iteration: 108 | converge: 0.0009995\n",
      "\n",
      "Fri Dec 13 12:34:37 2024 .. Calculating final decontaminated matrix\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Completed DecontX. Total time: 1.946547 mins\n",
      "\n",
      "--------------------------------------------------\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 5.2 GiB\"\n",
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 5.2 GiB\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in retrieveFeatureIndex(features, x, by = \"rownames\", exactMatch = exactMatch):\n",
      "\"The following features were not present in 'x': IGA1\"\n",
      "Scaling cell embeddings\n",
      "\n",
      "12:40:39 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:40:39 Read 19128 rows and found 49 numeric columns\n",
      "\n",
      "12:40:39 Using Annoy for neighbor search, n_neighbors = 30\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:40:39 Building Annoy index with metric = cosine, n_trees = 50\n",
      "\n",
      "0%   10   20   30   40   50   60   70   80   90   100%\n",
      "\n",
      "[----|----|----|----|----|----|----|----|----|----|\n",
      "\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "*\n",
      "|\n",
      "\n",
      "12:40:42 Writing NN index file to temp file /tmp/Rtmp95L5vt/file3203774bfb9dc6\n",
      "\n",
      "12:40:42 Searching Annoy index using 1 thread, search_k = 3000\n",
      "\n",
      "12:40:49 Annoy recall = 100%\n",
      "\n",
      "12:40:53 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 30\n",
      "\n",
      "12:40:59 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "12:41:00 Commencing optimization for 200 epochs, with 830408 positive edges\n",
      "\n",
      "12:41:13 Optimization finished\n",
      "\n",
      "Calculating cell-specific modality weights\n",
      "\n",
      "Finding 20 nearest neighbors for each modality.\n",
      "\n",
      "Calculating kernel bandwidths\n",
      "\n",
      "Finding multimodal neighbors\n",
      "\n",
      "Constructing multimodal KNN graph\n",
      "\n",
      "Constructing multimodal SNN graph\n",
      "\n",
      "12:42:23 UMAP embedding parameters a = 0.9922 b = 1.112\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by 'BiocGenerics'\n",
      "\n",
      "12:42:27 Commencing smooth kNN distance calibration using 1 thread\n",
      " with target n_neighbors = 20\n",
      "\n",
      "12:42:33 Initializing from normalized Laplacian + noise (using RSpectra)\n",
      "\n",
      "12:42:35 Commencing optimization for 200 epochs, with 590510 positive edges\n",
      "\n",
      "12:42:48 Optimization finished\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "\"sparse->dense coercion: allocating vector of size 2.7 GiB\"\n"
     ]
    }
   ],
   "source": [
    "#for (samp in samples[1]){\n",
    "for (samp in samples){\n",
    "    pdf(paste0(pdf_dir, samp,\"_\",modality,'_decontX_delta_plots.pdf'), width=9, height=8)\n",
    "    \n",
    "    #Load in adata\n",
    "    name <- samp\n",
    "    adata <- readRDS(paste0(indir, samp,\"_\",modality,\".RDS\"))\n",
    "    DefaultAssay(adata) <- 'RNA'\n",
    "\n",
    "    adata_to_test <- adata\n",
    "    adata_to_test$ATAC <- NULL\n",
    "    \n",
    "    #Run DecontX\n",
    "    adata_test_delta <- decontX(x = as.SingleCellExperiment(adata_to_test), dbscanEps=.5,\n",
    "                               delta = c(9, 20), estimateDelta = FALSE)\n",
    "    adata_test_delta\n",
    "    \n",
    "    umap <- reducedDim(adata_test_delta, \"decontX_UMAP\")\n",
    "    p1 <- plotDimReduceCluster(x = adata_test_delta$decontX_clusters,\n",
    "        dim1 = umap[, 1], dim2 = umap[, 2]) + ggtitle(paste0(\"Clustering for \", samp))\n",
    "    print(p1)\n",
    "    \n",
    "    p1 <- plotDecontXContamination(adata_test_delta) + ggtitle(paste0(\"Contamination for \", samp))\n",
    "    print(p1)\n",
    "    \n",
    "    in_data_counts <- scuttle::logNormCounts(adata_test_delta)\n",
    "    in_data_decon <- scuttle::logNormCounts(adata_test_delta, assay.type='decontXcounts')\n",
    "    p1 <- plotDimReduceFeature(as.matrix(logcounts(in_data_counts)),\n",
    "        dim1 = umap[, 1],\n",
    "        dim2 = umap[, 2],\n",
    "        features = marker.genes,\n",
    "        exactMatch = TRUE) + ggtitle(paste0(\"Raw Markers for \", samp))\n",
    "    print(p1)\n",
    "    \n",
    "    p1 <- plotDimReduceFeature(as.matrix(logcounts(in_data_decon)),\n",
    "        dim1 = umap[, 1],\n",
    "        dim2 = umap[, 2],\n",
    "        features = marker.genes,\n",
    "        exactMatch = TRUE) + ggtitle(paste0(\"Corrected Markers for \", samp))\n",
    "    print(p1)\n",
    "    \n",
    "    #Tidy up adata\n",
    "    adata$RNA.raw <- adata$RNA\n",
    "    DefaultAssay(adata) <- \"RNA.raw\"\n",
    "    adata$RNA <- NULL\n",
    "    adata   \n",
    "    \n",
    "    #Add DecontX Results in\n",
    "    adata[['RNA']] <- CreateAssayObject(counts=round(assay(in_data_decon, 'decontXcounts')))\n",
    "    adata[['RNA']] <- SetAssayData(adata[['RNA']], slot='data', new.data=assay(in_data_decon, 'decontXcounts'))\n",
    "    adata[['decontX_contamination']] <- colData(in_data_decon)[,'decontX_contamination']\n",
    "    adata[['decontX_clusters']]  <- colData(in_data_decon)[,'decontX_clusters']\n",
    "    adata[['decontX_sizeFactor']] <- colData(in_data_decon)[,'sizeFactor']\n",
    "    DefaultAssay(adata) <- 'RNA'\n",
    "\n",
    "    #Check for cells that lost all counts\n",
    "    if (sum(colSums(GetAssayData(adata, slot='counts'))==0) > 0) {\n",
    "        print(\"WARNING: CELLS REMOVED BECAUSE ALL COUNTS REMOVED\")\n",
    "        adata <- subset(\n",
    "          x = adata,\n",
    "          subset = nCount_RNA != 0\n",
    "        )\n",
    "    }\n",
    "    \n",
    "    #Dim reduction and transform\n",
    "    adata <- multiome_cluster_single_samp(adata)\n",
    "    saveRDS(adata, paste0(outdir, samp,\"_\",modality,\".RDS\"))\n",
    "\n",
    "    dev.off()\n",
    "    \n",
    "    gc()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Multiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"PPY\"\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"PPY\"\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"INS\"\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"PPY\"\"\n",
      "Warning message in SingleExIPlot(type = type, data = data[, x, drop = FALSE], idents = idents, :\n",
      "\"All cells have the same value of INS.\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"INS\"\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"PPY\"\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"PPY\"\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"PPY\"\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"PPY\"\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"PPY\"\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"PPY\"\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"PPY\"\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"INS\"\"\n",
      "Warning message in SingleExIPlot(type = type, data = data[, x, drop = FALSE], idents = idents, :\n",
      "\"All cells have the same value of INS.\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"INS\"\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message:\n",
      "\"The following requested variables were not found: IGA1\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_RNA), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"PPY\"\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nCount_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message in geom_boxplot(width = 0.6, fill = \"white\", alpha = 0.6, pt.size = 0):\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `pt.size`\"\n",
      "Warning message in geom_hline(yintercept = median(adata$nFeature_ATAC), linetype = \"dashed\", :\n",
      "\"\u001b[1m\u001b[22mIgnoring unknown parameters: `lw`\"\n",
      "Warning message:\n",
      "\"All cells have the same value (0) of \"PPY\"\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n",
      "Warning message in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :\n",
      "\"font width unknown for character 0x9\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 6 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>used</th><th scope=col>(Mb)</th><th scope=col>gc trigger</th><th scope=col>(Mb)</th><th scope=col>max used</th><th scope=col>(Mb)</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>Ncells</th><td> 16233297</td><td> 867.0</td><td>  43885333</td><td> 2343.8</td><td>  43885333</td><td> 2343.8</td></tr>\n",
       "\t<tr><th scope=row>Vcells</th><td>716543586</td><td>5466.8</td><td>2928321400</td><td>22341.4</td><td>5307748528</td><td>40495.0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 6 of type dbl\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & used & (Mb) & gc trigger & (Mb) & max used & (Mb)\\\\\n",
       "\\hline\n",
       "\tNcells &  16233297 &  867.0 &   43885333 &  2343.8 &   43885333 &  2343.8\\\\\n",
       "\tVcells & 716543586 & 5466.8 & 2928321400 & 22341.4 & 5307748528 & 40495.0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 6 of type dbl\n",
       "\n",
       "| <!--/--> | used | (Mb) | gc trigger | (Mb) | max used | (Mb) |\n",
       "|---|---|---|---|---|---|---|\n",
       "| Ncells |  16233297 |  867.0 |   43885333 |  2343.8 |   43885333 |  2343.8 |\n",
       "| Vcells | 716543586 | 5466.8 | 2928321400 | 22341.4 | 5307748528 | 40495.0 |\n",
       "\n"
      ],
      "text/plain": [
       "       used      (Mb)   gc trigger (Mb)    max used   (Mb)   \n",
       "Ncells  16233297  867.0   43885333  2343.8   43885333  2343.8\n",
       "Vcells 716543586 5466.8 2928321400 22341.4 5307748528 40495.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Generate Plots and save in a nested set of lists. \n",
    "#Do this for each sample and cleanup each time to reduce memory.\n",
    "\n",
    "dataset <- 'nPOD'\n",
    "modality <- 'Multiome'\n",
    "\n",
    "indir <- paste0(\"/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/\",dataset,'/231121_WE_SSP_Intermediates/DecontX_adatas/')\n",
    "samp_dir <- paste0(\"/nfs/lab/projects/mega_pancreas/data/cellranger/\",dataset,\"/\", modality, \"/\")\n",
    "pdf_dir <- (paste0('/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/', dataset,'/231121_WE_SSP_Intermediates/DecontX_Plots/'))\n",
    "\n",
    "samples <- list.dirs(samp_dir, full.names = FALSE, recursive = FALSE)\n",
    "\n",
    "plots <- list()\n",
    "\n",
    "for (samp in samples) {\n",
    "    #Load in adata and cleanup\n",
    "    adata <- readRDS(paste0(indir, samp,\"_\",modality,\".RDS\"))\n",
    "    gc()\n",
    "    \n",
    "    plots[[samp]] <- multiome_single_samp_plots(adata, samp, marker.genes)\n",
    "    \n",
    "    print_single_sample_pdf_multiome(samp, pdf_dir, plots, '_Post_DecontX_Plots.pdf')\n",
    "}\n",
    " \n",
    "print_pdfs_for_each_graph_type_multiome(plots, samples, pdf_dir, modality)\n",
    "\n",
    "plots <- NA\n",
    "gc()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From here make windows and merge"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save out barcode lists for windows"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Multiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generate Plots and save in a nested set of lists. \n",
    "#Do this for each sample and cleanup each time to reduce memory.\n",
    "\n",
    "dataset <- 'nPOD'\n",
    "modality <- 'Multiome'\n",
    "\n",
    "indir <- paste0(\"/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/\",dataset,'/231121_WE_SSP_Intermediates/DoubFind_adatas/')\n",
    "outdir <- paste0(\"/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/\",dataset,'/231121_WE_SSP_Intermediates/ATAC_Windows/')\n",
    "samp_dir <- paste0(\"/nfs/lab/projects/mega_pancreas/data/cellranger/\",dataset,\"/\", modality, \"/\")\n",
    "\n",
    "samples <- list.dirs(samp_dir, full.names = FALSE, recursive = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for (samp in samples[1]) {\n",
    "for (samp in samples) {\n",
    "    adata <- readRDS(paste0(indir, samp,\"_\",modality,\".RDS\"))\n",
    "    gc()\n",
    "    \n",
    "    system(paste0('mkdir ',outdir, samp,'/'))\n",
    "    write.table(rownames(adata[[]]),paste0(outdir, samp,'/', samp, \"_filtered_barcodes.txt\"), \n",
    "            row.names=F, col.names=F, quote=F)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run demuxlet before windows to only use final good barcodes in window making script\n",
    "#even if we choose to do doubletfinder afterwards we can filter windows by good barcodes\n",
    "#add in a demuxlet script\n",
    "#will also hop in and out for a bash window\n",
    "#/home/.conda/envs/mamba/envs/multiome_pipeline "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run window script"
   ]
  },
  {
   "attachments": {
    "image.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAA8UAAACgCAIAAACqgVlPAAAgAElEQVR4Ae19u45cR7KtPuB4Y4wxBuegnTbGkiFH7ljzERqgCZDm8Y9JNAQ2aBIDQRhZvCAd0uAnkA6vQfDQkQRcSLiQobkESIPgC0MSaNRFRGZErMjHftSrq6tCEMjcuTMjI1asiIydtav4xSL+CwQCgUAgEAgEAoFAIBAIBAKBZRH4YtmJMS8QCAQCgUAgEAgEAoFAIBAIBBZRTwcJAoFAIBAIBAKBQCAQCAQCgeURiHp6eexw5sdPn9++//D6zbv4/3AQePv+w8dPn5EG0Q4EAoFAIBAIBAKBA0Qg6ulVnX5+fh6V9OHU0LWlb99/OD8/X5VGMT8QCAQCgUAgEAgELi0CUU+v6roopusS89B63r7/sCqNYn4gEAgEAoFAIBAIXFoEop5eyXUfP33G2vHf//jP+P9wEEDXx4sfKwVSTA4EAoFAIBAIBC4zAhuop5+eXTk6e7azoLx4cP3o+PbT9ehXHE4fTikZlv77H/+J9XQcUa8nokJKIBAIBAKBQCBwCREo6+mX969dOTq2/08evJxr1ebraa/ktYcv5qi41noaK6rXb95FlXlQCBTen8PCGBsIBAKBQCAQCAQC+4NAq562GvrVw5PjK3bZNpuqWxyz4Xr62U2n0sv7Z/Pq6bYRjd7SrsaQRVFRHVQ1uaqx//3HP3/xxRd/+dOqci7uHZvC+y2CRF8gEAgEAoFAIBAI7D8Cw/X0YkGnuSMHwGXdudl6+vntMX3W5bTSrpbcoqJKpeE///JF/u+Pf/wfrPZSBWmdV/7rjzJS/v7zX6/k+vLvf6C+stz809/qkbjEJWpvv54u8V/1ZffC+y2CRF8gEAgEAoFAIBAI7D8CY/X04vltftuYToVvPgc86Oj6+v3/QwfY+n5IOqVO9TS/VpFu+ZeVSaBMgUo9F+529/r9V7CcNmlA59ZiseADdZGf1k2a059pIj4h5NK/XtTJSafvxbl4UqioqKgU/vsfpAjmchkK4v/563/8+a9/+q8//sd//bev5Kh0/sM/fSn8z7988be//+lv2M/loBbc//PXP5RyvITLe+67Gc3JHX/7yx++sOcZ74X56BXeV45GIxAIBAKBQCAQCAQOCoGxelqrz+LUWfsXi/Icl0baKxl0V7+eSLOgGuaRudrmWzaSbkG1jT5J8l1xn25zWaz9T8+snsYvIILmiySqo15h19R6Gsqy//nrf0D1xvXc3/+Tq2o5hE6DG/X0n/72BZXdXFXnso+OvaE6Hy460yp6Uv63vychWQfpt8peer74gtdV4WSC/CdCyATpw8cAOzsHq3HwFypBl9PHg3//4wo/aagQkwzL8bJdEMg6L9Ausy/+PlpPN9VoS456GuMy2oFAIBAIBAKBwMEiMFxPY4VKR7Z60ky1plSuRd3JRSqWwiTESluZlRBPh8fU5npa5aeT5v45tJwog7RSDXFpWQeX9TSq6p4NegJFMP1dVFRahqZGruFS0Uyny1wj1iVdXU/LGJKQy0cqNLUeLRaqL1MNmotLk0914Rdf5CqTiloWjnryxFzLYtuWMGlcK8tx7z//YsWrDVar4TFD7/opSbdU4kP9ShJy3U8Ky3IqBBtoyL//kZ9JaIAKEWBxlm+jGvQ8U0Ok0grvIzGiHQgEAoFAIBAIBAKHg0Crnpb3JfILEgIG1NBVbT30fUStp92sLPXpWf4uI5a5dI8G9+vpNDtV1fm3+Z7dbI+3kj1NwoXofNr/tB/0rFpPUw1nFTCVelYZ28EwFXNQoabazgpNq0fn19N2jquVJReL2p+Ly0IyjeHCvei3Y3Io61Uyl551setB8JUrTSmOk1WywuVK5AqoQqCroe3FG1hoWj2tapBrslFmqeoW9XSO4vgrEAgEAoFAIBA4bARa9TQWx4gOVaJcfb54cB3GlHUnlKQ8e3P1tCu7d6yepmK0Vyz6OrKup61047cghqrbqqDMVa/WfDxAK2OtlfMwuYv1vYzRM113tMwVubztwX/rXLtl9Wh+WuCBVbXtcZB1eTnTn2poOJ/WhwGnlZmjMu1VGauJGepKDY+hUwMfdURyeiGEVox6GtNDtAOBQCAQCAQCgYNFYE49vVg8u0lvbry8fw1PjifX0zRd3xJJiFsPHhvTvSnn0zROJZRqiEt1QO7AhcrS36Qt6vfCRSD+XVRUUplR/Sqn0Vzq8TGtq0KxqiuOXenS/0dFJFerY9WkKMBvYthgLdB9sZhLUq22U1WqlzrLqlWo77HTtztn0vXbGlKhpulON6unHXT2UrVaWjbs0D0PpnWr/1zF70pzpwacT0st/t9//LP4rvA+EiPagUAgEAgEAoFAIHA4CMyrp+nd6JsPHp64d475hWl4a6IsUvV8Or8kbbU4jRRRWOYS/L16+vlteGfav3XNr3/oXfw+onbmF7VlUVLAvjqZvp5o73B7Q6guh1P5RJGiouLariqm0xmtFGE0pjj69fW0lZKpztO7XGdrITj8+x4oBApZXyxKHQkDylei5W0Hq5hJMv7qiAiBulYrcpv17384yWnwlHraj3ECYUXs5x9F+fsf4MMBuItn1ax5qrYVVf/AQIaAHELvz3/8Dx1ceP9wskZYGggEAoFAIBAIBAKIwMx6mn8+rzhjTrWvFaa+DF3IL+7lValu1t/Lgyp8ej1t04+tHM/S0xvVSX4WPn4+napqFmvFNAmkml7tmlhPc7npjkP/9nd4fzdXn1yZuZ+a1pPXuuSF8hSOrqHOg3pRqlunhpXytfA8Fw5xVRO6Bf3+XXA1MQvnpwjpVN2cGlaFu8Hy/UinG01M5+vufPqLsVKedOZF9S0UD860elrswGJaJRs+UU9jKol2IBAIBAKBQCBwsAiU9fQYEHDYPDb0EtwvS//ZKhcVVefE1Jd0UvJudLDVo1tZbnO2FOfTVN/beyybANaV9YVdBaqF92ezJyYEAoFAIBAIBAKBwF4gMLOeXrkA3S3QVjbn7fsPWFQV5dcFXhaV3wVqstrSVN3qaTf/fAdebreeLt7S8d9HfPv+w25xO7QJBAKBQCAQCAQCgW0hMKue7r3TvC1l177OyvX0x0+fL7KeLt+FyO8p/PmvV/alnk7vmuv7F6mYLl8XybfXc27dOp8WnPXN6fSQgK7/+Onz2ukZAgOBQCAQCAQCgUDgUiAwtZ6mX884Kn+d41JYOKTkyvX0YrEojqixxor2gSAQh9NDURb3AoFAIBAIBAKBfUdgaj297zgsb9/5+XmU1AdSNzfNfPv+w/n5+fIEipmBQCAQCAQCgUAgcMkRiHp6PQ78+OlzVNXNcnOPO9++/xCveawnfkJKIBAIBAKBQCBwmRGIevoyey90DwQCgUAgEAgEAoFAIBC4aASinr5oD8T6gUAgEAgEAoFAIBAIBAKXGYGopy+r9+INkz1+k2TAtHjJZKMRG2E1wL09uHVo4RN83gPSDphwaHzeaPJfXXjU06tjuG0J8Q3IgfxyILfiS5Brj7oIqwOJnddv3h1C+ASfg89rT5IhcBiBqKeH8dnFu/HFx8NJlAOWxo/0rTc4I6wGyLZ/t/Y+fILP+0faAYv2ns/rzfYbkhb19IaA3ZTY4l+QGQiwuLX3CMSvi6wrzCKs9j5YagP3OHyCz7W7975nj/m8rjy/aTlRTy+N8MX8a5Fx6rD3aXG6gZs+k3h285j+FaeTBy+XjpJLMjHCajrr9mbkJsPnYnYHjbbg896wdLohm+SzMmt+48WD60fXHr7gifQv6El7vqTdn7Htepr+ncU92Z63nTH5n6g8+99v3k0PsBi57wj89vBkU/9q6R6F6lAejrDa9xgZSpj0xLiR/Wjbu4NSPFn0f2ObOEgENsZn5df8RtTT8zGbOmPTm/Sm5YOd282YQspD3vzWZfvjG8cnd37rS/vt3jfHtx4NbcP9ucOz1i95sXh+++j49lMg5pqalJpvPl+TsG2IWSb2dzCsHp3SZwL0/9V7vwzTKe6uisBisaE0viGxY3FE539nzxaLfoKiFJQJduOJDHty6+j4yjd3f7UalHsuhoSjSfLJrQsMjUenHqhVGSguWI+cjfF5gHhjVJccOyBib24tfz79408/f/f9D999/8OPP/08HY5l9rzp0heLTcsHXcZoBENXb5JdXNzMC7/dDv55tliut9QzVhnbSFxrbNZoQm+LxSU67fVLXmyM84dQTy8XVmP8WZoe716/oTpmM89yq2jFc9eQTNbP/06gTTWWMrPUoKtnaZCw1d1B13128/j6/VeLfj39652rrXKwrqcTgI3KdZPkT4uOkqSh1Yo0mDF9DVHgyDkLz9HBG+OzUqxujFE96ukas6Lnu+9/+PKrr0+/PTv99uzLr77+7vsfigG9S1fvpkRGcOcnZjtmyzmOzt7Sw3RKEyz2+W3/Co5s/ORXOdrRT/GwM724QzJBWk/T0f5MI7IoK0kHA/ofaSXKm12JW0/VZJ5CxqbBToKKSg+dSciMyH/z7vW6g3/e6q0KeL0SRvNLc7mxWaMJ3SXE5hKdzvVLJpKsP2FZ3F1JweJ4m1iKYzxvIaKv33/+8ETCzQk5TgdpXM20yA9C4Jg8B50FV/6wHsN8IPYhpKhJs5YIqzH+LE2Pd69/uXtydPp484HT4eeg5mtIJuvn/zKGALxMCGKypeiCI0teZqKm2YmuuoSx129ksJWAPilqXgDDu58a2c7Yg2UmdRuV60wJg4wCR4DCoyRpaAXTl1hxzpQ1RIFbbhaeo4NX5bOVIppFWeTTM/9alIQMZmkoq4DJ1x5StSPvTOe6TmQenT1LK27knasUfFv9c5nz6R9/+vnLr77+/fd/JU1///1fX3719cRT6qqeNrexD2Rjzn6VS3abFMGWNZIClJ4kxTj5i4W/dcYvxRMVRNQqWOcEp6I4S5rCt/mcgBYgW4RPiX9FEYCXYojTjGZlyZ3EoR/PHV9Jn+LRxmwFPdfruk9Xg3Neg3732Z+L/9dv3v165+rJnd8e38jy7TgNF7UPE2m8PDDg2VtzuVZnLjLkFuuGMrNwXbH9iTmlaVVjwvseshzWNyZZDHl0euXGXZZ89d4jxnwIurRVNCSDOflTfgb5Ln0Oe3T6OK2rBsI+xDyR7OZIs+oFxg6X7BaqfCieoolWoZGaEJnhRemQY2Q6+WmkhAxXvRLgOehEvovlCbHvMZkfVuAm4ZI6peZGchP0D7KOo4wGa5xyT4v8nASAz0C5UsMbT+ihukdR003fLfnt3jdX7/3iKYpxnbOKKqkjixcGyqSh6aKOQdQ5JRPqKYyCy0YOLDUk9VAsLaqeKsPH1b6eIktfmUy3r/ktCbcGx14MohQ1Fg7EeQkHrx4UPQ2IyGriTEFCQ6mBj6tcbaTuLDqlQaR3qfjTWbZNtAanT2aUGzYYnCVGsVYp2frXVHQteFcqpVxShoUrz/NnQXlFZRcSSa0ruKSD+dOkJAFRBTVkm2hYUfFTKMqqSnyxPid3fkOZWWdQT5B5x4Qw7nl+jFwxUTXrclaXqoOIqhmexOCOUy7nCc90VfaW9XSHySOa7u7tZerp777/4fTbM7Tp9NuziUfULmtgoVk4qbyFL3JMr6dLT6POK7dJuM9rSDIUD/2uUPCl9sAnj8BmjRxsPL4h4Vd8Xtx6mO4Mpn0RkwLKL9optvNgqwBIguRBkEYZQdWzXbb1qI3pnsuFlDJyjkupEFfJKdupR4MlaZpunE8lAbWWNsXSrqMZmbJbnvjklkigGiUZxXvDrUcpX58+Jvxl9UYOZaNkgEn+5e4teZ+bsOV8zSBfvfcLT/nm7q8dGJlnG+F5XU9LFYvc5jZkSTeLboJuk8mvH1vnlUx+GXSYT7Dt1q30NbGySTgKieM6kdJi3ZsWN9IHROLu5hLa2d0vO+T/9c6pvF0NUQPkJ3aljb9HURhsfOaSS8gPkjsfdo2FEoZVruckRcitJrFRtzdNwGU6OQv0JGPrAHxyq5WCXr9J9Yc7dumRZWZ/pj3XFlaj8HOpnLmQRI2O/qaGtXVSwsLBKUVryXGM8koalDZzHZYbDiWinyY34T/XuG7Y66Yv0FmaGHkkrchiNa3xJzAi0waDB9mbJUlMJa2D054CE1tZVJJ53tos5fJGWe13uLnAFmZ64ke+sDRuu00+O/2Ruk1uk2TGDZegWaPhlghRJWHHk84FVCl5BPRABcI34ZZxOE3DW9yDaR+pS20Ijbzo5f7rwutpTC7gCcQ9IWw9/dRTv0tKs+ST5QFPcc7y6abIPv6StmFNhSrX93iZuRZBbtUFtNmoMrkBbJb86GISO13Iterp5mDLd/3I14k++eYKspSgS1N+0Trb1Lb9XlfEtPUGPvv2EtBAbCf10tm5qEr5iFKzz3H1LBmf1JNZSbFCq6yt1M1ipsiUfjXKNSZIFlUFZJki/V5VKwj0cxJPHbriTd0TuNxc3d0kyiXlgre8hhebAplCwJfdEBSFkILtdklTqmA0+Wgp6SA1MbZJwdHYXzasxNdGZu8U5QD5rtqze7O4vyZbn/y6qFCFT7z08ExFdSjajhRfzahkWkvk6Lppg8eDZLzVaguZMSi8gToLQO5Ww3kwSZASDSsbWeXxjUb+SXNTsBBzpBJNPe5Pn8krZlZc1d3h5FpZMWROuinE5+YSSaUiavrvd6EVCqNvdAnpHC24TaynO0Si4s+4IemrPdhRq0USUynV0+puz3kdJsvlelofFWQhslcjRWaVnXmw10ck0J4CEgzADp+9IywJAM+tMyN/p/zuY2ewTZzEZxpEtZYwmYtaohkWYzTIzjUgVfISUKoVhVAtBwlsSX5D31tIAFzYn8vU02t93wNdCE5C3BM41jOnnua5lGimVNXzvACFQp5oPVSL2LMX2IXcWmM9TXunhgds4Rr8kjIoqluD4cHdgrOXAix3kNj8TO87/e6rK0IC0pODK1ptF9pqTtQGWsHtOr+QIQBFFu4l1LO8pT6BwlwvnHO66CwytZZqwtiTTP2gNh29CJ4yBdRAbRP1LPHN4/DQ6MF6mtOxFLJcuVq9u5Z62gtRPS3EUhfW0NjWCUOxD5sEQmpt5S0zSitj8bVzcYMbEhomsCJweYtW9B/mdPzOj4jAmVQx0PRcalhEdyjqFSZRfCgofGNVhYRsqcjxOht1B88UE1ZOuMlRnDE/6HL0vor+EoXDXCR0jhIT2oPTjUUD9bSSaUaDnwlPrl0v3swGyjlhdRWit4stY/fq6Q6R2oep7cHqa3JZhyQWOz7H2lyjIudSO8CumWnRYWI15QrHsmSvjy6npNXtRlmqt5DPsJDwlhZqJhMaALGs47uDRXhiDSW9uXxuMdC2lZK3UNJEPa2hulgsU08vFot1fh/RtAEnWfWcb8PuDsPoJmcuYU9zT2URxSxbddlWubXD18KKteCySI6FmcWlagZs1tCyBu24dgDjQk6DX+ItbcOaX3Sw2zh1cKfhB+fsRp2QPorLpC1lMRiTTVD9KYNAMaGXvZKilYzUIsMHj7rZovYYM7adQMkiU09yuiAsMqXfpEl2pp62ZAeLGCsgyxTpd3blD6wrKipzVmhAxFVfeSyIapcuGHlxIn8+UZ5Mfre0M6G0FOMd226Se9sP7iwbVuJr82ybG4W725QwIeRZ5bwObvsdKkh79Eqfw2iRLaHUoWhtBVNL+MYKCAlZSZFTMDBfkp6WhdpjxgBxgUBPI/RsMHC6LKuQzvo4qsmN746cbSc29PkGbJnXFKJSaMAnNkUImEzYI6yTW9WUHs+pX/ZBQcazi/HXx0Ic4xyt3BMX4MiaNnVPGt/sb3b6jz4cA3Fpabscq5o78ljUtKXpLJFJQFEnbE9y6SVoFGjD4CrQ9if0nWFtQJL+VTZoDwbJiTtL8blmIPRYhldOKquF6mltoqve4i6aK+91oBxsp7mX/88l6+nFYrGG38srAS38B19yopHmJKKLnI1REsHvZziZrx7e1H/aTYVTAz8yXtaJXDqYVnyZtXIMI211WJEcnbb9T0BoVj7Ix/jPbYo9+fyLIhDOp/FWijrswcHUP7od5nyByciymJOQD60LbSVDFXlHB2OVQFksp36SLMUB5I46A9JyZJSgYYN1CU6aCJGNUa0wgZoaqDxZnVaRrCqZzuV6HoaotiXTsHy2QQNS1S4gy5QOCMxe5TZTSEJjWWLneS4pF7x1l7R6/rEOectCT5cT+efW0+lVDRWyePFAvt3rIqv8rUAXTc3Y95AsG1bIhMRw7DFu5K8xtalbhEa+rHbQ4lUlmWV8zgfVzB8hoTKZGz2KtiNF+MZxISRkOcTAOrJ0LVCJ5iYml+OB6jrRGghjOrq79ejJLShx2PxKshgo4KhAZ0t1l4YxIZBU1Dbieb7MuQKZbvOift28+CP1nNg5UuzT2pf35fu+xFKIL75s71/Dz4filBXr6cJHhGqbSJ2T1+ZgoFaVM9Wb2sAca6wDaiWGpLjrcIBWxMzMwl1nSzIPkJqbBjTBVKaVWJHtmuqzOeUY7yYwisa3Biss1Kj4XH6YP8BhLqWk8C2+Zc5HEsI65rCWNMVXbNOllCvyYomIxRSN7QG1LtWt5evp5cx0z9YloL4s0N9S4c9TfI5LWzjVqdfvv3Ibfzquzv9IcnZ8OroQCTRXmLGcEWlWypj0c2D5aATrmJwEUyUNdlG/cGv6+x5wwKaxio2Ug0iNb+7e8/9YCUVg/kAq7+vdwSlZpMHlHuaCFmT682aQoInGDbaymJKRKIZZCfr14zMS2ytKUupkUTo+pS1vNSf9PMy/w+dMY1RBpsuYptvJnbv5e4eyl0spg7mezydc1u5IBtxu3cnGEm5kkewHHRCIi0Aq+4RuFWrzXBdWsEQSnJ9jCeT0m0dWCsiLy4n8KUzoN3FRT7oswr9xqQxR4SCN9XD5ZDz2k+76pwUmRpO2u5GSi0VkXYsbWpIK1TUodImyQdT1VO/4HcLq9J5+mQyIxMHFogYo2ogU4Zsqr2GVHw+SU5KSZrV7jJeaoPdkmwI/oQGG5CdJw4TNqUBLQYSVugsreXElPzlrkimBZQOZCkYDpqiSTXmyRMMRNQdLPjymW6oV7kTp4TPfkpPmHDX2E6v9zQuC1DBkM+WSgPJ4ltDlB3tHDDk7yKJginLDjc8ElpRYJdjWYKVB+uUo/yFDIcGxzkYC+TWLWv50OLBAGC8lsvuEB4DSFU8fu9emtZ8cmjRRQ9iPPpaT4QpaG0+GV/ffpKRNaYEPppV8hkesKSSGrF781kJK14m31x6+gJDJiZ1uKZmByWfPgJYu5xcJf4p+Oz9m2/X0VED2EeuptrfGaXEjmbFIMVu9pJRhEb7VpXfB/J3SwZ/Rvnp4Ak9rLSJtvc8VFltffWTBnQqr1XlVVDDF4dbq8ndUgjwwZPWoavF1DBQchQnED9xrnp7pKxMj1NnabSxHRha1WCvMjMsDQaDg88v717TGHeFO3F4HAlFPrwPFLciQrPr2/YcLTw1RT1+4C5ICb99/8D8F+vw2fkiyBVqOLiG8HR14MQNEvV0Iq5VJRWdXxaEaXO7tcy+lIz3PSz/TBpcDqHL4WA2aHk3lY8yL4WNjVaFo41bdJc8Ge8HnvWXsACdXuVXz+dnNXTteqSm7Vz1RT18ad/JnMWdPP31eJeRmzMWPw+QD6/Sp7n7U02QF2CVt/Bx51xP6x0//jz471o+Gd4HLT8/gRIQ+Ftwt9SqIthxWm2Wdj9lDKKY5ocGH4P4f+BhOdx8/faYPKHbtEbSg6Kx6Wl57/dfWton+2f8w+KN37W0rl6WnfvIwKn//BlwOPhf03q/LqKcvnz/j7GH/UuESFvFpxO6xl7Z/ez10x4tphC/CagkSXt4pOxo+yMjUnllPq4Dg8+Ul5xKaXxo+K0H3sbGr9fQ+Yr0um87PzyNXLpFx9mnK2/cfzs/P18WokLNYLCKs9ilAhm05hPAJPg9zYJ/uHgKfL8UmFfX0pXBTQ8mPnz5HVb1POXGiLW/ff/j46XODENG1DgQirCby8JIOO7TwCT5fUqJOVPvQ+LyOHL9BGVFPbxDcEB0IBAKBQCAQCAQCgUAgsPcIRD299y4OAwOBQCAQCAQCgUAgEAgENohA1NMbBDdEBwKBQCAQCAQCgUAgEAjsPQJRT++9i8PAQCAQCAQCgUAgEAgEAoENIhD19AbBDdGBQCAQCAQCgUAgEAgEAnuPQNTTe+/iMDAQCAQCgUAgEAgEAoFAYIMIRD29QXBDdCAQCAQCgUAgEAgEAoHA3iMQ9fTeuzgMDAQCgUAgEAgEAoFAIBDYIAJRT28Q3BAdCAQCgUAgEAgEAoFAILD3CEQ9vfcuDgMDgUAgEAgEAoFAIBAIBDaIwKWvp5/dPL5y8uDlJiB6enbl6OxZLfnFg+tHx7ef1jd2u4fUvvbwBSuJbaf1q4cnx9fvv3J9W72YrAB74crSjug5d8PGbpCuG9acxG8VtMlM2ILhscQFI4BkoPaVm88vWCNdfiAoyp2CNT/aJeXVioFGacXA0C3e6m1hu6ntFoFZaimMLxSwG7E2EGKo7A60L6ieJoCO9f9VCrgNFig9L+5kxL68f61d/SvJMAFhWwdQoxdXbtAmLyYq8Pz20pV00r7n3E3atlgsNkHXTchsw7BV0CYyoa1p9O4XAkgGal/GepridHceA6bzYyf3u0VvC9tNbaejfTEjMb5Qg92Ita3uO2j+7Pb262n2kDv3pdpo8hlzz/GzLR+fcFFeXGLdXnJBI3EMtnHMZamnu/o7Y4YuHMgrV+e6khO7teeTTQUFPaThhz+ldWr2JhqbMmoTuobMDSPQI0Ovf8PqoPipQVGour6cg8rMbu+IGrP17tbT8yXt7Iwy/W5Q0YKcG1xpGdFTQ2wZ2euds3w9/eNPP3/3/Q/fff/Djz/9PF2nzjHqdHdOHzldqc7Ii/Li/HUJ1dGTD6xBse2s3yK8bl29mKZAV3+VM9ZwIK9vU3Fit1ZPb+pNjDKhl2IYlQwAACAASURBVNaNgbzS/WlMWGmJmHxZEOiRode/RbumBkWh6vpyzkq27oga821YfQuYv+aWZ5Tpd4PLF+Tc4ErLiJ4aYsvIXu+cJevp777/4cuvvj799uz027Mvv/r6u+9/mKZWP3oBsvy5GPWkd0Lwrd/yLZHiQzSu12UMlphJPgVhvosvQJOQuh9UctZhJOc22ZUk2Lsr+RYxNQs3fagTFdDIQU30/Q3sxFmiVUtaVx+ehCaIFP47x5WtiMeTNEtskX4aaXYVFaTB4o453YoLPmlQ9J4XL3CbJvLyt3Nx/qADFoKPPmgk6oZWq3ONZqRDA96W1fnNjYxG5ieoeky+w4lH+a10gGsQ6sUCzLz28Cm8+86fIQhdJTQWGFnUNh4WaE+9JPVkFfn4aCCI0q0EJnLDQ5QWR6Ac4IDY9fueCegm9OlUc/rjEisI4WQvf2XClvPfoAANjVqzJLCnBFh1nwYOWU08/F/llzc0RfQtmXUn0S8vJ8y3OOr5BQI5ExiI2sXKefPpGQhZLBx1XTaAYXkttlDa6AsmPARXgkJGzgJmcHCDuon5oIxBl4jxwhuVnWvBZeMHl55805yoALKPjGx8efbMSM6US98Iws4UaGqFo27SX2U2Fp2s8ISBQ1plCuXt0rRNsF97+IJoUG3BvUXBENlxFi4bu/0uUa4OgWEqAosUQH4V8ObzdOv6/f9jOs/49L5nlO8HrgpDXKQkHZiW0J+BnQxmiovkON4REiZqO+9QJq3YsHQYuTWJIiNoPMYLIS/bjXy9xzxYyPQobORqmXr6x59+/vKrr3///V9Jo99//9eXX3096ZQauV6YA7cylLJr+lwADmYJyUlJGE/UnM6uclhLWZDDQ0a+eHBbv4FHnhOKmxe9rqCq1IIiCqfTsGOTlpJRNmoKLfKiBWOQTHkELSQKUNfz2wJdUi9PKdUWM51xDJqVlURNJeWzm7oKD0urFCjZJc1Vbck16gtckVFyw9yKkLxKbE3/l/fP8vcs/QvKBJ1CsdD0ysubntWOjuq5zf7Vw5v5m6/IOg5jUcaJ1SLJvt8JE4egZs4r2ilB5CVAwgIMJ2niKec1Z83MC8e9/H1E86NTkgwHZyVHKPjed+1wG2ACUhfpPdOc9nBeV8iZneIu1QqnBoTAPAnqpvSpgsZIWlqI5IjXIFLblhm9bjl2pSUr51lypVIRrJbnOmEdFwSaiBiTdlwP1tMDa8lCpEOrXX1K4/w1A5reUBcOT8+ydYn5ktwcdE4BVHs45/TWn9LfzbqWN5xWbrx/iUL2ETdedcCJ2N7AV0ScArVWyEmf5GmisTo9ubkdQa1JDR4v1LLsStApsRPtxd18y7Exyy82Arg0R+SMCokddsxcxMtChaYrXKKzdEczcjoCs7EZkLlgprjQ5Jl2BDUw3XWXmvo8hfK6KQWRnppVSoiyTElW3psrIDZj6jL19Hff/3D67Rkucvrt2aQjarJQrMX51H5+WwpZIpyjkTm7/sKcsZOEm0tIJPYQ1ngXWYWqQD9NaWmLYqmNDgZVy1uY7kdpYeuagagmtsv9Ce8V+ggCaAIOT8kCYsBtITAS+gExqKJoAMhxvgA5lXWFwoaDc31XfwSZj3h7OjjnOhNAO08hvUGrdxRzYhtlENhLliJEPUhpWbMX8FF9uAGS/Y0VrkAlljIQRMWtAYicPoZ8pT9YWqLqRKx6Ydi2bISln93UjVNHMg3mSECP5085MkVbfFD2lniuarSLJhJGjtBKIl3ivqXrASVKhXHWiDddejcO6CqpUawl6gExcL+nOXRL1abpCmAheqnLyiiWUjAfH4QcMVDtTdXTpcmoQG4PqgFsN4RQSO4lIYrt0KImZYXWkFasCdIJtaW28YE0aIoS1dr+RYFpJPTQFFzd5DtWm2Sa2947SlHp1A+Fi54r/Q3Kg5zMCnKlK5OALTRxBpjumIlXMhDo0lHIpaNaQ0OVZmmAt+ppqXB4RTJn7QACanVz+/W0M9gUAhA97jQEesDBPNluNcpKQN9cktZ0dOd6RT4SUtKUU0RZUBUKnXQX1MNh6ab1gGJ8yzm+WJdm4UO2qKF/Nwwn6+QzZdkmbXWsz1RKaoD+3OEUk0fMLFmYSmPyvuUfikAHnlK7vsTBxRXhYFakNjwrO2mkgw3OCQsUY2MQAQeyJ4OHhAjmT145U3QUc2LH62kpEWhJg5r0xJzr/ZVh8QPS9IHqIbOoVBtA87fYuaZSwqS0DnArbmUlnUwz1ivD+XGQCSn5uoLPO0muPA3c6k1LSSVkRb3jml2kYSWEvTBHAm4GpLUFbxl6rFiHyWJv/tvjWSlZqU3OLZYDV5LQ4jKdPYucHPiFBGQ73fKWwmAzOak/dy0QVVqBj9C1Dg61ZalSpWJjSGWOIwaq3UDYKccXS2iYk5VlQqdtFih5mxcpwCclrxSB5qygSSzH8s/woqVdy3K1qdX1E1/OuqMHnzaTHmYLGS6RkraSDmFKurrHNis/knwgA6FUbYsDe0cpCneEEkS9RivUnIEGeS37y20WTM6TopjGiB4GM9Mm4ZkDH6BI6noDeUV9LwADuQbcRJU+IpCVzzZM8Kl75I7QeACr8pbtXyCkaC5TTy//vkeVpk0bsNzjTkOgp3AD3KrdwE7qeNfyCNNLizPrZ+pb1jBVLSZrkoF6OCxNtp6ZtMjbvEuOpo83PCVN2cw6+pgmJoZbMJ6vja80xZ9JKI/pFgMFmpSHecU6+bLEoayndYliutOfc4qOBCJBRuP5OAuG1dVDsVp+mFbzwcxypBPr8xEPHWCyh9oTDzVnOcnL+PlJaWyp2TLXplKaXVrXD5Y+RJ1wG2RC1p0dXWz2y5jl5xTYFjbaZa2hyFlFggFVhl5iHQfyq4cnmqBk0VX/LpYDV5JkuCTrmoFfSEC211jBYDM52TB3LRCF23ASRtpy7Lx4cF1zwqpY+flECUjFxpCmOeo4VNsj7MWvcjWSdZPmDhYA3xamTqtfC3oXl7w7Tyk1TPySrVqr4+tUAvZTZaVq+fDsNKlJy7dLuk6tp+15GCVg260OlYz0l+lX+tfxNxlLNM5VNV+eXCtPoDG+5oHZ+CgA9j7MFckaiI4aIgux0kcOIhsmCNU9cmdDfy9TTy8Wi2W/jwiHcM4gQNNVz2kQxrwb6UrtYX+XyKpMbVRrlVNEY1wI23Qf1Ctvoe0wjKUS1TTN9dZF4aIL/e34V0iGS9QH2yiqWkL5qo003F/SKrefUkbQxOoHuDXggiZKSKduckcW0lXSP8YUcMFlqQPdku0NhrnqAZQrmzplQDEdkycD/twDOaW8ZdqSfH+8h5qbWo66INlGrNgylZKg0jpQoLjVhQimkEy9HGQCmFGqBLeWbBaqFobAZRfhFSSAzJIPZM7TM4qOjZSGxXLqiISiXRaAw2UhAaNy0JsAKS0GbAfhdAcuca1eO2lOWej2U5qruSjfWOdfoENhjlEaAfFbA2liCK9RL0Cslpo+PATNB9UwUY7exXQSYSPrNdfdY2uJVhRBWFJLP62MbdbEprcUg2CE25UQFFtOcWQgrIptEefCGtQsRW0BWNNW3Eo9uPVI/3ww6+NIbyBIJuvhknSQbZoxIq9lF8MwvkUy+4WTX5EnbPiPJevpxWKx3O/lyVEfPlNSZjFQErG0qMqXNr7ACC+pbaFF0FutZtRJiGo6cx5iCcKncoq4AgMM23QfpNEt0IcvNcUDRdKTHCDgZRZf0FEJok3KGm18eJVWeeqXMFGoP/daAkI0kmnK47zxn9028PM2CdrK90hgMWqSWAG8dD17EFYxKFB/bPMuZRxgPeWonmmmgYrmVFabjuV3VRPO0xRjKchPnzSBKjzSoC45j5rrN0iKLRml0XhA3qyZ3XIoJWcZ01xNUIxkSH1QG3SqWxVuHSbod79wj3GPkbMtswmOP5WNaBe1TcOFcmO6BCakms/k150DPajaPb99dPbQlYY0TCitw5ZoFMtpPkyi4BIRYP3FrSQBAOFLjVaPVXK0GI789EIG15LpTvMivkh7egh54E/07T20JZDSKZZ/MGOgzjQUoHPEcGq7nSLPUibogvMbBcHsy7uMczqMJIV1La9VM9DACkJbXazaDS66Bq5O0Uq3HtA2PaoV24GwSLWHhjdEv+3N7NW8B0imRI0vThRkICr6bbGbGBv1tC9Jad01gKlZK2++VVomE3QhoAeDMwPMAooSK5BMHsBLahvN/LpDhRNrbhPRkHVtFkCWZnP5eropbmJnqvPk7SV1Xp6dUiSTOL3ColTmARnfXDGkwbouzPIlRendIuulhY5vP4X+coosQgpIPsI23QdapFv2O1xenxRX6RWum8+xlspVl9TiXYtEHZfBqZNMSNi6n6lCVbFtcrz+3I+KmSYnD57he0u6KGYWOXkSLwtibjm+SGHAOt9+CgDSTQ4tM0d+KMPrD3TC39ah+e4WziqcKzqUCYumZDAtlQwoJo8ENlgkpDwOdC0sLY95DO3803sJQAcIaAu8ZQIMbRuM+rQ/ZLm0iRagYelQ3hr2XSvc5OEqEcYxQbxDt2Q7xw9DptnSGYWscHsMjy/sQk2aW/ioBOEDm4mZreRDUpdpAMNoOlx2bJrQXSyH/PFFoVGawIfATxLsF/fUNXl1wMp50zn62sMXbmmjfWOtFP5ec8ETCM8JEHPRmnZT0w2fVwuGYFA4anm1AQSK4jVpSMgLIEywnHULFrlLcVNWQ9OdBJodqcrIFKEov7kod66Dq7hurRXZTNjmKhAxT+3uFtyKEkRP18IQQNeXNWL1NJ72YqQiaJtgVN6SU3ojTx68JMXWASYaaAIdOfOmScpA/1wwy7goHhhAsmCiUKhDCaKTBy+dqOxrulUUTmkYsEX3x7VtFi3KYN/F1NOoQd1uEasetfM9iX8vtqHnniC2Dagu5xqjXHIZ53LaOE/rTbxSPE+D7YwuQzu9AbKdtUdWKXbE4dGzBg+LGr3rCvT0UA1b9ej0Cxiw4bdTLsAiWvLCuTqaNi8ImGWWDTBHUetugtvbLKKeHvXSsgO2GczbXGtZPGLe0gjgpwQtIdssVlrrX0Df89twenQB629pyUZpqIcuW1Khu8ws1s0a3F1y0o1qW312s//h2CSJGx+0+xouAcHL+9cumKt7tC0GmOMMrAJfpmxvs4h6WjBf+9/bDWb+jGYdnwetHYcQOB8Bt79SmijfiUKRdIR5EMUlGn0Q7bHnqIsFYVaJPGvwKnZtbaFVlIy5W0Fgu1vwVky6uEV2H8xuPb090KKe3hjWu8+/jZkegldEIL/BJu8yXvAxz4rGxPQlEKDsAd9mXkLCxqfMqlxnDV5S9Rw15UuoS0qLaZcegdiC1+jC3Qcz6uk1ujtEBQKBQCAQCAQCgUAgEAgEAttHYBfPp7ePQqwYCAQCgUAgEAgEAoFAIBAILIdA1NPL4RazAoFAIBAIBAKBQCAQCAQCAUIg6ungQSAQCAQCgUAgEAgEAoFAILA8AlFPL49dzAwEAoFAIBAIBAKBQCAQCASing4OBAKBQCAQCAQCgUAgEAgEAssjEPX08tjFzEAgEAgEAoFAIBAIBAKBQCDq6eBAIBAIBAKBQCAQCAQCgUAgsDwCu1BP937qn/qv7OKP86PCO6tk4gSqCizhfy1id/+VEPph9l3/J4IBzWgGAoFAIBAIBAKBwEEjEPX0Eu7HIjXq6SUAHJsS9fQYQnE/EAgEAoFAIBAIBHYHgV2up1dCif7t2ZMHL1eS0ZuM9XRvzI7076aqz28fHe/u6fiOuC7UCAQCgUAgEAgEAoFLgkDU00s4ajeL1KYhu6lq1NNNZ0VnIBAIBAKBQCAQCFxKBLZeT/Obu1eOjun/fH7sar5nN+kWH15CP8269vAF9eS5Q+9VwzBbhWq4PPfo7Bk6C1S6fv/5w5Pj6/df5ftwS7RdLBagmGsvFgtY5eZzskX0TG06Nc9qeB1Qn6Xb9JqE2JjXRVUXhm3Gk1eah+1iQaucPdO12Ilml/tMANBQzHUiqwqOJuSz69MSiwWLtRep+XIDuC0NeEwMBAKBQCAQCAQCgUBg6/8+Ih5Mvnp4M72PYTWfL5isf5HrWi2tuFCTUrXpRxIFtd3L+2cPX+SBVFbqLZas7x6kilPr6Wc3tXojZaQ4BsVcPU1a6Vy2RafkWlbuorSm+vM7sURePL9d1dMOWxw8F9tUEGf5UjHDpdi46GLOTx2KeXo+cV9AlHp6sWDcknDUeT48MSMQCAQCgUAgEAgEAoENIbDd8+l2SZTLUy74tGL2x8Bc80EFJqekfVSKetoNbJZreQTWym4SCMQx1oYBeWI6k04XrohfjOvv1p5yAUbB8A626Ii52PqF2Gv61MEnyvqsAnrkU+3cg49V2dHyrMIjcImsquGMUqMdCAQCgUAgEAgEAoHAhSOw3Xo6HTce2akt28+l0sk1d0JJN6CEwvovYVb3eCzr6pYrP3kdIr9+QEu4Mh0XZYHpxDq/R5ErRVAMxj+7aYfTSZeyns6HuHwTS0avOV1xjWsvb+hbHL0GKUZaXYEDcpbKqtbYInrYTppYT5aZNMlAec0LnBuXprOW3Y16Wk+1SYV6CXtvJ6kYfwYCgUAgEAgEAoFAILArCGy7nma7pUrL9SVfnly7Xv7mA9eC6VVmq/AEuLpH7qS/fWHHryXouamVa7REt56mJaz0B4Gg2IbqaW/LnKv8AoaUpx1sET1sp5XqHtTA0KNegKW47GGe3zIH2BFPXskvweU1vKKDykQ7EAgEAoFAIBAIBAKBi0bgQuppNtpqJimnqAerW+nP57X4KkhZxtUwujrP1mos7d404Fd7UzHqJLjCERTz9bS9lk3r0DAVjmfVdLNQqTZghR7QXFQtsMWKGdu8KExvKeE1LwbbpR/m7Z11Pv38tnwVVR4SWlpFXyAQCAQCgUAgEAgEAheEwHbr6RcPbutPZ1i9JTVfrjK1pIZ+qvmO6WclEkx8OVJdmfz0+oSW4+kEV0TRMF3Rf2uwlKBHpKAY1NPpJQ3ViipLON7u19MkDQ5rlyXC0zMVYkUtqoeWYg09F1uExT1mkOa2NC6Rf/ZEMEetaBLiyebbEnQrP5NQpzpxWZRiXiAQCAQCgUAgEAgEAutGYNv1NJfF6SXmdnUFZSiUWak4e5qq6vot4SYqXIrJe7dZbCrKrVzjiXSZ36u+/RQW1be9Wcgz+8EQHINt997z9fuvsIbGNq2qOpBpCkXTkGmdYAUckzv1DFssdudiq5qzXlZAV5dTMOdnAKckiZElCDQAp7ichkuMCgQCgUAgEAgEAoFAYLMIbLeeXtoWrP+WFjJpYlXbTZrVHlTW0M1RT8/0nZDm/Y13bg/bjZsSCwQCgUAgEAgEAoFAILB9BC57PZ2/fqcHzPiKxTJorrO4LN4Sbqvz8v41fU+jPWLTves0edO6hvxAIBAIBAKBQCAQCAR2DoHLXk+vDOjTM33jOf/rhvirdrPEv3hw3ebSOTe8dzFL0HYHRz29XbxjtUAgEAgEAoFAIBDYMwQOvp6majK/PL3q2Tb+Y+Py3vYloEvU05fASaFiIBAIBAKBQCAQCOwuApeknt5dAEOzQCAQCAQCgUAgEAgEAoGDRiDq6YN2fxgfCAQCgUAgEAgEAoFAILAiAlFPrwhgTA8EAoFAIBAIBAKBQCAQOGgEop4+aPeH8YFAIBAIBAKBQCAQCAQCKyIQ9fSKAMb0QCAQCAQCgUAgEAgEAoGDRiDq6YN2fxgfCAQCgUAgEAgEAoFAILAiAlFPrwhgTA8EAoFAIBAIBAKBQCAQOGgEop4+aPeH8YFAIBAIBAKBQCAQCAQCKyIQ9fSKANL0j58+v33/4fWbd/F/IBAIrBeBt+8/fPz0eThKIwDXi3lIayIwSsXgYRO36NxLBEbDYThp7+XdqKdXcuv5+XlU0nuZLMKonULg7fsP5+fndaxGAO6Umw5BmSYVg4eH4PqwsUagGQ51oj6QnqinV3J0FNN1gEVPILAJBN6+/1DHagTgJqAOmcMI1FQMHg4jFnf3GIE6HOpEfSA9UU8v7+iPnz7vcZCEaYHAriFQvPgRAbhrDjocfZCKwcPD8XtY2kQAw2H5iuryz4x6enkfxplEM7SiMxDYEALFQUgE4IZwDrGjCCAVg4ejcMWA/UYAw2H5iuryz7ygevrp2ZWjY/3/+v1XlxHJ/Y6QsC4Q2EEEMFHsoHqh0uEgoFQ8HJPD0kCgh4CGwyE3tl9Pv3p4cnzl6OyZof789tHxlZMHL62n36JCHOf2Rw7fWYecHrEuov/JLXk+Obnz20UosOHfNnl0Kk9fV+/9ktf69c7VK0fHtx7Z0qmHRt54sm0Qfrl7cnT6uP8bL6Tb9rXK+vx27xsH1LbB6cMyVxMM67lzNzCe4g4ZuIEljN7rFz5G2vWvOMaEgRCu43376uGKSkXsjPZSCIwmqCe3jizzL7XESBwZ8SCNP75xfMWvyz10FLj9ffbxjTUuOgr4CFy1CzQcDrmxfD39408/f/f9D999/8OPP/08HcGX96+1CmIqsiedUq+jDiZt1yGnptRQz6PTK9/c/XVsOxmS0J9LQX5htdpg4K1nw25XLb39lfoLNNajxkqWNrTqO3Q5GvRnrZA9N8nbvsJdqDHVzJm+AgJDbmozc45iXUu3IWSFuFjL7p6qk/qBpBksvXjfBlAtDigVZymwFtxmrbjdwcsF2uisDdfTFAiNer2upxOYtRPrnrXDvtYlRgGfnZc0HA65sWQ9/d33P3z51den356dfnv25Vdff/f9D9NApKPo209bY63Apdoax1AJzqfXz27aKyK5KH/x4PrRtYcv0pk33735XKTPkbNYoPC0OvUMnprPC5gN1iXrj415prU2myxhhQ3bdJgppLEZz5RgSw+YVtwaW6KhVSFhg5crMGSDvJ2dsl+/eSfRTX/PcdMKCAz5JerpZZwojnty6+j08aPT8un3zbsLDZapFikVxZxJE9daFU1acZZ6Kw9eLtBGZ224np6Z5Won1j0rI1k6d61LjAJerj5qjobDITeWqad//OnnL7/6+vff/5WA+/33f3351deTTqlz+dsC3G5162CaZmU3C6FZx1eopE4y+dWRXFLPkKMle1pitXqamJpfTkgH0lRpSU9u6IsBtB/nwe5IFfrHT7XrTT1Hi3w4ZU/e6YyHV7TO16WGp4/f0FZ3j1+oOLnzhC3S8aabfub1652rJ3d+k+XSJ+A2LBtoBwAAEXy41o1YetlDEUtxrsJVK4t/vxnrSMG5qYaCn8timWXgg846+M271/YiynGlpKn0WkoED1FhC3zIkPK7+gVWBA/amwYq1n86KVYwx/QUECQUqDqFK1aggSA5QVT4qLhslKQggdRjPzowj5uf56Q4T3+2CQNCkskATuYA8rYiJ4EAEBnIreXIigxscpYQBiQgyGY16lDFjncEotclRkPy6zdAWlGMrACIgLQwuAw3pw+YJjGl/ATJyrfXlExkpPo6qUFacVWNNkqwAOAqoYh37U/y6W4rFzn9QSz1F/RgvzTEFrPw0a6+RT0VFPNxc0KUMK3lfrv3zdV7v6jaRjlcFDzS8TXoDMupWAjJVp4skMSXIkAN060gBqhX+4vr6UeykwKZQXLmBhPgLlOOn9YmvP5HQkAmIay5txULWNqCAkJyDQfP6uS4NFdnsdXsvjv8TuONJwlGQQM9Je97uOpfnzS6HOgQ5vjWI/UsOAU4kHWgJ967uQZILiiwYjMxMx9se5l6+rvvfzj99gwhO/32bNIRNZW/vbefn9/OZfGMOnjB9TQeZkPBPUMOHUXbwTZaNtRu0fTdr3dO5QVfigTLSi4Gcr54fEN5TMyWEKK2TWzFZF7aYl4iOQd/DsIkhOIzRfgvd2/J29WQQUBPiqWUlTjSbjxJu8KtR7TrsDQYzNtk0jllh7QcSE5ZSW3MVtOAwYyjwCaxUvGwjW6iphKXfxvyc/Z3wwwWLjsy4BlSAyEZ2B+ckzjn39JSNURLtLwK4ZwHtzmQkxqPIZVsq9CJKhwBZ8SSZPQUtS05ytI4UaWVjTZva34igcfPFw1PBlyYrw5Ke4NeWgMDslSVIuXJLWUIgJwqy3KVdjikYnrIlbCuWF1YAUsDyOgRmSi1e00MWMXMlyqtIEZPcisXAZ2QtKSnQjeQc+SWJATQrQN+z9ciAcgpwjvKlPHelExzJX8C+KgntM1TpEaNQKrOm75QKrbu9nioiRR06ODGvpb0Isi01nonD055sMFi1qXSHNJIw9clvLwWUgsgYsLjZiHB1fAmG5IDCjxSShYJiIy2eT+C3U3Tdb2jMQGu3vuFtf3m7q+kajecCSv3sOczW2euUFfVa7rV7qLX8oqMv6CBqpIC1M8DzJX84GdZwipa9RoLkZ0CJ+Lq0u4NbvGWWETbB2ueztoatNRwOOTG9utpPUv2sC9/Pu0FLi8Hz7m9bp0roWY7bPRIMg9r1SUoQUNUYqwrFmdx2/ZmucXRohmzubRmCmpoeKgoiVKZm9X75e6JRbKVTRr/pIBKLtqyJdDgfoITE8B83BVEiBiumtt4p0wajyplCf5UTMxMxxKa2cUp7cFyl5duLGEqlWSARxG01wSSyWoaeZNVUu+g5KJTBqtFZK90clJW6/h8SFdBmdB2cnI/JVagQTIBYO9Ww2Is6ky65X1CnDuliJnwvodw2CMgOoCNjqiomx8j6oEEHnyHjs0AVdp44FLUmBI7RIxBjzSJ0ZEMelqoGsec1bOeIggWJ6eNTDIE8URfm6XAnAx43cO22BS5VJxNsps7FpVYdtDEgtUeWMRz/HzaMHFqT8bNLCrWbV3SYKNcjtlSgi5NljYyMHmqiEQshWldxZMatqJKxlSjerbDwSUWr79BpwHoMGx4CnQTAohM1bkhFvhWuD4N7swFe7OEukfNLxoueWb5parZBL+6LeGgU2RESNLcjVEYtVENbvOB7Bf+YgAAG5hJREFUQ1hEiQK6okqjRqdKOqzuZerp5d/3WKz0/jR5pvG+xzrq6eR0Ej6jqi7iJF9SDMAj73BRy09++nycchmm+PYSZV7AHStR3EeLjad+XU5SKqRRq1wlZopY8gqTKH3OVksxC2Db1EhPuqRJmb5hjNluWrkA7pWDkk9hcK1G4aYjKRDrkTlTI255sOQXXqg5EczxWoGDPKQZEIHdQMhq1MWWeCqvJZKdBOnEFwAyS2uBgJt9KO87QY7t4orAYBXCFjEPE2FoltdhcDqmZweOQE3hYwGokhUBNITVsMHFoS+O7LUpdng5PAArxHbeZmnGTudByyx1bhWtPIU0KtPDoaHBgA+QVqGbEpVOzhD4HV+jIcockeODRcwsHzbakt3cSrIhyWvR4FxJkDRv+MhjoVKxkJkuFUzGX3nYfg5pDa6zuuJQN0h5i8ScMH2nfwTSFZ3JBJfnM7oJalYrrMVlAkK5qJydV1HpJNezCht9lrO5NNEYzq4UAojMMQ7kw44t1tMOcwKwVDWb4DW3iDPz+RO5nD9FSPJId9NMwPrBsJASw3gry4kC3hdCAA2HQ24sU08vFotlv4+4cG8qG/D0bob8vge2aYT7XuBYPQ3y58hpa2K9zZZkEIx8qE2LtwCFlDaLk5cmQSGrnSTZSKFsp6fOvD5aZDqFimYNCyEarClJ9JGYEbWzenJZaCIpjKEwyXCeITr4iQ4ufwtQ7aYGUdILd8qkW6hSr2egv57Og9VlpHlnjBrltRLNaZZthCawjbPMSqrmPwvvy6WToHzQBsDrpFX9Tk511+mfhQ+cLisamLX97jK1iGmeTxPIdtCCcDUM74QDzqrsLbHKaDtRujsWgztItolRzNXLppBmJ5fmiq2uYhzrknZSVDo5rF4P/KavsTMlH8k8hLlqq4ThRukaFFJbSlPGolLen+EEqLlRzFGZXo3MCt0U6rs9KJrn+p3BDcbWC0mPH5z5QJ1gQnGZrGj72lQqcq9edoH1mhCSdQ8v7UjbGaO0949SSg8Xd6KS3BWZ0i9YZffhJU3x3s93O3Nr8tc9KB/brZGlqtkEv7pNdNBpUIiQBJobU5vcHkzr1vlTRIkCuqITq+FwyI0l6+nFYrHc7+UtFlTm+p/MK39/mmpifc06nRnr72zY6xzsNX5/2gbzpdTlXLtPk/Pspr7VbVW4q+NbHMEgkbZUMzmVwwt5FBt2SkHjsYfylOQ+6rcaSyQ77vpOWDQnIB8tkpUo++RTZBqQI0eipZLJ2spdiCXRU8SW+5/LArVuaAipARk/n1vjzkpaaRKHFVnbdmBLPsWFajUYAT1TV8lOeZXQHgxZmORDJtKJ1kCtbGKPAwK7dwrjU+V93FSaklPZkVClAZYxTT3BWRkit1BDRckaDliupJ/c+wYPa0kOKkALtUHOI5EPhfn5U8WnZ+mHd+q7Zn5e1CIOyJ9Ngx4IhzSxArlei3vUfMeQDsg0uLaOBgsPSaXhpdvEaEpW3RLg8FFSXoIGtMhAtqCepfvY+wh1Agd7eAqD3/Y1LYFhDr4g7yAmgLyP97ZkP7cYw2lW0e6Wet2PZTJzkkoL+riVPh0FDfOANhSzcMsv1pehVK/FPYinkYHU0GBvZ9HS11m+DTZpCa5MjAJYSwgc7MLnJM2poSNJQg7PJsG8peh6IzbQhgxJxgp5BBOnahpmacE0bMadm2vel93QetDjqHltWj3XeCjLiQlmKWMoUWneQXPEXkIYvUZKVmrgYGont6IVPIWBkpwjmqMvDAENh1atdCh9y9fTqyDEFbOdibovFJLcVHPzgJvP4ciZ7lGZS9sAV8CpvH6afuWD+rWYZvWmyhGZTsJS9TRnc/nIjH4fAzJLCglWPmfJxHLq+ebuPfy1doqrZObY/irBg5uTxacmr9QAsbfopc+kRopJWS6X8hIzZSzlWE1HSvrPqUj8c3RJUshpJe1hZE7KYn45wKcV9izQ0odEr8lMarcky+fpHTX0SzzZ8LF9wqkt1YZ2pi/XD21+4H3n0zYHBHZMzalt4+GhCzpNB10x/XyEMkT7E/HsN9ETqt4jubgRSvM/WEPJWgkgUCTX8K1KQlJPFVBHl0LabhWnM4dTBkhhXoOTNpIk9uTOXf8usqkNvs6GQDjQcoCnK/uqFUmm2JXkZ/w7IJsOLnYUz+amjlHcJcawZP6tnuyXJmm1kwHxHqzdxzjAlDzedEDwEcx8alCkCHtmBpkJkyS5Q4yG5KIWLxaqGV5Idk8aPZJnQtLLh3zQU7HCJUmEYhZuPvSklkIyWNvhhiEJPNQi0g22HcpBoYOdLTa4ANY0ceNVE1DDsp92FglqGE8NHBpGapQhTGKJNmQmxaZTNdmO1pE3aYoPPdVNclSKa2O4JBn919MAVQifOnakKsW0VqoqJsAPvNx4gp/7KfP5Nzoa258in8Cs1ABt3UO+GWi8lZwjmktt4Jz+TsNhlcrwss+9mHp6bagVx9VrkztJUCvsMUguSVuiJZtDWdVqsj2x0Ud+GHXBCLgdLm145Q7X05AftvNXJnpjLle/7Z17ydIVfD3ixzVIljomI0+XRSEyoMPL+9f6z3VbTv6FIVtePZbbBQQumAMaDpOKpz0dFPX08o59+/7DQLa9LLeKR3N6kPVP6pfFkNDz0iDgn9n4KGjSI9zb9x/oU0V5+2t/AhAOtC6NEydW/8v6ehyHNUim0zj5eCEfdtrloIFv3394djM/1+0ADy+4lhp31iCYMX0dCFwkBzAclq+oLv/MqKeX9+HHT5/XEQYTHm3pJMY+fJRPoGYcpQzqSXGoMqOYHsRqzFmb9dTY6pdn09LPK5l4k4rp12/effz0GcN1awHotdVgmar2MKMa59P7xSKP3npAS5CuQTIV5epQqK3HQgmpuDUe8pOnaSua599anvgkMMzGi73rHaqWrpMzF2vgJle/yHoawwFT9KG1o55eyeM7cDKxPzXWJnNNoHTpEeDD6TJaIwAjaraPQE3F4OH2vRAr7ggCdTiUafpgri95PX3Rfjo/P49MuiNRHWrsMQJv3384Pz+vwz0CcI+dvpumNakYPNxNZ4VWm0agGQ51oj6Qnqin1+Doj58+R1W96bgN+YeJwNv3H0Y/TIwAPExubNnqUSoGD7fskVjuAhEYDYc1lFaXTUTU05fNY6FvIBAIBAKBQCAQCAQCgcAuIRD19C55I3QJBAKBQCAQCAQCgUAgELhsCFxMPR2fi13gxzSx9C4gMPfDsgiZXfBa6BAIbA6ByAmbwzYkBwJNBOYG3XCFv+16Or630XRqdB4mAlO+zBEhc5jcCKsPE4HICYfp97D6AhGYEnTDlXS6u+16Or63d4GkiaV3EIHRHxuKkNlBr4VKgcDmEIicsDlsQ3Ig0ERgNOh2rp7e2u/eN/GKzkBgNxEY+P2KCJnddFloFQhsFIHICRuFN4QHAjUCA0E3pZheLBZbPZ+Ok7bahdETCAw8GUfIBD0CgQNEIHLCATo9TL5YBAaCbhfr6YsFK1YPBHYWgV647qzCoVggEAhsFIHICRuFN4QHAjUCvaCb2L/V8+la++iZicBv9745vvWo/29HPzq98s3dX9/0B6x6ixS4csT/33giyj+5dXTs1+UeGnb13i8bU2bM2F/vXL1iSq5TjSUkP74xhEYvXAXhdSp/WDJ/uXtydPp4VdrPwP/xjeOTO78tC/IuxVcbtLEU9ObJrY1GfVur7KAlwR/LJF1vbpJdO5wTtpLea0c/Os1bz7oJRvk8bWqQK+qMzT20960Q4DMyiWPdGEWX2JKc/Brt6T2bjIK1KTnNnF7QTezf3Xp6ycw4DbUtO6m13Oi21Ay8sVljUdfSpLlQu5OCtlGv1/V0mt7YWdfp1jFjN5dilpBcZ2f0RS9cccxoe53Y7mIcjZG/qfPWc/0qXlg5vpaCqIlbt3N0iUbUj1J3XQOWBH8sk3TV2yS71pETRp3VzvNdex0rtuxo2mWGzpJAt3k0ICc2zn16GbshfGn+gM4jmI8tscSWNLLidN3WHgVrFzjZll7QTeyPenqVhLLK3OUy3dissahbMYQaqWSIqY2EO1PCIMJjxm4uxaxdci9cZ/lrndgOuXXQKRucOEb+5tJbT82reGHm3Dq+loKoiVu3c3SJWqvtEWYmgKLYWCbphuEm2bWOnDDqLEGg6+6BAdt19Byo59FgpvcbwmdK6NJpwAtjS6x9S5qh5BzXTBK7doEDwPpbvaCb2L8D9bR9iJOfPokZ+cOX6tWCanByD04ZfIR9cuvo9DF5iyXDUat+lONfUdAPttz7DO3lVOzRsb1mkMJAb/HrB7BWNlA/PwLJ+CE1qDH8jM7LPRYAVezrN5RYM6rZ6t/ufXP13h3+BO3Gk6TSIHTvkhCQSanWFG68WeESro1U5zamSPp+dHrlxl3W+eq9R+wvcZaTI53NKKWRYtoVhxvgmXUgYtxj3E7uPMnr9t9UmSP53Wv1PnyYWCjcC9diWL6sosBhkuAVbOGWO4OB/uFTn8sZMgDRlT7shCd55/QxvaVQ5AQIGQGTxoNkiBcYjB8Hw+AicFqeJSHFMHMT6pD3ABdfM7MKrXXr0TuZZdywFR1ugg+jBIZLwNq2xFqlmPWvgYFkXa6rxmv1CAqxUIIci8mtDb4uV2vLPVPSJuIPbh1hF2ZIC0zHFp+aTMNZOaHGVjwrad+QgdVd/oR+53pTyZPW0Y9vAUlUMmXyJ6oe0BsGEzITfOS0au9fupCVEOg4Y6kZRVNU2zTAaIa7cJ7i6mkbqTjrFDPQrG7VAwV0NrggT6Gkt4WsaG52wFUN22SCYqX9rk4wtM0QBxRIhigAFg0j397iYa0UL4kYjLMiw9xWnM2Vxs/SLzRY7c3caKnXC7qJ/RdeTz+5pVaRewwjx9pMnfZghskmGqaecMDaNJgcnzxEEoSsKK2lg27AhReJB8JLamffZ87xiuRjzRq0uowXUYAAqGR6JrqXs9DMtFyCFJb79c6pvMes0qhBVgtTaUX1BcqkdoflMqwzt064tH9rVAx5ig2R/T6VOwwdQESVjXitKYpdKcvBxMc3lC3qNTbwxpNULaV1B/ScI1mcSzjrutIpAPbCtWVXOwpev2lh+8vdW/IuL+kscLH+XWX8osn1abCSh3NTS1rbv23bFfxMsLWFDDBfyuUScLOR+S/bualEKTuHg1lNgaAx4hilg9ELDf7buuJ67kkg65ZclhfkL13XJtbySdUyP4CewAEOf0lTZmx/sIRDawlTSXNFSZjXbSp21OCEIyuq78w7KSOlAaY8hgByAIxq408DxIMwESSTnlkfGDCNXTnnkzQJGVODy027BCRn5IQ2tvkEpOBDay/II1scU/DrRkE/gCg94STGIrbgCMOWA7BQEgHhZIWhkTzFzOnsX+0sBNgm+aRDLtpyw1GOFGskyYbwxk6EaABvExpJrBFpYLAo0FjCeSShlPU3nPv5Sghf5oQyz6BujHkaYMq7isjcyhxweBb4MxSNLd7OOJyBvDVXNUAhM1+CzrSKFl0qkA+JWnN7QTex/8LrabUwJWKzvMFaZ78GM5DVDUDJ2tZZ1ENMInIUEsgZKbzbGbAV/0BKXkvZ79xpkpuV8eMbuBeKqiqKrEMJahQ0qsE1ocVqESVZQ/pBWoknTakFApLFXDEB5Iy5VSSIITI+iRKdk0AZg8kX296iwst5IS+fTx95z5N+0QdMqOwdliwSBGfUUNu9cNUBnYZDeFhn2PvbCk9ZQiAtJJhfdiFkHA6DsJPJPpxlrk+4baYJ+H4JkeAfEjx/OlCvHl/mCF2inVWw4kmH7kJ7qGzEQGd+Ywldixsyi00mwtRFpCFG0qyMk4Was8rONNhEUaAp+L/euQqZakxnWZf1p8E8t80BXYIG+9U9DqSP09mqHMkJtPVIwVQxZMmc4FQaMZzUs6q3q0ltF/c4R3O5AxJUDYetTsEEooArLK1GiZ6YJguZLUCDjuZOvvMRekEkF0IcAdJ4ZyML/+XuCdDedGvWA+3BYmBvCVTViplkGsKrxir4FCYWlWpmiTBPLDplsANBOrmUAg7UsKDOcldEmXq9sGIY6VNryFFqHTRIH60kG1D4lAgT37zrBd3E/ouvp8m19oyoKFhmRDY3BjvsHDQ4UdrgM0t2rtPXrMTppJ5zIZGM+yVmLGAK9gtpRAHV0EcLzbK1BBBGw0moZ6lAbrjBgCGhBDhTDhVREgylCUj93KYpsEvZ0p25BbA0XoLH5lbIOENkfBIlOhcgN1QlIV4rmKvuY0xk73Roy7ptPedIFgmCc9PeXrg2BzeigBFo6UxWC52O8+c/+x4yDodB2Ane5gDqRNxs+2mA7yVUqyc5sM106JpCcrX4ApLnVQoC6Ml3PXJNKYhqREvj8CFSoUkCpKkG7/epzgOsfNDJo28HfO8mcoFL4CCWONBMmx0OVP4d9CylmowGqeSNGtjXX/e39lZOaGLbOX8p7GKVCFvZy1ryJYk53HqO5sHqF4etTmGFExo0EgjjloB1i8JucP9yPuoJlP6u7WqCjEzINIQ7G1lnv8VQBk7G1iP9K2Q5V9NgHxrNiaCYjw6b6wPBqFjnmTYOxbqCiQNBOtPZBGw3ljYbpBLJIkq50cnJZCzQBmyvhFMNnQ0sacPBnnzRktDbiCf2X3A9TS60B3QAtFV4dQa7WRWyEJAEnxssFCyeYIpLzQ6NjExk5TRUcNEuhTSVYsZ4uVX3SGRapuuMUWa45XQw0MuqTLkrwSBoFIjhJU2p47CqXHWKQzuZKcGjYzoNMUTGJ1Gic7JXxgiApShvkShD9porvXzbWaW/lJnWmiNZJAjOTW174VoPpqUnh4zyk+SYAgKF0mao4QaL4UWMFJdsssdZDVGVSJQRG07yum713ied6x5a2vnOrBZHFMY2BzQ7U+zU4PvBbnVZy3utowmbs1p81YDUPWn1Zn+z04KCndgZI5Z2ciyXyOpuQ6wtTWjmgGoTxkTRYAVfG0q8oYajnKjkJet0J7kzRgf7CsNX3v03PdL06TlBY4omOpXEFucdy+GGs0PAwW62mJA0wGWG8nxa6xgn2aaQzvLI2uR8ua4KzGqIaWKv2QI0KIWUJpAhBa9siki2Hp7uCJAEOhsZnLqnN7J6nJPlxMCBiekW/4nmawxSZ52vevg0dS5gl0sHggKlDVBMzKlIJcuJKOOG57BN5JH07aYxwhB0yi7/ID30idDAQ2wvGIv+HainJclygNlzas1y7MHBLpuMOLLwWa6rUAKuAlRoVQwYisQkrdJgsJAGRGV+0KL+OcmzX2hEku2xcsZBi1Cfo0t0Yz15XQlXiQFaXfVJw8Q1ojxNaVLZzTUXANrS2YGXt1sD0LbwItgMtFpDstdBilrZujRRaMZT2CJRVfwl67IXVpKMfvQbqmDCYfzq4cnx7adFeC4EeRGClOO0qKdfzY3B4EpP9pJbkfD1Er5HkMlb9SUIGfM1v8oFO4rBaDYK/62H/MIZWcNBPAWSE2MTkSzeaYAejcgskmzByDrU7JVFV4wv8Hg2llWqiSfhj0qWDwwCF4QMycc49dMZQyAM45/2M1CMsc1UbKuRylC/EaZKUfKYSe6AT4BLmBdK1uBL1HtPjXKAlnbsqiS7NIJq0MjaKQI4j1wsnt8+uvbwxXhO6GBL0uBWEm5w5Vo/k5z6m9zLw8q9oDycSruMSCDochux1fOsCeb7eKwiSJ+lRRTxXAIW4xTkJOKVrOgMLp5MzDUNn5IOhdgOng4NldkebL4j+YNnvSlyK/PRNJImSjZMIMo11cBOcytIpgESBe2QSS5gBTR+u1s8DyaZZfhrCq3Q7kkG1yeoSb1SLEZl/0OhMgg71xdcTycXpieJkzt33QeFuQLg5wyI+ebgBGi6NYhX8n1+dsGRIEHTnBssmYIfZ+XBWmjE3kqk51s6uPgk0TvY5Ot4oqkKlxSmnSd3fhv+iJC3bX0yU0NQZ/4VC8JTuEVq00haRQKyk0MtnMQQjh9VWMvZVIBaP+YamKLL5dIQuC55RyJfN2kFjX+nRSAifdKiIFNxI0hhpPn6m7v38vcjRX65Lnt2JckIfnKN+UVgfLd48eD60dmzKkp1ADQUgePxkAFO3rrjtnADoc5cLsXYckUhBRLUIje4zWosI0A9HbyOkFGOXb33y8iBBFNd9dcdTkpq4bCoZwY68JXw/FMGebB2khAMgVwaIifZv2uJL9Qw/7MyrUCgtTABKsFag43D4ymId2VNYrYE+Bqo2FUjp6CEvwYvCBGPSOBz8nGvTTv8wcVJiMrU1JF9DSNxD7Jys8+uWjIozJgk4SpBcjVkLXXE4unZlZMHL6fkBFgFsE1kLvkA/tW9oBzZ2Ne6cCG3bS3bSiSjsl2SZiXhK0/Mmy75QDySN0vXELvIduonuwxGQLjs9JGYJqJ1qUcCH7wG1vFd1BlRlX89yo3PgeDQAOvaUaMSqs2uQgkUwM1OJbjNQnZVVGCAAyakhXCRYwF5CxkSnvYLSwgChShj3CCqWPCyv/hS57Io82Mp2YetblsOIvIg0ilDUUXbvI4Lr6drj260x/usIqXmsmgEAttEgPbOm8/r2N2mDp21ImQ2mpFC+D4jILVCtpE2fivvRgx/ef/a9fuvdjInjGjeySQwS4rgPJIurTwanx4bdyAwgIBU6plI5ZMY8LASUofbrJ6t1tNv33+46FCJ4mCITBftnQPV7e37Dy/vX6tf9lgsFhEywclA4NIiQGd1cIpJ53xwOZTu3r7/8Oxm42WP3cgJQ5pPcpYvcfjgsHFYOElUVRLFrANHgOgEHzjQQyxcDoDz9v2HWdVzPXir9fTHT58HjFnjrXT+r58lSQN+xvhyByF8/oIfS00++Vgj1CFqdQQ+fvpcR2bqiZBZHV6WECGzcg3kc2Y/x655oTUR4IK0gjcx2i/We1TV2F3ICarMJhqeP/LCBm5nuR3n1kBdTycpbKY+pG3Cjzsp071zMrGYfv3m3UDQ9Tboon+r9fSePFt3MuBOEgtCMdTeSQRGn4l34Ig6WBQIBALbQyByQmymgcCWERgNuqJ0bl5uu54+Pz+P+mDLRInldhaBt+8/nJ+fNyNTOyNkdtZ9oVggsHYEIiesHdIQGAgMIzAl6HRHHmhsu55Oqnz89Dmq6mEHx939RuDt+w+zPl2KkNlvPoR1gUDkhOBAILBlBOYG3UAxvVgsLqaeHtYp7gYCgUAgEAgEAoFAIBAIBAKXBYGopy+Lp0LPQCAQCAQCgUAgEAgEAoFdRCDq6V30SugUCAQCgUAgEAgEAoFAIHBZEIh6+rJ4KvQMBAKBQCAQCAQCgUAgENhFBKKe3kWvhE6BQCAQCAQCgUAgEAgEApcFgainL4unQs9AIBAIBAKBQCAQCAQCgV1EIOrpXfTKRnWKX17b8i/yxHKBwOoIrPd3nTaaYUJ4IBAIBAIHiMD/ByMo8vXAo82OAAAAAElFTkSuQmCC"
    }
   },
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![image.png](attachment:image.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_window_script <- function(samp, outdir, is.multiome=TRUE, script_path, cellranger_path) {\n",
    "    bam <- paste0(cellranger_path,samp,'/outs/atac_possorted_bam.bam')\n",
    "    if (!is.multiome) {\n",
    "        bam <- paste0(cellranger_path,samp,'/outs/possorted_bam.bam')\n",
    "    }\n",
    "    \n",
    "    fragments <- paste0(cellranger_path,samp,'/outs/atac_fragments.tsv.gz')\n",
    "    if (!is.multiome) {\n",
    "        fragments <- paste0(cellranger_path,samp,'/outs/fragments.tsv.gz')\n",
    "    }\n",
    "    \n",
    "    system(paste0('mkdir ',outdir, samp,'/'))\n",
    "    #setwd(paste0(outdir, samp,'/'))\n",
    "    print(paste0('conda run -n windows_multiome_v4 python ', script_path, \" --bam \", bam, \" --fragments \", fragments, \n",
    "                  \" --keep \", outdir, samp, '/', samp, \"_filtered_barcodes.txt -o \", outdir, samp, \"/ &> \", \n",
    "                  outdir, samp, \"/\",samp,\"_log.txt \"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'MM_510'</li><li>'MM_660'</li><li>'MM_661'</li><li>'MM_662'</li><li>'MM_663'</li><li>'MM_664'</li><li>'MM_665'</li><li>'MM_666'</li><li>'MM_667'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'MM\\_510'\n",
       "\\item 'MM\\_660'\n",
       "\\item 'MM\\_661'\n",
       "\\item 'MM\\_662'\n",
       "\\item 'MM\\_663'\n",
       "\\item 'MM\\_664'\n",
       "\\item 'MM\\_665'\n",
       "\\item 'MM\\_666'\n",
       "\\item 'MM\\_667'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'MM_510'\n",
       "2. 'MM_660'\n",
       "3. 'MM_661'\n",
       "4. 'MM_662'\n",
       "5. 'MM_663'\n",
       "6. 'MM_664'\n",
       "7. 'MM_665'\n",
       "8. 'MM_666'\n",
       "9. 'MM_667'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"MM_510\" \"MM_660\" \"MM_661\" \"MM_662\" \"MM_663\" \"MM_664\" \"MM_665\" \"MM_666\"\n",
       "[9] \"MM_667\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Generate Plots and save in a nested set of lists. \n",
    "#Do this for each sample and cleanup each time to reduce memory.\n",
    "\n",
    "dataset <- 'nPOD'\n",
    "modality <- 'Multiome'\n",
    "\n",
    "indir <- paste0(\"/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/\",dataset,'/231121_WE_SSP_Intermediates/Amulet_adatas/')\n",
    "outdir <- paste0(\"/nfs/lab/mega_pancreas/intermediates/Single_Sample_Processing_Intermediates/\",dataset,'/231121_WE_SSP_Intermediates/ATAC_Windows/')\n",
    "script_path <- '/nfs/lab/mega_pancreas/notebooks/single_sample_processing/2ATAC_processing_v4.py'\n",
    "cellranger_path <- paste0('/nfs/lab/projects/mega_pancreas/data/cellranger/', dataset,'/',modality,'/')\n",
    "samp_dir <- paste0(\"/nfs/lab/projects/mega_pancreas/data/cellranger/\",dataset,\"/\", modality, \"/\")\n",
    "\n",
    "samples <- list.dirs(samp_dir, full.names = FALSE, recursive = FALSE)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(data.frame(unlist(parallel::mclapply(samples, run_window_script, outdir, is.multiome=(modality=='Multiome'), \n",
    "                   script_path, cellranger_path, mc.cores=16))),\n",
    "            paste0(outdir, dataset, \"_\", modality, \"_windows_commands_2.sh\"),\n",
    "            row.names=F, col.names=F, quote=F)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Seurat5.0",
   "language": "R",
   "name": "seurat5.0"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
